
## Grouping By Main Effects {.tabset .tabset-fade .tabset-pills}

The following heatmap shows that the transcripts can be grouped into 
six groups with distinct patterns of main effect.

The second tab shows the interactions sorted by main effect groups.
the interaction patterns are also distinct in each group.

```{r main_effects}
n.groups = 6
just.main <- var.inf[,(nrow(var.inf)+1):ncol(var.inf)]
just.main[which(is.na(just.main))] <- 0
#pheatmap(just.main, cutree_rows = n.groups)

dist.mat <- dist(just.main)
clust.mat <- as.dendrogram(hclust(dist.mat))
trans.groups <- cutree(clust.mat, k = n.groups)

group.locale <- lapply(1:max(trans.groups), 
function(x) which(trans.groups == x))

#I tried doing the following with lapply, but there was 
#some weird behavior that I couldn't figure out. Below,
#I just go group by group to sort the transcripts.

ordered.int <- NULL
ordered.main <- NULL
group.mem <- NULL
group.genes <- NULL
nets <- vector(mode = "list", length = n.groups)
for(i in 1:length(group.locale)){
  int.group <- just.int[group.locale[[i]],]
  int.adj <- just.int
  out.group <- setdiff(1:nrow(var.inf), group.locale[[i]])
  int.adj[out.group,] <- 0
  int.adj[which(is.na(int.adj))] <- 0
  net <- graph_from_adjacency_matrix(int.adj, mode = "directed", weighted = TRUE)  
  pos.neg <- rep(1, ecount(net))
  neg.locale <- which(E(net)$weight < 0)
  pos.neg[neg.locale] <- -1
  E(net)$weight <- abs(E(net)$weight)
  E(net)$pos.neg <- pos.neg
  no.deg <- which(degree(net) == 0)
  nets[[i]] <- delete.vertices(net, no.deg)

  main.group <- just.main[group.locale[[i]],]
  ordered.int <- rbind(ordered.int, int.group)
  ordered.main <- rbind(ordered.main, main.group)
  group.mem <- c(group.mem, rep(i, length(group.locale[[i]])))
  group.genes <- c(group.genes, names(group.locale[[i]]))
}
group.df <- data.frame(as.factor(group.mem))
rownames(group.df) <- group.genes
colnames(group.df) <- "group"
```

### Grouped Main Effects

```{r grouped_main, fig.width = 4, fig.height = 7}
pheatmap(ordered.main, cluster_rows = FALSE, cluster_cols = FALSE,
annotation_row = group.df)
```


### Interactions Grouped by Main Effect

The interactions are also distinct among these groups.

```{r grouped_int, fig.width = 7, fig.height = 7}
pheatmap(ordered.int, cluster_rows = FALSE, cluster_cols = FALSE,
annotation_row = group.df)
```

## Transcript Group Networks {.tabset .tabset-fade .tabset-pills}

The following plots show the interaction networks associated with 
each of these groups. They also show three additional plots showing
in-degree, out-degree, and total degree vs. betweenness for each network.
These plots give us an idea of how the transcripts are interacting, and which
might be central to the function of the group. Keep in mind that the different
networks contain many of the same transcripts. The networks interact with each
other.


```{r net_plots, results = "asis", fig.width = 8, fig.height = 8}
for(i in 1:length(nets)){
  
  cat("### Group", i, "\n")
  
  simp.net <- nets[[i]]
  par(mfrow = c(2,2))

  par(mar = c(0,0,0,0))
  edge.weights <- E(simp.net)$weight
  pos.neg <- E(simp.net)$pos.neg
  edge.col <- colors.from.values(edge.weights*pos.neg, split.at.vals = TRUE,
  split.point = 0, col.scale = c("blue", "brown"), grad.dir = "ends")
  plot(simp.net, vertex.size = 3, layout = layout_nicely, edge.color = edge.col)
  par(mar = c(4,4,2,2))
  #barplot(degree(net))
  in.deg <- degree(simp.net, mode = "in")
  out.deg <- degree(simp.net, mode = "out")
  all.deg <- degree(simp.net, mode = "all")
  bet <- betweenness(simp.net)
  
  max.deg <- max(c(in.deg, out.deg, all.deg))
  max.bet <- max(bet)
  xmax <- max.deg*1.05
  ymax <- max.bet*1.05

  plot(in.deg, bet, xlab = "In Degree", ylab = "Betweenness", pch = 16, 
  col = "darkgray", xlim = c(0, xmax), ylim = c(0, ymax))
  text(in.deg, bet, labels = V(simp.net)$name, pos = 3)

  plot(out.deg, bet, xlab = "Out Degree", ylab = "Betweenness", pch = 16, 
  col = "darkgray", xlim = c(0, xmax), ylim = c(0, ymax))
  text(out.deg, bet, labels = V(simp.net)$name, pos = 3)

  plot(all.deg, bet, xlab = "Total Degree", ylab = "Betweenness", pch = 16, 
  col = "darkgray", xlim = c(0, xmax), ylim = c(0, ymax))
  text(all.deg, bet, labels = V(simp.net)$name, pos = 3)

  cat("\n\n")

}
```


## Transcript Group Enrichment

The following heat map shows functional enrichments for the 
different groups of transcripts.

Group1, which has positive effects on ET1 and negative effects on the other
two ETs, is the most obviously related to diabetes. ET1 describes how all 
the traits in the trait matrix vary together. The traits include insulin 
homeostasis measurements, triglycerides, glucose, and islet counts. Group1 
also has the majority of the positive interactions. 

Group2, which has negative effects on ET1 and positive effects on the two 
other ETs, seems to be mostly enriched for transcription factors and the 
Swr1 complex, which is involved in chromatin remodeling. This is the group
that has the bulk of the negative interaction banding.



```{r main_enrich, fig.height = 7, fig.width = 7}
group.enrich <- lapply(group.locale, 
function(x) gost(names(x), organism = "mmusculus"))
names(group.enrich) <- paste0("Group", 1:max(trans.groups))
plot.enrichment.group(group.enrich, n.terms = 40)
```

## Gene group expression with ETs

```{r, eval = FALSE}
expr <- readRDS(here("Results", "all_gene_int", "cross_geno.RData"))

et1 <- data.frame(cross$ET[,1])
colnames(et1) <- "ET1"

layout(get.layout.mat(n.groups))
for(i in 1:length(nets)){
  vnames <- V(nets[[i]])$name
  v.locale <- match(vnames, dimnames(expr)[[3]])
  v.locale <- v.locale[which(!is.na(v.locale))]
  net.expr <- expr[,2,v.locale]  
  #pheatmap(net.expr, show_rownames = FALSE)
  #plot.decomp(t(net.expr))
  #with.ET <- cbind(cross$ET, net.expr)
  ET.cor <- lapply(1:ncol(cross$ET), function(y) apply(net.expr, 2, function(x) cor(x, cross$ET[,y])))
  names(ET.cor) <- colnames(cross$ET)
  stripchart(ET.cor, vertical = TRUE, pch = 16, method = "jitter", main = paste("Group", i))
  #pheatmap(net.expr[order(cross$ET[,1], decreasing = TRUE),], cluster_rows = FALSE, 
  #show_rownames = FALSE, annotation_row = et1, main = paste("Group", i))

  #pheatmap(cor(net.expr))
}

plotEffects(cross, expr, "BC003331_B", "Mrpl19_B", pheno.type = "ET",
plot.type ="l")

```



Further quesetions
  which genes are only in 1 module or the other 
  which genes have connections to both modules

```{r all_verts, eval = FALSE}
i = 1
vnames <- V(et.nets[[i]])$name
all.vert.results <- lapply(vnames, 
function(x) ind_weights(et.nets[[i]], all.net.mods[[i]], x))

#in and out degree going to each module
all.in.out <- t(sapply(all.vert.results, function(x) x[[1]]))
rownames(all.in.out) <- vnames
colnames(all.in.out) <- paste0("Mod", 1:ncol(all.in.out))
pheatmap(all.in.out)

#proportion of total in edges that are from each module
all.mod.in <- t(sapply(all.vert.results, function(x) sapply(1:2, function(y) length(which(x$mods.in.out[[1]] == y))/sum(x$num.in.out))))
rownames(all.mod.in) <- vnames
colnames(all.mod.in) <- paste0("Mod.In.", 1:ncol(all.mod.in))
pheatmap(all.mod.in)

#proportion of total out edges that are to each module
all.mod.out <- t(sapply(all.vert.results, function(x) sapply(1:2, function(y) length(which(x$mods.in.out[[2]] == y))/sum(x$num.in.out))))
rownames(all.mod.out) <- vnames
colnames(all.mod.out) <- paste0("Mod.Out.", 1:ncol(all.mod.out))
pheatmap(all.mod.out)



full.in.out <- cbind(all.mod.in, all.mod.out)
pheatmap(full.in.out)
```

```{r classify}
#which genes are only within module 1?
#these will have edges going to module 1, but 0 going to module 2
with.edges <- apply(full.in.out, 2, function(x) which(x > 0))
no.edges <- apply(full.in.out, 2, function(x) which(x == 0))
```

The following plot shows how expression across modules corresponds
with each ET. ET1 is most highly correlated with average expression
across the modules. It is positively correlated with module 1 and 
negatively correlated with module 2. In other words, individuals with
high expression in module 1 also have high values of ET1. Individuals
with high expression in module 2 have low values of ET2.

```{r module_expression}
expr <- readRDS(here("Results", "all_gene_int", "cross_geno.RData"))
par(mfrow = c(3,2))
for(et in 1:3){
  net.names <- V(et.nets[[et]])$name
  vertex.locale <- match(net.names, dimnames(expr)[[3]])
  vertex.locale <- vertex.locale[which(!is.na(vertex.locale))]
  net.exp <- expr[,2,vertex.locale]

  vert.mods <- all.net.mods[[et]]
  names(vert.mods) <- net.names

  mod.expr <- lapply(unique(vert.mods), function(x) expr[,2,match(rownames(mod.df)[which(vert.mods == x)], dimnames(expr)[[3]])])
  mean.mod <- lapply(mod.expr, function(x) rowMeans(x, na.rm = TRUE))

  for(m in 1:2){
    common.ind <- intersect(rownames(pheno), names(mean.mod[[m]]))
    pheno.locale <- match(common.ind, rownames(pheno))
    mod.locale <- match(common.ind, names(mean.mod[[m]]))
    plot.with.model(mean.mod[[m]][mod.locale], pheno[pheno.locale,et], 
    xlab = paste("Module", m, "Expression"), ylab = paste("ET", et),
    report = "cor.test")
    #plot3d(mean.mod[[1]][mod.locale], mean.mod[[2]][mod.locale], pheno[pheno.locale,et])
  }
}
```

The following figure shows how overall module expression correlates 
with expression of a gene of interest.

```{r gene_mod_expr}
gene.name <- "Ins1"

par(mfrow = c(3,2))
for(et in 1:3){
  net.names <- V(et.nets[[et]])$name
  vertex.locale <- match(net.names, dimnames(expr)[[3]])
  vertex.locale <- vertex.locale[which(!is.na(vertex.locale))]
  net.exp <- expr[,2,vertex.locale]

  vert.mods <- all.net.mods[[et]]
  names(vert.mods) <- net.names

  mod.expr <- lapply(unique(vert.mods), function(x) expr[,2,match(rownames(mod.df)[which(vert.mods == x)], dimnames(expr)[[3]])])
  mean.mod <- lapply(mod.expr, function(x) rowMeans(x, na.rm = TRUE))

  gene.locale <- which(dimnames(expr)[[3]] == gene.name)
  gene.expr <- expr[,2,gene.locale]

  for(m in 1:2){
    common.ind <- intersect(rownames(pheno), names(mean.mod[[m]]))
    pheno.locale <- match(common.ind, rownames(pheno))
    mod.locale <- match(common.ind, names(mean.mod[[m]]))
    plot.with.model(gene.expr[mod.locale], mean.mod[[m]][mod.locale], xlab = gene.name,
    ylab = paste("Module", m, "Expression"))
    }
}
```

## Word cloud enrichment

We can do a related, but independent analysis by looking at wordclouds generated
from pubmed abstracts. Below we find PubMed abstracts that contain each of our
gene names and the word 'pancreas'. 


```{r lit_searc, eval = FALSE}
papers.file <- here("Results", "Abstract_Count.RDS")

if(!file.exists(papers.file)){
  all.pubmed <- lapply(vnames, function(x) get_pubmed_ids(paste(x, "AND (pancreas OR diabetes OR insulin OR glucose OR tryglyceride OR islet) AND english[la]")))
  saveRDS(all.pubmed, papers.file)
}else{
  all.pubmed <- readRDS(papers.file)
}
num.papers <- unlist(lapply(all.pubmed, function(x) x$Count))
with.papers <- which(num.papers != 0)
n.papers <- as.numeric(num.papers[with.papers])
names(n.papers) <- vnames[with.papers]
barplot(sort(n.papers), las = 2, main = "Number of papers per gene")

#wordcloud

has.data <- all.pubmed[with.papers]

all.abstract.file <- here("Results", "abstract.info.all.RDS")
if(!file.exists(all.abstract.file)){
  all.abstracts <- lapply(has.data, function(x) fetch_pubmed_data(x, format = "abstract"))
  names(all.abstracts) <- names(n.papers)
  saveRDS(all.abstracts, all.abstract.file)
}else{
  all.abstracts <- readRDS(all.abstract.file)
}

just.abstract.file <- here("Results", "abstract.info.trimmed.RDS")
if(!file.exists(just.abstract.file)){
  just.abstracts <- lapply(1:length(all.abstracts), 
  function(x) parse.abstracts(abstract.list = all.abstracts[[x]], keyword = names(all.abstracts)[x]))
  names(just.abstracts) <- names(all.abstracts)
  saveRDS(just.abstracts, just.abstract.file)
}else{
  just.abstracts <- readRDS(just.abstract.file)
}


word.bags <- vector(mode = "list", length = length(just.abstracts))

for(i in 1:length(just.abstracts)){
  if(length(just.abstracts[[i]]) > 0){
    word.bags[[i]] <- strsplit(unlist(just.abstracts[[i]]), " ")
  }
}

remove.punct <- c("(", ")", ".", ",", ";", ":", "[", "]")
trimmed.words <- word.bags
for(i in 1:length(trimmed.words)){
  small.bag <- unlist(word.bags[[i]])
  for(j in 1:length(remove.punct)){
    small.bag <- gsub(remove.punct[j], "", small.bag, fixed = TRUE)
  }
 trimmed.words[[i]] <- small.bag
}

```

We now have a bag of words for each gene.
We calculate the term frequency for each gene's bag of words. This is the


```{r, eval = FALSE}
total.bag <- unlist(trimmed.words)


term.freq <- function(small.bag, big.bag){
  small.bag <- str_to_lower(small.bag)
  big.bag <- str_to_lower(big.bag)
  u_small <- unique(small.bag)
  small_count <- sapply(u_small, function(x) length(which(small.bag == x)))
  total_count <- sapply(u_small, function(x) length(which(big.bag == x)))
  term_freq <- small_count/total_count
  return(term_freq[which(is.finite(term_freq))])
}

inverse.doc.freq <- function(all.bags, term){
  term.inc <- lapply(all.bags, function(x) grep(term, x))
  total.appearances <- length(unlist(term.inc))
  total.abstracts <- sum(sapply(all.bags, length))
  idf <- log(total.abstracts/total.appearances)
  return(idf)
}

```

Calculate the term frequencies for the abstracts associated with each gene.

```{r all_term_freq, eval = FALSE}
term.freq.file <- here("Results", "Term.Freq.RDS")
if(!file.exists(term.freq.file)){
  term.freq.by.gene <- lapply_pb(trimmed.words, function(x) table(str_to_lower(unlist(x))))
  term.freq.by.gene <- lapply_pb(trimmed.words, function(x) term.freq(unlist(x), total.bag))
  saveRDS(term.freq.by.gene, term.freq.file)
}else{
  term.freq.by.gene <- readRDS(term.freq.file)
}

u_terms <- unique(total.bag)
idf.file <- here("Results", "IDF.RDS")
if(!file.exists(idf.file)){
  all.term.idf <- unlist(lapply_pb(u_terms, function(x) inverse.doc.freq(just.science, x)))
  saveRDS(all.term.idf, idf.file)
}else{
  all.term.idf <- readRDS(idf.file)
}

names(all.term.idf) <- u_terms
## Individual Vertices of Interest
```

Now we build a big matrix with one row per gene, and one column per term.
Each entry contains the term frequency-inverse document frequency, which is the product
of the term frequency (for each gene) and the inverse document frequency.

```{r term_mat, eval = FALSE}
tfidf.mat <- matrix(0, nrow = length(just.science), ncol = length(u_terms))
rownames(tfidf.mat) <- names(just.science)
colnames(tfidf.mat) <- u_terms

for(i in 1:length(just.science)){
 gene.term.freq <- term.freq.by.gene[[i]]
 term.locale <- match(names(gene.term.freq), u_terms)
 term.idf <- all.term.idf[term.locale]
 tfidf <- gene.term.freq * term.idf
 tfidf.mat[i,term.locale] <- tfidf
}
```

To see how the genes cluster with each other by function, we can do an SVD
on the tf-idf matrix.

```{r decomp_tfidf, eval = FALSE}
decomp <- plot.decomp(tfidf.mat, pc = 5, plot.results = TRUE)
plot(decomp$u[,1:2])
text(decomp$u[,1:2], labels = names(decomp$rows.used))

plot.decomp(t(tfidf.mat), 5)

plot(decomp$v[,1:2])
text(decomp$v[,1:2], labels = colnames(tfidf.mat), cex = 0.5)

boxplot(t(tfidf.mat), las = 2)
gene.cor <- cor(t(tfidf.mat))
diag(gene.cor)  <- 0
pheatmap(gene.cor)
```

## Enrichment stuff from cape module follow-up

## Module Enrichment  {.tabset .tabset-fade .tabset-pills}

The following plots show enrichments for each transcript module.


```{r get_mod_enrich}
all.var <- ls()
  data.loaded <- as.logical(length(which(all.var == "dataset.clinical.phenotypes")))
  if(!data.loaded){
    islets <- load(here("data", "Attie_DO378_eQTL_viewer_v6.Rdata"))
    }
  transcript.info <- dataset.islet.rnaseq$annots
  u_modules <- unique(transcript.info[,"module"])
  module_transcripts <- lapply(u_modules, function(x) transcript.info[which(transcript.info[,"module"] == x),"symbol"])
  names(module_transcripts) <- u_modules
  module.enrich.file <- here("Results", exp.dir, "Module.Enrichment.RDS")
  if(!file.exists(module.enrich.file)){
    module.enrich <- lapply(module_transcripts, function(x) gost(x, "mmusculus", sources = "GO"))
    names(module.enrich) <- paste0("ME", u_modules)
    saveRDS(module.enrich, module.enrich.file)
  }else{
    module.enrich <- readRDS(module.enrich.file)
  }
```

```{r go_sum}
summ_go <- function(enrichment.table, n.terms = 10){
  if(is.null(enrichment.table)){
    return("none")
  }else{
    if(class(enrichment.table) == "list"){
      enrichment.table <- enrichment.table[[1]]
    }
    terms <- enrichment.table[order(-log10(as.numeric(enrichment.table[,"p_value"])), decreasing = TRUE),"term_name"]
    split.terms <- strsplit(terms, " ")
    summ_term <- paste(unique(unlist(split.terms[1:n.terms])), collapse = "_")
    return(summ_term)
  }
}

enrich.summ <- sapply(module.enrich, function(x) summ_go(x, 2))
```


```{r plot_mod_enrich, results = "asis", fig.width = 10, fig.height = 7}

for(i in 1:length(module.enrich)){
  cat("###", names(module.enrich)[i], "\n")
  #quartz(width = 10, height = 7)
  #par(mfrow = c(1,2))
  plot.enrichment(module.enrich[[i]], num.terms = 20, order.by = "p_value",
  plot.label = "")
  
  #split.terms <- unlist(strsplit(module.enrich[[i]]$result[,"term_name"], " "))
  #par(mar = c(0,0,0,0))
  #wordcloud(split.terms)
  cat("\n\n")
}

```

## Module Enrichment Based on Main Effects  {.tabset .tabset-fade .tabset-pills}

We will go through each trait and examine the networks associated with
upregulation and downregulation of each. The following heatmaps show
functional enrichments for ordered lists of transcripts that are associated
with up and down regulation of each eigentrait. There is a lot of overlap 
because the same transcripts are represented in each eigentrait.

```{r trait_networks, results = "asis", fig.height = 7, fig.width = 7}

for(i in 1:ncol(just.main)){
  cat("### ET", i, "\n")
  up.locale <- which(just.main[,i] > 0)
  down.locale <- which(just.main[,i] < 0)

  up.mod.locale <- match(rownames(just.main)[up.locale], names(module.enrich))
  up.mod.locale <- up.mod.locale[which(!is.na(up.mod.locale))]
  down.mod.locale <- match(rownames(just.main)[down.locale], names(module.enrich))
  down.mod.locale <- down.mod.locale[which(!is.na(down.mod.locale))]
  #quartz(width = 7, height = 20)
  all.up <- Reduce("rbind", lapply(module.enrich[up.locale], function(x) x$result))
  all.down <- Reduce("rbind", lapply(module.enrich[down.locale], function(x) x$result))
  
  enrich.group <- list("Up" = all.up, "Down" = all.down)
  plot.enrichment.group(enrich.group, plot.label = colnames(just.main)[i], 
  n.terms = 20, sort.by = "p_value")
  cat("\n\n")
}

```

## Cell Type Specificity {.tabset .tabset-fade .tabset-pills}

We were interested to see which of the genes might be cell-type specific,
and whether cell-specific genes clustered in either module. To do this, 
we downloaded cell-type specific gene lists from 
[GSEA](https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=C8)

We downloaded all text files listing pancreas cell types.

The following plots show the network for ET1 with the vertices
colored by cell type. Alpha and Beta cell-specific transcripts are scattered
pretty randomly throughout the network.

```{r cell_types, results = "asis"}
cell.files <- list.files(here("Data"), pattern = "GeneSet")
cell.types <- sapply(strsplit(cell.files, "_"), function(x) x[2])
cell.genes <- lapply(cell.files, 
function(x) read.table(here("Data", x), header = TRUE, stringsAsFactors = FALSE))
ortho.file <- here("Data", "Cell.Type.Orthologs.RDS")
if(!file.exists(ortho.file)){
  cell.orthos <- lapply(cell.genes, function(x) gorth(x[,1]))
  saveRDS(cell.orthos, ortho.file)
}else{
  cell.orthos <- readRDS(ortho.file)
}

vnames <- V(et.nets[[et]])$name
gene.cell.type <- matrix(0, nrow = length(vnames), ncol = length(cell.types))
rownames(gene.cell.type) <- vnames
colnames(gene.cell.type) <- cell.types
for(i in 1:length(vnames)){
  type.locale <- grep(vnames[i], cell.orthos)
  if(length(type.locale) > 0){
    gene.cell.type[i,type.locale] <- 1
  }
}

simp.net <- et.nets[[et]]
vert.weights <- V(simp.net)$main.effect

mod.df <- data.frame(cbind(as.factor(all.net.mods[[et]]), vert.weights))
rownames(mod.df) <- vnames
colnames(mod.df) <- c("Module", "Main_Effect")
pheatmap(gene.cell.type[order(mod.df[,1]),], cluster_rows =FALSE, annotation_row = mod.df)

edge.weights <- E(simp.net)$weight
pos.neg <- E(simp.net)$pos.neg
edge.col <- colors.from.values(edge.weights*pos.neg, split.at.vals = TRUE,
split.point = 0, col.scale = c("blue", "brown"), grad.dir = "ends")

for(i in 1:length(cell.types)){
  cat("### ", cell.types[i], "\n")
  #quartz()
  vertex.col <- gene.cell.type[,i]*3
  
  plot(simp.net, vertex.size = 5, vertex.color = vertex.col, 
  layout = layout_with_kk, edge.color = edge.col, vertex.label.dist = 1,
  vertex.label.cex = 0.5, edge.arrow.size = 0.3)
  cat("\n\n")
}
```


## Use regular genetic mapping to map CAPE expression modules

Because the two modules are extremely highly correlated, 
we will just look at the eigengene for the whole network.
Mapping the modules separately gives almost identical results.

The plot below shows the LOD plot for the network eigengene.
There is one mQTL on chromosome 18 that has a LOD score greater
than 6. Could this be a trans eQTL hotspot? Could a gene or
genes encoded here be master regulators of our network?

```{r mod_eigengenes}
#the expression is stored as the 'genotype' matrix.
expr <- readRDS(here("Results", exp.name, "cross_geno.RData"))
net.eigengene <- matrix(NA, nrow = nrow(expr), ncol = 1)
rownames(net.eigengene) <- rownames(expr)
colnames(net.eigengene) <- "Network_Eigengene"

net.verts <- V(et.nets[[et]])$name
vert.locale <- match(net.verts, dimnames(expr)[[3]])
vert.locale <- vert.locale[which(!is.na(vert.locale))]
net.expr <- expr[,2,vert.locale]
net.decomp <- plot.decomp(net.expr, plot.results = FALSE)
net.eigengene[,1] <- net.decomp$u[,1]
```


```{r map_mods, fig.width = 10, fig.height = 4}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.clinical.phenotypes")))
if(!data.loaded){
  islets <- load(here("Data", "Attie_DO378_eQTL_viewer_v6.Rdata"))
  kin <- readRDS(here("Data", "cross_kinship.RData"))
  }
covar <- dataset.clinical.phenotypes$covar

mapping.file <- here("Results", exp.name, "Mod_Mapping.RData")
if(!file.exists(mapping.file)){
  mapping.results <- scan1(genoprobs, pheno = net.eigengene[,1], kin = kin, 
  addcovar = covar[,"sexM"])
  saveRDS(mapping.results, mapping.file)
}else{
  mapping.results <- readRDS(mapping.file)
}

plot(mapping.results, map = map)
```

To investigate this further, we performed an eQTL scan for 
all the transcripts in the network and looked at the cis-trans plot.

(transcripts on Chr 18?)

```{r eqtl}
peak.table <- find_peaks(mapping.results, map =  map, threshold = 6, drop = 1.3)

for(i in 1:nrow(peak.table)){
  peak.chr <- as.numeric(peak.table[i,"chr"])
  chr.locale <- which(cross$chromosome == peak.chr)
  chr.markers <- cross$geno_names[[3]][chr.locale]
  chr.coord <- as.numeric(cross$marker_location[chr.locale])
  peak.min <- peak.table[i,"ci_lo"]
  peak.max <- peak.table[i,"ci_hi"]
  chr.region <- intersect(which(chr.coord >= peak.min), which(chr.coord <= peak.max))
  chr.trans <- chr.markers[chr.region]
  expr.locale <- match(chr.trans, dimnames(expr)[[3]])
  chr.locale <- which(names(genoprobs) == peak.chr)

  mapping.file <- here("Results", exp.name, 
  paste0("Peak.Transcript.Mapping.Chr", peak.chr, ".", signif(peak.min, 2), 
  ".to.", signif(peak.max, 2)))

  if(!file.exists(mapping.file)){
    eqtl.results <- lapply(expr.locale[1:5], 
    function(x) scan1coef(genoprobs[,chr.locale], pheno = expr[,2,x], kin = kin[[chr.locale]], 
    addcovar = covar[,"sexM"], cores = 3))
    #i = 2
    #plot_coefCC(eqtl.results[[i]], map = map, main = chr.trans[i])
    saveRDS(eqtl.results, mapping.file)
  }else{
    eqtl.results <- readRDS(mapping.file)
  }

}

```

## Haplotype coefficients {.tabset .tabset-fade .tabset-pills}

The haplotype coefficients of many of the transcripts in the 
network match the coefficients in the ET, but are not driving 
the pattern. Coefficient patterns match across the whole chromosome,
without there being cis eQTL. Does this mean the transcripts are
downstream of another process that is driving the eigentrait and 
the gene expression?

```{r coef, eval = TRUE, results = "asis", fig.width = 10, fig.height = 4.5}
transcript.chr  <- get_marker_chr(cross, rownames(var.inf))

for(chr in 1:19){
  cat("### Chr", chr, "\n")

  chr.trans <- rownames(var.inf)[which(transcript.chr == chr)] 
  chr.locale <- which(names(genoprobs) == chr)
  trans.locale <- match(chr.trans, dimnames(expr)[[3]])
  trans.pos <- as.numeric(get_marker_location(cross, chr.trans))

  chr.map.file <- here("Results", exp.name, paste0("Chr.", chr, ".Mapping.RDS"))
  if(!file.exists(chr.map.file)){
    mapping.results <- vector(mode = "list", length = 3)
    names(mapping.results) <- c("Eigentrait", "Net_Eigengene", "Transcripts")

    #map the first eigentrait  
    mapping.results[[1]] <- scan1coef(genoprobs[,chr.locale], pheno = cross$ET[,1],
    kin = kin[[chr.locale]], addcovar = covar[,"sexM"])

    #map network eigengene
    mapping.results[[2]] <- scan1coef(genoprobs[,chr.locale], pheno = net.eigengene, 
    kin = kin[[chr.locale]], addcovar = covar[,"sexM"])

    #Map the local transcripts
    mapping.results[[3]] <- lapply(trans.locale, 
    function(x) scan1coef(genoprobs[,chr.locale], pheno = expr[,2,x], 
    kin = kin[[chr.locale]], addcovar = covar[,"sexM"]))
  
    saveRDS(mapping.results, chr.map.file)
  }else{
    mapping.results <- readRDS(chr.map.file)
  }
  
  #quartz(width = 10, height = 4.5)
  par(mar = c(4,4,2,2))
  plot_coefCC(mapping.results$Eigentrait, map = map, main = "ET 1")
  plot_coefCC(mapping.results$Net_Eigengene, map = map, main = "Net Eigengene")
  trans.results <- mapping.results$Transcripts
  for(tr in 1:length(trans.results)){
    plot_coefCC(trans.results[[tr]], map = map, main = chr.trans[tr])
    abline(v = trans.pos[tr])
  }

  cat("\n\n")
}

```

