---
title: "CAPE with Transcripts as Genotypes"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---


The purpose of this workflow is to run CAPE using the output from 
cape_setup_expression.Rmd. That workflow creates different data sets
using gene expression in place of, or in addition to, true genotypes.

```{r setup}
set.seed(2398347)
library("here")
```

```{r set_params}
#geno.type = "transcripts"
geno.type = "modules"
#exp.name = "multi_tissue_test"
exp.name = "CCA"
ex.path <- here("Results", "CAPE", paste(geno.type, exp.name, sep = "_"))

#args <- commandArgs(trailingOnly=T)
#delete.old.results <- as.logical(args[1])
#delete.old.results = FALSE
delete.old.results <- TRUE
```

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, error = FALSE, message = FALSE, echo = FALSE}
needed.packages <- c("cape", "pheatmap", "gprofiler2")
load_libraries(needed.packages)
```


```{r load_data}
data.obj <- readRDS(here("Data", paste0("cape_data_", geno.type, ".RDS")))
geno.obj <- readRDS(here("Data", paste0("cape_geno_", geno.type, ".RDS")))
geno.names <- dimnames(geno.obj)
names(geno.names) <- c("mouse", "allele", "locus")
data.obj$geno_names <- geno.names
#tail(data.obj$geno_names[[3]], 15)
```


Subsample the cross for testing if requested

```{r num}
cross <- data.obj
geno <- geno.obj

num.pheno <- apply(cross$pheno, 2, as.numeric)
dimnames(num.pheno) <- dimnames(cross$pheno)
cross$pheno <- num.pheno
```


```{r run.cape}
#remove all previous file except the 
#kinship file and the yml file
if(delete.old.results){
    results.files <- list.files(ex.path, full.names = TRUE)
    to.keep <- file.path(ex.path, "cape.parameters.yml")
    to.delete <- setdiff(results.files, to.keep)
    if(length(to.delete) > 0){
        unlink(to.delete)
    }
}

#copy over kinship matrix in case we want to use it
link.line <- paste("ln -s", here("Data", "cross_kinship.RData"), ex.path)
system(link.line)

sink(file.path(ex.path, "progress.txt"))

final.cross <- run_cape(pheno_obj = cross, geno_obj = geno, p_or_q = 0.05, 
verbose = TRUE, results_path = ex.path, 
param_file = file.path(ex.path, "cape.parameters.yml"))

sink(NULL)

final.cross <- readRDS(file.path(ex.path, "cross.RData"))

pdf(file.path(ex.path, "Variant_Influences_Labeled.pdf"), width = 18, height = 12)
par(mar = c(35, 35, 4, 4))
plot_variant_influences(final.cross, show_marker_labels =TRUE)
dev.off()
```

