---
title: "Adjusting Transcripts and Traits for Further Analysis"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction
The purpose of this workflow is to set up transcripts and traits 
for downstream analysis. 
Both cluster_transcripts_CCA.Rmd and cluster_transcripts.Rmd operate
on the same adjusted traits and transcripts. This workflow sets up 
those files for both analyses.

```{r get_args}
args <- commandArgs(trailingOnly=T)
exp.name = args[1]
delete_previous <- as.logical(args[2])

if(is.na(delete_previous)){
    #exp.name = "all_traits"; weight.adjusted = FALSE
    #exp.name = "weight_adjusted"; weight.adjusted = TRUE
    exp.name = "imputed_genetic"; weight.adjusted = FALSE
    #exp.name = "imputed_local"; weight.adjusted = FALSE
    delete_previous <- FALSE
}

#traits to analyze can either be full names, a pattern to search for, like "TG",
#or numeric values indicating the column numbers of the traits.
#the analysis results will be put in a folder named for exp.name
#adj.traits are full names, partial names, or column numbers of 
#traits to adjust out of the phenotype matrix. This could be weight,
#for example, or NULL to not adjust for anything

if(weight.adjusted){
    keep.traits <- 1:20 #all traits;
    adj.traits <- "weight" #traits to adjust for, set to NULL to not adjust
}else{
    keep.traits <- 1:20 #all traits;
    adj.traits <- NULL #traits to adjust for, set to NULL to not adjust
}

use.imputed <- as.logical(length(grep("imputed", exp.name)))
```

```{r load_code}
is.interactive = FALSE
#is.interactive = TRUE
library("here")

results.dir <- here("Results", "CCA_Clusters", exp.name)
if(!file.exists(results.dir)){dir.create(results.dir)}

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r delete_previous}
if(delete_previous){
    results.files <- list.files(results.dir, full.names = TRUE)
    unlink(results.files)
}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "gprofiler2", "cape", "PMA", "abind", "cluster",
"RColorBrewer", "igraph", "corpcor", "easyPubMed", "knitr", "kableExtra",
"ape")
load_libraries(all.packages)
```


```{r read_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.clinical.phenotypes")))
if(!data.loaded){
    gen.data <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
    tissue.files <- list.files(here("Data", "RDS_datasets_tissues"), full.names = TRUE) #DO measured expression
    tissue.names <- gsub(".RDS", "", basename(tissue.files))
    tissue.data <- lapply(tissue.files, readRDS)
    names(tissue.data) <- tissue.names
    tissue.covar <- lapply(tissue.data, function(x) x$covar.matrix)
    data.set.names <- lapply(1:length(tissue.names), 
    function(x) paste0("dataset.DO.Cube.", tissue.names[x]))
    gene.table <- as.matrix(unique(Reduce("rbind", lapply(data.set.names, 
        function(x) get(x)$annot.mrna))))
}
```

## Select Traits

Extract clinical traits, expression traits, and covariates from the 
islet data set. Adjust for covariates.

Here we also select the traits for analysis and adjust for any traits
that we want to use as covariates, for example, weight.

```{r select_traits}
pheno <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE))
num.pheno <- apply(pheno[,11:(ncol(pheno)-1)], 2, as.numeric)
rownames(num.pheno) <- pheno[,1]

max.n <- which.max(sapply(tissue.covar, nrow)) #adjust with the largest covariate matrix
covar <- tissue.covar[[max.n]]

if(is.numeric(keep.traits[1])){
    keep.trait.locale <- keep.traits
}else{
    keep.trait.locale <- sapply(keep.traits, function(x) grep(x, colnames(num.pheno), ignore.case = TRUE))
}

if(length(adj.traits) > 0){
    adj.trait.locale <- as.vector(sapply(adj.traits, function(x) grep(x, colnames(num.pheno), ignore.case = TRUE)))
    keep.trait.locale <- setdiff(keep.trait.locale, adj.trait.locale)
    all.covar <- Reduce("cbind", get.xz(covar, num.pheno[,adj.trait.locale]))
}else{
    all.covar <- covar
}

#remove batch effects from transcript data and 
#phenotype data.
common.ind <- intersect(rownames(num.pheno), rownames(covar))
pheno.idx <- match(common.ind, rownames(num.pheno))
covar.idx <- match(common.ind, rownames(covar))

adj.pheno <- adjust(matX = num.pheno[,keep.trait.locale], 
    adj.mat = all.covar, retain.intercept = TRUE)

cca.pheno <- apply(adj.pheno, 2, scale)
rownames(cca.pheno) <- rownames(adj.pheno)

#save these in data so they are easily accessible to other workflows
pheno.file <- here("Data", "Adjusted_Phenotypes.RDS")
saveRDS(cca.pheno, pheno.file)

covar.file <- here("Data", "Covariates.RDS")
saveRDS(all.covar, covar.file)
```

## Adjust Expression for Covariates

Adjust expression for sex and generation. If we are using 
imputed data, read it in here and overwrite the measured
expression data.

```{r adjust_covar}
adj.expr.file <- here("Data", "imputed", paste0("Adjusted_Expression_", exp.name, ".RDS"))

if(use.imputed){
    if(exp.name == "imputed_local"){
        imp.expr.files <- get.files(here("Data", "imputed"), want = c("RDS", "local"), 
            dont.want = c("CC", "Adjusted"), full.names = TRUE)
    }
    if(exp.name == "imputed_full"){
        imp.expr.files <- get.files(here("Data", "imputed"), want = c("RDS", "genetic"), 
            dont.want = c("adjusted", "CC"), full.names = TRUE)
    }
    tissue.norm <- lapply(imp.expr.files, readRDS)
    #remove any transcripts without any data
    for(i in 1:length(tissue.norm)){
        has.data <- which(apply(tissue.norm[[i]], 2, function(x) !all(is.na(x))))
        tissue.norm[[i]] <- tissue.norm[[i]][,has.data]
    }
}else{
    tissue.norm <- lapply(tissue.data, function(x) x$data$norm)
}

if(is.interactive){
#test a few to make sure we're not mapping to the end of the chromosome
    tx = 1
    rnd.id  <- sample(colnames(tissue.norm[[tx]]), 1)
    #rnd.id = colnames(tissue.norm[[tx]])[1]
    measured.expr <- adjust(tissue.data[[tx]]$data$norm[,rnd.id,drop=FALSE], tissue.data[[tx]]$covar.matrix)
    imp.expr <- adjust(tissue.norm[[tx]][,rnd.id,drop=FALSE], tissue.data[[tx]]$covar.matrix)
    #imp.expr <- tissue.norm[[tx]][,rnd.id,drop=FALSE]
    rnd.chr <- gene.table[match(rnd.id, gene.table[,1]), "chr"]
    rnd.pos <- as.numeric(gene.table[match(rnd.id, gene.table[,1]), "start"])
    rnd.name <- gene.table[match(rnd.id, gene.table[,1]), "symbol"]
    test.imp <- scan1(genoprobs[,rnd.chr], imp.expr, kinship = K[[rnd.chr]])
    test.measured <- scan1(genoprobs[,rnd.chr], measured.expr, kinship = K[[rnd.chr]])
    #test <- scan1(genoprobs, tissue.norm[[tx]][,rnd.id], kinship = K);plot(test, map = map)
    ylim <- c(0, max(c(test.imp, test.measured)))

    quartz(width = 8, height = 8)
    par(mfrow = c(2,2))
    plot(test.measured, map = map, main = paste(rnd.name, "Chr", rnd.chr))
    abline(v = rnd.pos)
    plot(test.imp, map = map, main = "Imputed")
    abline(v = rnd.pos)
    legend("topleft", col = c("red", "darkblue"), lty = 1, legend = c("measured", "imputed"), lwd = 2)

    matched.var <- get.xz(imp.expr, measured.expr)
    plot.with.model(matched.var[[1]], matched.var[[2]], 
        xlab = "Imputed Expression", ylab = "Measured Expression", col = tissue.data[[tx]]$covar.matrix[,1]+1)
}

#if we haven't generated the adjusted expression file yet,
#create it here. Otherwise, read it in.
if(!file.exists(adj.expr.file)){
    #adjust expression for covarates
    adj.expr <- lapply(tissue.norm, function(x) adjust(x, all.covar, TRUE))
    names(adj.expr) <- c("Adipose", "Islet", "Liver", "SkeletalMuscle")
    saveRDS(adj.expr, adj.expr.file)
}else{
    adj.expr <- readRDS(adj.expr.file)
}
```


## Tissue Gene Expression Decomposition

```{r expr_decomp}
transcript.list <- lapply(adj.expr, colnames)

if(is.interactive){quartz()}
plotVenn(transcript.list)

common.transcripts <- Reduce("intersect", transcript.list)
common.transcript.idx <- lapply(adj.expr, function(x) match(common.transcripts, colnames(x)))
common.ind <- Reduce("intersect", lapply(adj.expr, rownames))
common.ind.idx <- lapply(adj.expr, function(x) match(common.ind, rownames(x)))

matched.expr <- lapply(1:length(adj.expr), 
function(x) adj.expr[[x]][common.ind.idx[[x]], common.transcript.idx[[x]]])

tissue.label <- unlist(lapply(1:length(matched.expr), 
function(x) rep(names(adj.expr)[x], ncol(matched.expr[[x]]))))

all.expr <- Reduce("cbind", matched.expr)

expr.decomp.file <- here(results.dir, paste0("all.expr.decomp.", exp.name, ".RDS"))
if(!file.exists(expr.decomp.file)){
    expr.decomp <- plot.decomp(t(all.expr), plot.results = FALSE)
    saveRDS(expr.decomp, expr.decomp.file)
}else{
    expr.decomp <- readRDS(expr.decomp.file)
}

plot(expr.decomp$u, col = as.numeric(as.factor(tissue.label)), pch = 16,
xlab = "PC1", ylab = "PC2")
legend("topleft", legend = levels(as.factor(tissue.label)), 
col = 1:length(unique(tissue.label)), pch = 16)

xlim <- c(min(expr.decomp$u[,1]), max(expr.decomp$u[,1]))
ylim <- c(min(expr.decomp$u[,2]), max(expr.decomp$u[,2]))
par(mfrow = c(2,2))
for(i in 1:length(adj.expr)){
    tissue.locale <- which(tissue.label == names(adj.expr)[i])
    plot(expr.decomp$u[tissue.locale,], pch = 16, xlim = xlim, ylim = ylim,
    xlab = "PC1", ylab = "PC2", main = names(adj.expr)[i])
}
```

## Transcript Correlation

How well does the expression of a single transcript in one 
tissue correlate with its expression in another tissue?

The following plot shows the distribution of correlations across
individuals of transcript i in one tissue with transcript i in another 
tissue. 

Transcripts in one tisse are expressed very differently across 
individuals than the same transcript in another tissue.

```{r plot_transcript_cor, fig.height = 10, fig.width = 3.5}
tissue.pairs <- pair.matrix(1:length(tissue.names))
tissue.expr.comp <- vector(mode = "list", length = nrow(tissue.pairs))
for(i in 1:nrow(tissue.pairs)){
    tissue.expr.comp[[i]] <- sapply(1:ncol(matched.expr[[1]]),
    function(x) cor(matched.expr[[tissue.pairs[i,1]]][,x], matched.expr[[tissue.pairs[i,2]]][,x]))
}

if(is.interactive){quartz(height = 9, width = 3.5)}
par(mfrow = c(length(matched.expr),1), mar = c(0, 2, 0, 2))
for(i in 1:length(matched.expr)){
    tissue1 <- tissue.names[tissue.pairs[i,1]]
    tissue2 <- tissue.names[tissue.pairs[i,2]]
    a <- hist(tissue.expr.comp[[i]], xlim = c(-1, 1), axes = FALSE, xlab = "", 
    ylab = "", main = "", breaks = 100, border = NA)
    points(x = median(tissue.expr.comp[[i]]), y = 0, pch = "*", 
    col = "#2b8cbe", cex = 3)
    text(x = -0.1, y = max(a$counts)*0.75, paste(tissue1, tissue2, sep = "-"), 
    adj = 1, cex = 2)
}
axis(1, cex.axis = 2)
par(xpd = NA)
abline(v = 0, col = "darkgray", lty = 2)
par(xpd = FALSE)
```

## Tissue Pairs

How well are the transcript means correlated across tissues? Are transcripts
that are highly expressed in one tissue also highly expressed in another
tissue? Do transcripts with low mean expression in one tissue also have
low mean expression in another tissue?

The following plots show how well mean expression for common transcripts
align across tissue pairs. Transcripts in different tissues have different
relative expression levels.

In each case below the expression of the tissues mean compared are adjusted
to have the same overall mean.


```{r tissue.correlation, fig.height = 8, fig.width = 8}
par(mfrow = c(2,2))
mean.tx <- sapply(adj.expr, colMeans)
for(i in 1:nrow(tissue.pairs)){
    tissue1.idx <- tissue.pairs[i,1]
    tissue2.idx <- tissue.pairs[i,2]
    tissue1.mean <- mean.tx[[tissue1.idx]]
    tissue2.mean <- mean.tx[[tissue2.idx]]
    tissue.means <- c(mean(tissue1.mean), mean(tissue2.mean))
    overall.mean <- mean(tissue.means)
    mean.diffs <- tissue.means - overall.mean
    tissue1.mean <- tissue1.mean - mean.diffs[1]
    tissue2.mean <- tissue2.mean - mean.diffs[2]

    common.tx <- intersect(names(tissue1.mean), names(tissue2.mean))

    plot.with.model(tissue1.mean[common.tx], tissue2.mean[common.tx], 
        main = paste(tissue.names[tissue1.idx], "vs.", tissue.names[tissue2.idx]),
        xlab = tissue.names[tissue1.idx], ylab = tissue.names[tissue2.idx],
        report = "cor.test")
}

```

## eQTL and Trait correlations {.tabset .tabset-fade .tabset-pills}

The following plots show a clear negative correlation between 
the strength of the eQTL and the correlation with traits. The 
correlation shown below is the maximum correlation across all
traits.

The genes shown are the genes with at least a minimum correlation 
threshold, ranked by LOD score and correlation. These are genes
with high LOD scores that are also correlated with at least one 
trait.

These lists show us that some of the genes with high correlations
also have eQTL that are considered significant, even if they are
not the highest LOD scores in the bunch. These are good candidates
for genes that influence traits in a way that we can trace back to
genotype through gene expression.

```{r eqtl, fig.width = 10, fig.height = 6, results = "asis"}
if(!use.imputed){
    get_max_eqtl <- function(gene.id){
        gene.locale <- which(tissue.eqtl[[1]] == gene.id)
        if(length(gene.locale) == 0){return(NA)}
        max.lod <- max(tissue.eqtl[[3]][gene.locale])
        return(max.lod)
    }
    cor.thresh = 0.4
    top.n <- 20

    for(i in 1:length(tissue.names)){
        cat("###", tissue.names[i], "\n")
        if(is.interactive){quartz(width = 10, height = 6)}
        layout.mat <- matrix(c(1:5, 0), nrow = 2, byrow = TRUE)
        layout(layout.mat)
        expr.trait.cor.file <- here("Results", "General", paste0("Trait.", tissue.names[i], ".Expr.Cor.RDS"))
        tissue.expr <- adj.expr[[i]]
        tissue.data <- get(paste0("dataset.DO.Cube.", tissue.names[i]))
        tissue.eqtl <- tissue.data$lod.peaks$additive
        common.ind <- intersect(rownames(cca.pheno), rownames(tissue.expr))
        expr.locale <- match(common.ind, rownames(tissue.expr))
        pheno.locale <- match(common.ind, rownames(cca.pheno))

        if(!file.exists(expr.trait.cor.file)){
            trait.cor <- apply(cca.pheno[pheno.locale,], 2, function(y) apply(tissue.expr[expr.locale,], 2, function(x) cor(x, y, use = "pairwise.complete.obs")))
            saveRDS(trait.cor, expr.trait.cor.file)
        }else{
            trait.cor <- readRDS(expr.trait.cor.file)
        }
        
        cor.mean <- apply(trait.cor, 2, function(x) mean(abs(x)))
        par(mar = c(4,8,2,2))
        boxplot(abs(trait.cor[,order(cor.mean)]), las = 2, horizontal = TRUE, 
            main = "Trait-Expression Correlation")

        par(mar = c(4,8,2,2))
        #plot.decomp(t(trait.cor), label.points = TRUE, main = "Decomposition of Trait-Expression Correlation")

        max.eqtl.file <- here("Results", "General", paste0("max.eQTL.LOD.", tissue.names[i], ".RDS"))
        if(!file.exists(max.eqtl.file)){
            max.eqtl <- sapply(colnames(tissue.expr), get_max_eqtl)
            saveRDS(max.eqtl, max.eqtl.file)
        }else{
            max.eqtl <- readRDS(max.eqtl.file)
        }
        
        max.cor <- apply(abs(trait.cor), 1, max)
        plot.with.model(max.eqtl, max.cor, report = "cor.test")
        abline(h = cor.thresh, col = "red")

        #ID genes with correlations > 0.4 and relatively high LOD scores
        high.cor <- which(max.cor >= cor.thresh)
        high.table <- cbind(max.cor[high.cor], max.eqtl[high.cor])
        no.na <- which(!is.na(high.table[,2])) #take out genes with LOD scores < 6
        high.table <- high.table[no.na,]
        gene.ranks <- rank.2D(high.table)
        ranked.genes <- high.table[order(gene.ranks[,1], decreasing = FALSE),]
        top.genes <- ranked.genes[1:min(c(top.n, nrow(high.table))),]
        #plot names for the top 10 genes
        gene.names <- tissue.data$annot.mrna[[2]][match(rownames(top.genes), tissue.data$annot.mrna[[1]])]
        rownames(top.genes) <- gene.names
        colnames(top.genes) <- c("Max.Trait.Cor", "Max.LOD.Score")
        #cat(gene.names, sep = "\n")    

        gene.text <- paste(rownames(top.genes), collapse = "\n")
        cor.text <- paste(round(top.genes[,1], 2), collapse = "\n")
        lod.text <- paste(round(top.genes[,2], 1), collapse = "\n")
        par(mar = c(0,0,0,0))
        plot.text(paste("Top", top.n, "genes\n\n", gene.text), x = 0.35, add = FALSE)
        plot.text(paste("Corr\n\n", cor.text), x = 0.5, add = TRUE)
        plot.text(paste("LOD\n\n", lod.text), x = 0.6, add = TRUE)

        trait.eqtl.cor <- apply(abs(trait.cor), 2, 
            function(x) cor(max.eqtl, x, use = "pairwise.complete.obs"))
        
        #par(mfrow = c(4,5), mar = c(1,1,4,1))
        #for(tr in 1:ncol(trait.cor)){
        #    plot(max.eqtl, abs(trait.cor[,tr]), main = colnames(trait.cor)[tr])
        #}

        par(mar = c(8,4,2,2))
        barplot(trait.eqtl.cor, las = 2, main = "Correlation between LOD\nscore and trait correlation")
        
        #plot(max.eqtl, abs(trait.cor[,3]), main = tissue.names[i], xlab = "Maximum eQTL", ylab = "Trait Correlation Magnitude")
        expr.var <- apply(tissue.expr, 2, var)
        
        trait.var.cor <- apply(abs(trait.cor), 2, function(x) cor(expr.var, x, use = "pairwise.complete.obs"))
        #plot.with.model(expr.var, abs(trait.cor[,1]), main = tissue.names[i], xlab = "Expression Variance", ylab = "Trait Correlation Magnitude", report = "cor.test")
        #plot(expr.var, trait.cor[,1], xlab = "Maximum eQTL", ylab = "Variation in Expression", main = tissue.names[i])
        par(mar = c(8,4,2,2))
        barplot(trait.var.cor, las = 2, main = "Correlation between expression\nvariance and trait correlation")

        mtext(tissue.names[i], side = 3, outer = TRUE, line = -1)
        cat("\n\n")
    }
}
```