---
title: "Prioritize with trans eQTL"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction
The purpose of this workflow is to use TRiAGE to prioritize
plausible causal candidates in QTLs that map to traits and 
are trans bands, like the Chr 11 QTL.

We are also interested in whether the first PC of the trans
band is correlated with any of the traits.
    
The network used is the same as the specified tissue. 


```{r get_args}
args <- commandArgs(trailingOnly=T)
tissue.name = args[1]
n.cores <- as.numeric(args[2])
training.set <- args[3]

if(is.na(tissue.name) || length(tissue.name) == 0){
    tissue.name = "Adipose"
    n.cores = 4
    #training.set = "Trans-band"
    training.set = "SOM"
}

trait.name = "Ins_tAUC"
qtl.chr = 11
exp.name = paste0(tissue.name, "_Chr", qtl.chr, "_", training.set)
project.name = paste0("Chr", qtl.chr, "_QTL")

#delete_previous <- TRUE
delete_previous <- FALSE

network.names <- c("Adipose" = "adipose tissue", 
                    "Islet" = "islets of langerhans", 
                    "Liver" = "liver", 
                    "SkeletalMuscle" = "skeletal muscle")

network.name = network.names[which(names(network.names) == tissue.name)]
Mb.buffer = 1
```

```{r load_code}
is.interactive = FALSE
#is.interactive = TRUE
library("here")

data.dir <- here("Results", "CCA_Clusters", "all_traits")
results.dir <- here("Results", "Prioritization", exp.name)
if(!file.exists(results.dir)){dir.create(results.dir)}

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}

needed.libraries <- c("httr", "Matrix", "knitr", "here", "igraph", "DescTools", "pheatmap", 
"e1071", "parallel", "doParallel", "gprofiler2", "DT", "qtl2", "XML", "RCurl", 
"R.utils", "plotly") 

load_libraries(needed.libraries, personal.library = TRUE)

check_bmediatR <- require("bmediatR")

if(!check_bmediatR){
    lib.loc <- Sys.getenv("R_LIBS_USER")
    devtools::install_github("wesleycrouse/bmediatR", lib = lib.loc, build_vignettes = FALSE)
    library(bmediatR)
}

```


```{r load_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.DO.Cube.Adipose")))

if(!data.loaded){
    tissue.data <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
}
traits <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE))
num.traits <- apply(traits[,c(12:(ncol(traits)-1))], 2, as.numeric)
rownames(num.traits) <- traits[,1]

data.set <- get(paste0("dataset.DO.Cube.", tissue.name))
covar.mat <- data.set$covar.matrix
expr.data <- data.set$data$norm

adj.expr <- adjust(expr.data, covar.mat)
adj.pheno <- adjust(num.traits, covar.mat)

gene.table <- data.set$annot.mrna
eqtl.table <- data.set$lod.peaks$additive
```

The figure below shows the genetic mapping of `r trait.name`.
We define the QTL as the area between the red lines. We are
using a pretty broad definition of the QTL to make sure we 
capture all relevant trans-regulated transcripts.

```{r get_plot_idx}
#finds the x position of a mapping position on an
#existing QTL plot or the full genome.

get_plot_idx <- function(map, chr, Mb){
    full.map <- length(unlist(map))
    pt.marker <- find_marker(map, chr = chr, pos = Mb)
    idx <- which(unlist(lapply(map, names)) == pt.marker)
    plot.dim <- par("usr")
    x.coord <- (idx/full.map)*plot.dim[2]
    return(x.coord)
}

```

```{r trait_scan, fig.width = 8, fig.height = 5}
trait.scan <- scan1(genoprobs, adj.pheno[,trait.name], kinship = K)
peak.table <- find_peaks(trait.scan, map = map, threshold = 6, drop = 6)
chr.locale <- which(peak.table[,"chr"] == qtl.chr)
chr.peak <- peak.table[chr.locale,]
qtl.start <- (floor(chr.peak[,"ci_lo"])-Mb.buffer)*1e6
qtl.end <- (ceiling(chr.peak[,"ci_hi"])+Mb.buffer)*1e6

plot(trait.scan, map = map)
qtl.start.x <- get_plot_idx(map, chr = 11, Mb = qtl.start/10^6)
qtl.end.x <- get_plot_idx(map, chr = 11, Mb = qtl.end/10^6)
abline(v = c(qtl.start.x, qtl.end.x), col = "red", lwd = 2)
```

## cis-trans eQTL

The figure below shows the structure of the cis and
trans eQTL by loci across the genome.

```{r cis_trans, fig.width = 10, fig.height = 7}
split.markers <- strsplit(eqtl.table[[2]], "_")
eqtl.chr <- sapply(split.markers, function(x) x[1])
eqtl.pos <- as.numeric(sapply(split.markers, function(x) x[2]))/1e6
eqtl.info <- cbind(eqtl.table$gene.id, eqtl.chr, eqtl.pos, eqtl.table$lod)

transcript.info <- cbind(gene.table$gene.id, gene.table$chr, gene.table$start)
#pdf("~/Desktop/test.pdf")
plot.cistrans.table(eqtl.info, transcript.info, map = map)
#dev.off()
```

## eQTL counts {.tabset .tabset-fade .tabset-pills}

The following plot shows counts of eQTL across the genome. 

```{r eqtl_count, results = "asis"}
eqtl.chr <- eqtl.info[,2]
eqtl.chr[which(eqtl.chr == "X")] <- 20
mb.window = 4
mb.gap = 2
u_chr <- sort(unique(as.numeric(eqtl.chr)))
eqtl.counts <- vector(mode = "list", length = length(u_chr))
eqtl.position <- vector(mode = "list", length = length(u_chr))
for(ch in 1:length(u_chr)){
    chr.idx <- which(eqtl.chr == u_chr[ch])
    chr.table <- eqtl.info[chr.idx,]
    sorted.table <- chr.table[order(as.numeric(chr.table[,3]), decreasing = FALSE),]
    num.pos <- as.numeric(sorted.table[,3])
    pos.bins <- sliding.window.el(num.pos, window.size = mb.window, gap.size = 2)
    eqtl.count <- sapply(pos.bins, function(x) length(intersect(which(num.pos >= min(x)), which(num.pos < max(x)))))
    pos.mid <- sapply(pos.bins, mean)
    eqtl.counts[[ch]] <- eqtl.count
    eqtl.position[[ch]] <- pos.mid
}
max.count <- max(unlist(eqtl.counts))

for(ch in 1:length(eqtl.counts)){
    cat("### Chr", ch, "\n")
    if(is.interactive){quartz(width = 10, height = 6)}
    plot(eqtl.position[[ch]], eqtl.counts[[ch]], ylim = c(0, max.count),
    main = paste("Chromosome", ch), type = "h", lwd = 3, 
    xlab = paste("Chr", ch, "Position (Mb)"), ylab = "eQTL count")
    cat("\n\n")
}

```

## Transcript Training Set Identificaction

For the trans band prioritization, we identified all transcripts 
that mapped to this interval using the data in the QTLviewer object.

For the SOM prioritization, we identified all transcripts that
were differentially expressed based on their genotype at the
Chr 11 locus. We compared those with up alleles at this locus 
to those with down alleles at this locus. See SOM_DO.Rmd for 
more details.

```{r interval_transcripts}
if(training.set == "Trans-band"){
    eqtl.markers <- strsplit(eqtl.table[[2]], "_")
    eqtl.chr <- sapply(eqtl.markers, function(x) x[1])
    eqtl.pos <- as.numeric(sapply(eqtl.markers, function(x) x[2]))
    chr.idx <- which(eqtl.chr == chr.peak[,"chr"])
    above.min <- which(eqtl.pos >= qtl.start)
    below.max <- which(eqtl.pos <= qtl.end)
    trans.idx <- Reduce("intersect", list(chr.idx, above.min, below.max))
    query.table <- eqtl.table[trans.idx,]
}
if(training.set == "SOM"){
    #fdr.thresh = 0.1
    pval.thresh = 0.01
    diff.genes <- read.csv(here("Results", "SOM", paste0("DO_", tissue.name, " - Results"), 
        "Summary Sheets - Differences", "CSV Sheets", "Gene Lists - Global", "down.vs.up.csv"), 
        skip = 17)
    pval <- as.numeric(sapply(strsplit(diff.genes[,"p.value"], " "), function(x) x[1]))
    fdr <- p.adjust(pval, "fdr")
    #gene.id <- diff.genes[which(fdr <= fdr.thresh),"ID"]
    gene.id <- diff.genes[which(pval <= pval.thresh),"ID"]
    trans.idx <- match(gene.id, eqtl.table[[1]])
    trans.idx <- trans.idx[which(!is.na(trans.idx))]
    query.table <- eqtl.table[trans.idx,]
}
```

We removed the transcripts that are located in the trait QTL.
We don't want to spike our gene list with local genes.

```{r remove_cis}
chr.locale <- which(gene.table[[4]] == chr.peak[,"chr"])
above.min <- which(gene.table[[5]]*10^6 >= qtl.start)
below.max <- which(gene.table[[6]]*10^6 <= qtl.end)
cis.idx <- Reduce("intersect", list(chr.locale, above.min, below.max))
cis.transcripts <- unique(gene.table[cis.idx,][[1]])

cis.locale <- which(query.table[[1]] %in% cis.transcripts)
trans.locale <- setdiff(1:nrow(query.table), cis.locale)
query.table <- query.table[trans.locale,]
```

There are `r nrow(query.table)` training genes.
The positions of eQTL markers for these genes is shown below.

```{r eqtl, width = 10, height = 5}

split.marker <- strsplit(query.table$marker.id, "_")
if(training.set == "Trans-band"){
    marker.pos <- as.numeric(sapply(split.marker, function(x) x[2]))
    a <- hist(marker.pos/1e6, breaks = 50, plot = FALSE)
    plot.new()
    plot.window(xlim = c(min(a$mids), max(a$mids)), ylim = c(0, max(a$counts)))
    points(a$mids, a$counts, type = "h", lwd = 20, col = "lightgray")
    text(a$mids, a$counts, labels = a$counts)
    axis(1);axis(2)
    mtext("Position (Mb)", side = 1, line = 2.5)
    mtext("Number of Trans Targets", side = 2, line = 2.5)
    abline(v = peak.table[1,"pos"])
    text(x = peak.table[1,"pos"], y = max(a$counts), labels = "Peak Trait QTL", adj = 0)
}
if(training.set == "SOM"){
    marker.chr <- sapply(split.marker, function(x) x[1])
    barplot_with_num(sort(table(marker.chr)), main = "Transcripts per Chromosome",
    xlab = "Chromosome", ylab = "Count")
}
```

## Enrichment of Selected Genes

The functional enrichment of the trans-regulated genes
is shown below.

```{r trans_enrich, fig.width = 7, fig.height = 7}
enrichment <- gost(query.table$gene.id, organism = "mmusculus")
plot.enrichment(enrichment, num.terms = 25, plot.label = 
    "Enrichment of trans band targets", max.term.size = 5000)
```

The allele effects of all transcripts in this region
are shown below. There are small groups of transcripts
that share allele effects. The F and G alleles have
the most consistent effects across transcripts.

```{r allele_effects}
allele.effects <- as.matrix(query.table[,LETTERS[1:8]])
pheatmap(allele.effects, scale = "row")
#allele.cor <- cor(t(allele.effects))
#pheatmap(allele.cor)
```

The figure below shows the first four PCs of the trans-regulated
gene expression matrix plotted against each other. Each dot represents
an individual animal. In some cases there are strange clusters of 
individuals. Further investigation has revealed that these clusters
tend to be related to pseudogenes that have expression patterns that
have very divergent from the rest of the transcripts. 

```{r trans_expr, fig.height = 8, fig.width = 8}
query.id <- unique(query.table[[1]])
query.info <- gene.table[match(query.id, gene.table[[1]]),]
query.expr <- adj.expr[,query.id] 
matched.mats <- get.xz(query.expr, adj.pheno)
trans.decomp <- plot.decomp(matched.mats$X, pc = 6)
#pheatmap(matched.mats$X)
```

In TRiAGE we only use protein coding genes with entrez IDs.
If we filter to only those genes here, most of the strainge
clustering patterns disappear. 


```{r filter_entrez, fig.height = 8, fig.width = 8}
entrezgene <- gconvert(query.id, organism = "mmusculus", target = "ENTREZGENE_ACC")
triage.id <- entrezgene[,"input"]

query.expr <- adj.expr[,triage.id] 
matched.mats <- get.xz(query.expr, adj.pheno)
query.decomp <- plot.decomp(matched.mats$X, pc = 6)
```

We asked whether the PCs of the trans-regulated 
transcripts were correlated with `r trait.name`.
This varied by tissue. Adipose expression was
the most highly correlated with `r trait.name`.

```{r pc_trait_cor, fig.width = 8, fig.height = 6}
if(is.interactive){quartz(width = 8, height = 6)}
par(mfrow = c(2,3))
for(i in 1:ncol(query.decomp$u)){
    plot.with.model(query.decomp$u[,i], matched.mats$Z[,trait.name], 
    xlab = paste("PC", i), ylab = trait.name, main = paste("PC", i))
}
```

The heatmap below shows the correlation structure of the transcripts
in our query.

```{r expr_decomp, fig.height = 7, fig.width = 7}
pheatmap(cor(matched.mats$X), show_rownames = FALSE, show_colnames = FALSE)
```

We used these transcripts to train SVMs to classify genes 
related to these transcripts. The following text explains 
this process.

```{r download_net, eval = TRUE, echo = FALSE}
#Download the network if it hasn't been previously downloaded. 
#full list available at:  http://fntm.princeton.edu
tissue.type = gsub(" ", "_", network.name)
net.file <- here("Data", paste0(tissue.type, "_top.RData"))
if(!file.exists(net.file)){
    download.tissue.net(tissue.type, organism = "mouse", top.edges.only = TRUE, 
    project.dir = here("Data"))
    }
```

```{r read_net, echo = FALSE}
#Read in the downloaded network.
all.var <- ls()
net.loaded <- as.logical(length(which(all.var == "tissue.net")))
if(!net.loaded){
    tissue.net <- readRDS(net.file)
}
```

## Setup Positive-Labeled Gene Features

Get the connection weights of the positive-labeled genes to each other
in the tissue network. Positive-labeled genes that are not connected 
to any other positive-labeled genes are removed from the analysis.


```{r pick_genes, echo = FALSE}
tissue.adj <- tissue.adj.mat(tissue.net, entrezgene[,"target"], inc.all.genes = FALSE)
```

After identifying all Entrezgene IDs and building the positive-labeled network,
we have `r nrow(tissue.adj)` positive-labeled genes for training.

## Train SVM
We used this positive-labeled network as features to train 100 SVMs to classify
TB- and not TB-related genes. In each of the 100 SVMs, we selected a balanced
set of unlabeled genes for training. We write each of the trained models to a 
file for further analysis.

```{r train_models, message = FALSE, warning = FALSE, error = FALSE, eval = TRUE}
generate.triage.models(path = results.dir, project.name = project.name, 
trait.genes = entrezgene[,"target"], fgn = tissue.net, n.trials = 100, 
cluster.modules = FALSE, cluster.threshold = 700, max.cluster.size = 600, 
min.cluster.size = 10, use.SVD = FALSE, verbose = TRUE, n.cores = n.cores)
```

## ROC Curves

We calculated ROC curves for each SVM, as well as the overal AUC for
the group. The plot below shows the ROC curves for all 100 models
along with the average AUC.

```{r roc, fig.height = 5, fig.width = 5}
all.fptp <- svm.ROC(results.dir, n.cores = n.cores, plot.results = TRUE)
all.auc <- sapply(all.fptp, function(x) smooth.auc(Reduce("rbind", x), plot = FALSE))
```

## Score all genes

We then used the trained models to classify genes that are encoded
in the QTL for `r trait.name` and expressed in `r tissue.name`.


```{r score_genes}
chr.locale <- which(gene.table$chr == qtl.chr)
mb.locale <- intersect(which(as.numeric(gene.table$start)*10^6 >= qtl.start), 
            which(as.numeric(gene.table$end)*10^6 <= qtl.end))
candidate.locale <- intersect(chr.locale, mb.locale)
candidate.id <- gene.table$gene.id[candidate.locale]
candidate.entrez <- gconvert(candidate.id, organism = "mmusculus", target = "ENTREZGENE_ACC")

score.candidates(results.dir, 
    candidate.genes = as.numeric(candidate.entrez[,"target"]),
    verbose = TRUE, plot.results = FALSE)

entrez.idx <- match(candidate.entrez[,"input"], gene.table$gene.id)
gene.info.mat <- as.matrix(gene.table[entrez.idx,])
#head(cbind(gene.info.mat[,1], candidate.entrez[c("input", "name")]))
candidate.entrez.id <- candidate.entrez[,"target",drop=FALSE]
colnames(candidate.entrez.id) <- "entrezgene_id"
gene.info.mat <- cbind(gene.info.mat, candidate.entrez.id)

merge.svm.gene.info(results.dir, gene.info.table = gene.info.mat,
    entrezgene.column = "entrezgene_id", gene.start.column = "start",
	gene.end.column = "end", gene.name.column = "symbol")
```

The following plot shows the mean SVM score for expressed genes
across the trait QTL.

```{r svm_fig, results = "asis"}
svm.image.path <- file.path(results.dir, project.name, "Module1", "Candidate.Gene.SVM.Results.jpg")
image.line <- paste0("![Mean SVM Scores](", svm.image.path, ")")
cat(image.line, "\n")
```

The following image shows the mean false positive rate across 
all expressed genes in the trait QTL. This may differ from the
mean SVM score shown above because both are means taken independently
and vary across SVM runs.

```{r get_fp, error = FALSE, warning = FALSE, message = FALSE, fig.width = 10, fig.height = 5}
fp.mat <- get.all.fp(results.dir, gene.name.column = "symbol")
cand.pos <- gene.table$start[match(rownames(fp.mat), gene.table$symbol)]

layout(matrix(c(1,2), ncol = 2), widths = c(1, 0.2))
par(mar = c(4,4,4,0))
plot.new()
plot.window(xlim = c(min(cand.pos/1e6, na.rm = TRUE), max(cand.pos/1e6, na.rm = TRUE)), 
    ylim = c(min(-log10(fp.mat), na.rm = TRUE), max(-log10(fp.mat), na.rm = TRUE)))
for(i in 1:ncol(fp.mat)){
    points(cand.pos/1e6, -log10(fp.mat[,i]), pch = 16, col = i)
    text(cand.pos/1e6, -log10(fp.mat[,i]), rownames(fp.mat), pos = 2, col = i)
}
axis(1);axis(2)
mtext(side = 1, paste("Chr", qtl.chr, "position (Mb)"), line = 2)
mtext(side = 2, "Functional Score", line = 2)
abline(h = -log10(0.05))
par(mar = c(0,0,0,0))
plot.new()
plot.window(xlim = c(0,1), ylim = c(0,1))
par(xpd = NA)
legend(x = -1, y = 1, legend = colnames(fp.mat), col = 1:ncol(fp.mat), lty = 1, lwd = 2)
par(xpd = TRUE)
```



## Mediation

The following code can be run by hand to check whether 
a candidate gene is a mediator of either the trans band
or the trait.

```{r check_gene, eval = FALSE}

gene.name = "Heatr6"
gene.name = "Ccdc92b"


gene.idx <- which(gene.table$symbol == gene.name)
gene.id <- gene.table$gene.id[gene.idx]

plot.with.model(adj.pheno[,trait.name], adj.expr[,gene.id], xlab = gene.name,
    ylab = trait.name)

nearest.marker <- gene.table$nearest.marker.id[gene.idx]
nearest.geno <- genoprobs[[11]][rownames(adj.expr),,nearest.marker]


trait_med <- bmediatR(
    y = adj.pheno[,trait.name,drop=FALSE], 
    M = adj.expr[,gene.id,drop=FALSE], 
    X = cbind(rep(1, nrow(nearest.geno)), nearest.geno),
    ln_prior_c = "reactive",
    options_X = list(sum_to_zero = TRUE, center = FALSE, scale = FALSE))

plot_posterior_bar(trait_med, mediator_id = gene.id, relabel_x = gene.name, 
    main = paste(gene.name, "mediating trait in", tissue.name), 
    med_var = "transcript.id")

trans.pc <- trans.decomp$u
rownames(trans.pc) <- rownames(matched.mats$X)
colnames(trans.pc) <- paste0("PC", 1:ncol(trans.pc))

trans_med <- bmediatR(
    y = trans.pc[,1,drop=FALSE], 
    M = adj.expr[,gene.id,drop=FALSE], 
    X = cbind(rep(1, nrow(nearest.geno)), nearest.geno),
    ln_prior_c = "reactive",
    options_X = list(sum_to_zero = TRUE, center = FALSE, scale = FALSE))

plot_posterior_bar(trans_med, mediator_id = gene.id, relabel_x = gene.name, 
    main = paste(gene.name, "Mediating trans band in", tissue.name), 
    med_var = "transcript.id")

#does the trans band mediate the effect of the peak marker on the trait?

nearest.marker <- find_marker(map, chr = peak.table[1,"chr"], pos = peak.table[1,"pos"]) 
nearest.geno <- genoprobs[[11]][rownames(adj.expr),,nearest.marker]

pc = 1
common.ind <- Reduce("intersect", list(rownames(adj.pheno), rownames(trans.pc), rownames(genoprobs[[1]])))

plot.with.model(trans.pc[common.ind,pc], adj.pheno[common.ind, trait.name])
adj.pc <- adjust(trans.pc[,1,drop=FALSE], nearest.geno)
adj.trait <- adjust(adj.pheno[,trait.name,drop=FALSE], nearest.geno)
plot.with.model(adj.pc[common.ind,], adj.trait[common.ind,])

trans_trait_med <- bmediatR(
    y = adj.pheno[common.ind,trait.name,drop=FALSE], 
    M = trans.pc[common.ind,1,drop=FALSE],
    X = nearest.geno[common.ind,],
    ln_prior_c = "reactive",
    options_X = list(sum_to_zero = TRUE, center = FALSE, scale = FALSE))

plot_posterior_bar(trans_trait_med, mediator_id = paste0("PC", pc), 
    relabel_x = paste("Trans band PC", pc), 
    main = "Trans Band Mediation of Peak Marker on Trait")

#post <- get_posterior(trans_trait_med, mediator_id = paste0("PC", pc), 
#    relabel_x = paste("Trans band PC", pc), 
#    main = "Trans Band Mediation of Peak Marker on Trait")
#par(mar = c(10, 4, 4, 4))
#barplot(post[,"post_p"], names = post[,"model"], las = 2)
```

```{r med_fun}

med <- function(y, M, X, model.type = "reactive"){

    test_med <- bmediatR(y = y, M = M, X = X, ln_prior_c = model.type,
        options_X = list(sum_to_zero = TRUE, center = FALSE, scale = FALSE))

    post <- get_posterior(test_med)
    post.prob <- post[,"post_p"]
    names(post.prob) <- post[,"model"]
    return(post.prob)
}
```

The following plot shows the posterior probabilities for each
model using expression of a locally encoded gene to mediate
the effect of the peak marker genotype on the **query expression**.
This is either the first PC of the trans band expression, or the
first PC of the expression of the differentially expressed genes.

```{r mediate_training_set, echo = FALSE, message = FALSE, results = "hide"}
query.pc <- query.decomp$u
rownames(query.pc) <- rownames(matched.mats$X)
colnames(query.pc) <- paste0("PC", 1:ncol(query.pc))

common.ind <- Reduce("intersect", list(rownames(query.pc), 
    rownames(genoprobs[[1]]), rownames(adj.expr)))

#mediate everything in terms of the peak marker for the trait.
peak.marker <- find_marker(map, chr = 11, peak.table[1,"pos"])
peak.geno <- genoprobs[[11]][,,peak.marker]

cis.names <- gene.table$symbol[match(cis.transcripts, gene.table$gene.id)]

expr_mediate_file <- file.path(results.dir, "Mediation_of_expr.RDS")
if(!file.exists(expr_mediate_file)){
    #mediate the trans band or differentially expressed genes with 
    #each of the cis transcripts using bmediatR
    all.cis.expr.med <- matrix(NA, ncol = 6, nrow = length(cis.transcripts))
    rownames(all.cis.expr.med) <- cis.names
    for(i in 1:length(cis.transcripts)){
        cis.idx <- which(gene.table$gene.id == cis.transcripts[i])
        cis.expr <- adj.expr[,cis.transcripts[i],drop=FALSE]
        med.post <- med(query.pc[common.ind,1], cis.expr[common.ind,1],
            peak.geno[common.ind,], "reactive")
        #barplot(med.post)
        all.cis.expr.med[i,] <- med.post
    }
    colnames(all.cis.expr.med) <- names(med.post)
    saveRDS(all.cis.expr.med, expr_mediate_file)
}else{
    all.cis.expr.med <- readRDS(expr_mediate_file)
}

pheatmap(all.cis.expr.med, cluster_cols = FALSE, show_rownames = FALSE)
#comp.expr.med <- which(all.cis.expr.med[,1] >= 0.8)
#cat(names(comp.expr.med), sep = "\n")
#comp.expr.react <- which(all.cis.exprtrans.med[,4] >= 0.8)
#cat(names(comp.expr.react), sep = "\n")
```

The following plot shows the posterior probabilities for each
model using expression of a locally encoded gene to mediate
the effect of the peak marker genotype on the **trait**.


```{r mediate_trait, echo = FALSE, message = FALSE, results = "hide"}
#mediate the trait with each of the local transcripts using bmediatR
#mediate everything in terms of the peak trait marker.
trait_mediate_file <- file.path(results.dir, "Mediation_of_trait.RDS")

if(!file.exists(trait_mediate_file)){
    all.cis.trait.med <- matrix(NA, ncol = 6, nrow = length(cis.transcripts))
    rownames(all.cis.trait.med) <- cis.names
    for(i in 1:length(cis.transcripts)){
        cis.idx <- which(gene.table$gene.id == cis.transcripts[i])
        cis.expr <- adj.expr[,cis.transcripts[i],drop=FALSE]
        med.post <- med(adj.pheno[common.ind,trait.name], cis.expr[common.ind,1],
            peak.geno[common.ind,], "reactive")
        all.cis.trait.med[i,] <- med.post
    }
    colnames(all.cis.trait.med) <- names(med.post)
    saveRDS(all.cis.trait.med, trait_mediate_file)
}else{
    all.cis.trait.med <- readRDS(trait_mediate_file)
}

pheatmap(all.cis.trait.med, cluster_cols = FALSE)

#comp.trait.med <- which(all.cis.trait.med[,1] >= 0.8)
#cat(names(comp.trait.med), sep = "\n")
#comp.trait.react <- which(all.cis.trait.med[,4] >= 0.8)
#cat(names(comp.trait.react), sep = "\n")
```

```{r rank_genes_for_both}
expr.med.prob <- rowSums(all.cis.expr.med[,c("complete med", "partial med")])
trait.med.prob <- rowSums(all.cis.trait.med[,c("complete med", "partial med")])
med.ranks <- rank.2D(cbind(expr.med.prob, trait.med.prob))
ordered.med <- med.ranks[order(med.ranks[,1]),1,drop=FALSE]
```

The following plot shows the posterior probabilites of each local 
gene mediating the trait plotted against the posterior probability 
of each local gene mediating the transcript. The top ranked genes
are labeled.

```{r med_against}
top.genes <- Reduce("union", list(rownames(ordered.med)[1:10], names(which(expr.med.prob > 0.8)), names(which(trait.med.prob > 0.8))))

df <- data.frame("Expression_Mediation_Prob" = expr.med.prob, "Trait_Mediation_Prob" = trait.med.prob)
pl <- plot_ly(df, x = ~Expression_Mediation_Prob, y = ~Trait_Mediation_Prob)
pl <- add_text(pl, x = ~Expression_Mediation_Prob, y = ~Trait_Mediation_Prob, text = names(expr.med.prob))
pl

#plot(expr.med.prob, trait.med.prob, xlim = c(0, 1), ylim = c(0, 1.05),
#    xlab = "Post. Prob. of Mediating Expression", ylab = "Post. Prob. of mediating Trait")
#text(expr.med.prob, trait.med.prob, labels = top.genes, pos = 3)
```

The following plots shows the posterior probabilities of mediating either
the trait or the trans bands as a function of the position in the locus. 
The peak trait locus is marked with a vertical line. A posterior probability
of 0.8 is marked with a horizontal line.

```{r trait_by_pos, fig.width = 10, fig.height = 6}
gene.pos <- gene.table$start[match(cis.names, gene.table$symbol)]

plot(gene.pos, expr.med.prob, main = "Expression mediators", 
    ylab = "Post. Prob. of Complete Mediation")
text(gene.pos, expr.med.prob, cis.names, cex = 0.5, pos = 4)
abline(v = peak.table[1,"pos"], h = 0.8)

plot(gene.pos, trait.med.prob, main = "Trait mediators",
    ylab = "Post. Prob. of Complete Mediation")
text(gene.pos, trait.med.prob, cis.names, cex = 0.5, pos = 4)
abline(v = peak.table[1,"pos"], h = 0.8)

```

## Mediation vs. TRiAGE

Finally, we compare the TRiAGE prioritization with the mediation results.

The following plot compares genes based on their functional score
and mediation of first PC of the expression of the selected genes.

```{r fp_v_expr}

common.genes <- intersect(rownames(fp.mat), rownames(all.cis.expr.med))
col <- colors.from.values(ordered.med[common.genes,1], grad.dir = "low")

df <- data.frame("Functional_Score" = -log10(fp.mat[common.genes,1]), 
    "Post_Prob" = expr.med.prob[common.genes], "name" = common.genes, "color" = col)
pl <- plot_ly(df, x = ~Functional_Score, y = ~Post_Prob, colors = col)
#pl <- add_markers(pl, text = ~name)
pl <- add_text(pl, x = ~Functional_Score, y = ~Post_Prob, text = ~name)
pl

#plot(-log10(fp.mat[common.genes,1]), expr.med.prob[common.genes], col = col,
#    xlab = "Functional Score", ylab = "Post. Prob. of Mediating Expression",
#    ylim = c(0,1), main = "Functional Score v. Expression Mediation", pch = 16)
#abline(v = 1, h = 0.8)
#top.genes <- which(expr.med.prob[common.genes] > 0.8)
#if(length(top.genes) > 0){
#    text(x = -log10(fp.mat[common.genes[top.genes],1]), expr.med.prob[common.genes[top.genes]],
#        labels = common.genes[top.genes], pos = 4)
#}
```

The following plot compares genes based on their functional score
and mediation of the trait.

```{r fp_v_trait}
df <- data.frame("Functional_Score" = -log10(fp.mat[common.genes,1]), 
    "Post_Prob" = trait.med.prob[common.genes], "name" = common.genes, "color" = col)
pl <- plot_ly(df, x = ~Functional_Score, y = ~Post_Prob, colors = col)
#pl <- add_markers(pl, text = ~name)
pl <- add_text(pl, x = ~Functional_Score, y = ~Post_Prob, text = ~name)
pl

#plot(-log10(fp.mat[common.genes,1]), trait.med.prob[common.genes],
#    xlab = "Functional Score", ylab = "Post. Prob. of Mediating Trait",
#    ylim = c(0,1), main = "Functional Score v. Trait Mediation")
#abline(v = 1, h = 0.8)
#top.genes <- which(trait.med.prob[common.genes] > 0.8)
#if(length(top.genes) > 0){
#    text(x = -log10(fp.mat[common.genes[top.genes],1]), trait.med.prob[common.genes[top.genes]],
#        labels = common.genes[top.genes], pos = 4)
#}

```