---
title: "CC-RIX validation"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction
The purpose of this workflow is to use wrangle some CC-RIX data in 
preparation for validation of the Sash3 hypothesis developed in the 
DO CCA analysis. The previous analysis found that a composite trait 
and composite transcript co-mapped to a locus on chromosome X. The 
composite transcript mediated the effect of the locus on the composite 
trait. Through multiple independent methods we prioritized Sash3 as the 
causal gene at the locus. Sash3 gene expression was positively correlated 
with the composite trait and mediated the effect of the locus on the 
composite trait. All effects through this chain were weak, but 
consistently pointed to Sash3. 

Here we use the CC-RIX data to test whether we can find independent 
evidence for an influence of Sash3 on metabolic phenotypes.

The data from the CC-RIX are unfortuantely very different from the DO
data, so we will need to think about this carefully.

The allele frequencies are a mess, so we probably won't be able to 
do any mapping. Also, the traits measured are different. Hopefully
there is enough overlap that we can still see an effect of Sash3.

We may have to shift the focus of the validation to something that
takes advantage of the data in the CC-RIX, namely different diets
and treatment with metformin.

This workflow also saves several processed files for use in other
workflows. All are saved in Data/CC-RIX. The files are as follows:
    
1. CC_RIX_Tissue_Annotations.RDS - Contains annotation data for 
    gene expression measurements in the CC-RIX. There are three
    elements of the list, one for each of the three tissues: 
    liver, adipose, and skeletal muscle. Each element contains a 
    table with sample.id, climb.id, sex, age, strain, diet, treatment,
    tissue, and RNAseq batch.

3. CC-RIX_Groups.RDS - A matrix of the unique groupings of CC-RIX mice
    based on sex, diet, and treatment.

2. CC-RIX_BW.RDS - Contains dimension-reduced body weights for 
    each mouse and is labeled with a JMUS mouse ID. These body
    weights are smoothed estimates of the body weight closest 
    to the time of sacrifice of eacn animal. The smoothing is
    to iron out measurement errors. The time point gives us a body
    weight closes to the time of gene expression measurement.

3. CC-RIX_chem.RDS - Contains a matrix of blood chemistries by individual
    mouse. The chemistry measurements are binned into an early timepoint (_1)
    and a late timepoint (_2). Each mouse had two collections of blood, and
    they were generally at an early time point and a late time point. However,
    the measurements were not at exactly the same age for each mice, so here
    we group the measurements into early (< 30 weeks) and late (> 30 weeks).


```{r load_code}
is.interactive = FALSE
#is.interactive = TRUE
library("here")

data.dir <- here("Data", "CC-RIX")
results.dir <- here("Results", "CCA_Clusters", "Validation")
if(!file.exists(results.dir)){dir.create(results.dir)}

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "gprofiler2")
load_libraries(all.packages, personal.library = TRUE)
```

The following code chunk is where the object of the analysis is set.
The analysis can either look at an individual gene, or a few genes 
set by hand, or it can read in results from the DO study, and pull
out a specified composite transcript to analyze. These parameters
are all set below. After setting the manual or CT parameter, set
the corresponding parameters in the appropriate if chunk.

```{r set_param}
manual.or.CT <- "CT" #set whether to set genes by hand or to use a CT from the DO study

if(manual.or.CT == "manual"){
    #gene.name <- "Sash3" #set gene name(s) here
    gene.name <- "Lep"
    gene.weights <- 1 #set weights here

    #convert to ensembl ID
    #gene.id.table <- gconvert(gene.name, organism = "mmusculus")
    #gene.id <- gene.id.table[,"target"]    
    gene.id <- "ENSMUSG00000031101"

    #create a weighted vector for the gene(s)    
    expr.vector  <- c(1)
    names(expr.vector) <- gene.id

    label.text <- paste(gene.name, "expression")
}

if(manual.or.CT == "CT"){
    #set tissue, CT, and region from DO study to analyze
    #doing this by hand now, but maybe can automate later.

    tissue.name <- "Adipose"; CT <- 1
    #tissue.name <- "Adipose"; CT <- 20
    #tissue.name <- "Islet"; CT <- 1
    #tissue.name <- "Liver"; CT <- 1
    #tissue.name <- "Liver"; CT <- 5 #only a few genes, but highly correlated with traits
    #tissue.name <- "SkeletalMuscle"; CT <- 1

    #read expression data to get transcript names
    do.expr <- readRDS(here("Results", "CCA_Clusters", "all_traits", "Adjusted_Expression.RDS"))

    #read CT data
    boot.CCA.results <- readRDS(here("Results", "CCA_Clusters", "all_traits", "Aggregate.Results.RDS"))
    tissue.idx <- which(names(boot.CCA.results) == tissue.name)
    expr.vector <- boot.CCA.results[[tissue.idx]]$u[,CT]
    names(expr.vector) <- colnames(do.expr[[tissue.idx]])

    label.text <- paste(tissue.name, "CT", CT)
}
```

Gene expression in this analysis was set to analyze `r label.text`.

```{r read_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "expr")))

if(!data.loaded){
    annot <- read.csv(file.path(data.dir, "final_sample_annotations.csv"))
    expr <- read.csv(file.path(data.dir, "norm_expression_matrix.csv"))
    genoprobs <- readRDS(file.path(data.dir, "Genoprobs.RDS")) #created by 3.5.CC-RIX_genotypes.Rmd
    K <- readRDS(file.path(data.dir, "Kinship.RDS"))
    map <- readRDS(file.path(data.dir, "Map.RDS"))
    manifest <- as.matrix(read.csv(file.path(data.dir, "Original_Samples_Manifest.csv")))
}
```

## Covariates

Pull out covariates we may want to use. We will at least
want to subdivide along these groups in the future.

```{r covar}
factor.covar <- annot[,c("Sex", "Age", "RNAseq_Batch", "Diet", "Treatment")]
rownames(factor.covar) <- annot[,1]

group.def <- c("Sex", "Diet", "Treatment") #use these factors to define groups
groups <- unique(factor.covar[,group.def])
rownames(groups) <- NULL
saveRDS(groups, file.path(data.dir, "CC-RIX_Groups.RDS"))
#covar.mat <- dummy_covar(factor.covar)
```

## Split Expression by Tissue

Split expression by tissue.
Create a matrix with individuals in rows and transcripts 
in columns to match other expression data sets we are 
working with.

```{r filter_expr}
assay.id <- colnames(expr)
assay.id.locale <- match(annot[,1], assay.id)
aligned.expr <- t(expr[,assay.id.locale]) #align expression data with annotation data
colnames(aligned.expr) <- expr[,1] #add transcript names

tissue.names <- unique(annot[,"Tissue"])
tissue.annot <- lapply(tissue.names, function(x) annot[which(annot[,"Tissue"] == x),])
tissue.expr <- lapply(tissue.names, function(x) aligned.expr[which(annot[,"Tissue"] == x),])
names(tissue.annot) <- names(tissue.expr) <- tissue.names
```


```{r mouse_id}
#use the manifest to label samples with a mouse ID
for(i in 1:length(tissue.expr)){
    expr.mouse.info.table <- t(sapply(rownames(tissue.expr[[i]]), function(x) get_CC_RIX_mouse_info(x, manifest, input.type = "User.Inputs")))
    rownames(tissue.expr[[i]]) <- expr.mouse.info.table[,1]
    covar.mouse.info.table <- t(sapply(tissue.annot[[i]][,1], function(x) get_CC_RIX_mouse_info(x, manifest, input.type = "User.Inputs")))
    rownames(tissue.annot[[i]]) <- covar.mouse.info.table[,1]
}
saveRDS(tissue.annot, file.path(data.dir, "CC_RIX_Tissue_Annotations.RDS"))
```

## Phenotypes

Read in the phenotypes.

```{r pheno_data}
bw <- as.matrix(read.csv(file.path(data.dir, "Cube Body Weights 8.13.21 FINAL.csv")))
bw <- bw[which(bw[,1] != ""),]
food <- as.matrix(read.csv(file.path(data.dir, "Cube Food Intake 7.08.21_FINAL.csv")))
water <- as.matrix(read.csv(file.path(data.dir, "Cube Water Intake 8.19.21 FINAL.csv")))
chem <- as.matrix(read.csv(file.path(data.dir, "Cube_RIX_clinical chemistries_FINAL_11.01.2021.csv")))
chem <- as.matrix(chem[which(chem[,1] != ""),])
```

### Body weight

Each body weight measurement is in a single row of the bw
matrix. Here we generate a list with one element per mouse.
Each element includes all the body weights for that mouse and
the week at which each body weight was measured.

The bw table does not have any JMUS IDs, and the Lot.Barcodes
in the bw table are not in the manifest. So I used CLIMB IDs
to match up mice. 

```{r weight, fig.width = 10, fig.height = 7, message = FALSE, warning = FALSE, error = FALSE}

get_bw_ind <- function(climb.ID, manifest){

    #convert mouse ID to climb ID
    id.locale <- which(manifest[,"CLIMB.ID"] == climb.ID)[1] #take the first, since each mouse will have multiple assays associated with it
    jmus.id <- manifest[id.locale,"Barcode.1"]
    climb.locale <- which(bw[,"CLIMB.ID"] == climb.ID)

    ind.meta <- c(jmus.id, climb.ID, as.matrix(bw[climb.locale[1],c("User.defined.strain", "Sex", "Diet", "Mouse.Treatment")]))
    names(ind.meta) <- c("JMUSID", "CLIMB.ID", "Strain", "Sex", "Diet", "Treatment")
    ind.bw <- bw[climb.locale,"Body.Weight..g."]
    bw.age <- bw[climb.locale,"Age..weeks."]
    age.order <- order(bw.age)
    bw.table <- cbind(bw.age[age.order], ind.bw[age.order])
    colnames(bw.table) <- c("Age_weeks", "BW")
    full.bw <- list("Individual" = ind.meta, "BW" = bw.table)
    return(full.bw)
}

bw.mouse.info <- t(sapply(as.numeric(bw[,"CLIMB.ID"]), 
    function(x) get_CC_RIX_mouse_info(x, manifest, input.type = "CLIMB.ID")))
#head(bw.mouse.info)
bw.mouse.info[which(bw.mouse.info[,"Diet"] == "44% fat + fiber"),"Diet"] <- "HFD"
bw.mouse.info[which(bw.mouse.info[,"Diet"] == "10% fat + fiber"),"Diet"] <- "LFD"

all_mice <- unlist(bw.mouse.info[,2])
u_mice <- unique(all_mice[which(!is.na(all_mice))])
bw_mouse_info <- bw.mouse.info[match(u_mice,bw.mouse.info[,"CLIMB.ID"]),]
#head(bw_mouse_info)
#length(u_mice)

#get body weight measurements for each mouse
all.bw <- lapply(u_mice, function(x) get_bw_ind(climb.ID = x, manifest = manifest))

weights <- unlist(lapply(all.bw, function(x) as.numeric(x[[2]][,2])))
weeks <- unlist(lapply(all.bw, function(x) as.numeric(x[[2]][,1])))
xlim = c(min(weeks), max(weeks))
ylim <- c(min(weights), max(weights))

layout.matrix <- matrix(c(1,8,6,4,3,7,2,5), nrow = 2, byrow = TRUE)
layout(layout.matrix)
for(g in 1:nrow(groups)){
    group.locale <- Reduce("intersect", lapply(group.def, function(x) 
        which(bw_mouse_info[,x] == groups[g,x])))    
    #bw_mouse_info[group.locale,]
    plot.new()
    plot.window(xlim = xlim, ylim = ylim)
    for(idx in group.locale){
        sex <- all.bw[[idx]][[1]]["Sex"]
        if(!is.na(sex)){
            if(sex == "Female"){col = "#7fc97f"}
            if(sex == "Male"){col = "#beaed4"}
            points(apply(all.bw[[idx]][[2]], 2, as.numeric), col = col, 
            cex = 0.5, pch = 16, type = "l", lwd = 2)
        }
    }
    axis(1);axis(2)
    mtext("Week", side = 1, line = 2.5)
    mtext("Weight", side = 2, line = 2.5)
    mtext(paste(groups[g,], collapse = " "), side = 3)
    #legend("topleft", legend = c("Male", "Female"), col = c("#beaed4", "#7fc97f"), pch = 16)
}

pdf(file.path(results.dir, "Weight.Individual.pdf"), width = 15, height = 15)
par(mfrow = c(5,5), mar = c(2,2,2,2))
for(i in 1:length(all.bw)){
    if(nrow(all.bw[[i]][[2]]) > 0){
        plot(apply(all.bw[[i]][[2]], 2, as.numeric), type = "l", 
            main = paste(i, paste(all.bw[[i]][[1]][c(1,4:6)], collapse = ", ")),
            ylim = ylim, lwd = 3)
    }
}
dev.off()

```

### Body weight dimension reduction

Get a single body weight value for each mouse. To do this fit
smooth splines and predict weekly weights for each 
mouse using the model. Then use the last value from the predicted
weight as the weight for the mouse. This is a noise-reduced
estimate of the body weight closest to the time that gene 
expression was measured.

```{r predict_weight, message = FALSE, warning = FALSE, error = FALSE}
all.weeks <- unique(unlist(lapply(all.bw, function(x) as.numeric(x[[2]][,1]))))
pred.weeks <- ceiling(min(all.weeks)):floor(max(all.weeks))
pred.weight.mat <- matrix(NA, nrow = length(all.bw), ncol = length(pred.weeks))
rownames(pred.weight.mat) <- sapply(all.bw, function(x) x[[1]][1])
colnames(pred.weight.mat) <- pred.weeks
for(i in 1:length(all.bw)){
    bw.mat <- apply(all.bw[[i]][[2]], 2, as.numeric)
    if(length(bw.mat) > 0){
        model <- smooth.spline(bw.mat[,1], bw.mat[,2], df = (nrow(bw.mat)-1), spar = 0.8)
        ind.pred.weeks <- ceiling(min(bw.mat[,1])):floor(max(bw.mat[,1]))
        pred.weight <- predict(model, x = ind.pred.weeks)
        pred.weight.mat[i,as.character(ind.pred.weeks)] <- pred.weight$y
    }
    #plot(all.bw[[i]][[2]]);points(pred.weight, type = "l")    
}

#remove mice with no weight data
has.vals <- which(apply(pred.weight.mat, 1, function(x) !all(is.na(x))))
pred.weight.mat <- pred.weight.mat[has.vals,] #dim(pred.weight.mat)
all.bw <- all.bw[has.vals] #length(all.bw)
bw_mouse_info <- bw_mouse_info[has.vals,] #dim(bw_mouse_info)

pdf(file.path(results.dir, "Smoothed.Weight.Individual.pdf"), width = 10, height = 10)
par(mfrow = c(5,5), mar = c(2,2,2,2))
for(i in 1:length(all.bw)){
    plot(all.bw[[i]][[2]]);points(pred.weight.mat[i,], type = "l")
}
dev.off()

weight.order <- order(apply(pred.weight.mat, 1, function(x) max(x, na.rm = TRUE)), decreasing = FALSE)
pheatmap(pred.weight.mat[weight.order,], cluster_rows = FALSE, cluster_cols = FALSE,
    show_rownames = FALSE, show_colnames = FALSE)

mouse.weight.val <- apply(pred.weight.mat, 1, function(x) x[max(which(!is.na(x)))])
saveRDS(mouse.weight.val, file.path(data.dir, "CC-RIX_BW.RDS"))
#hist(mouse.weight.val, breaks = 100)
```


## Blood Chemistries {.tabset .tabset-fade .tabset-pills}

```{r chem, results = "asis"}
#get blood chemistry assay data. Assays were measured at two main
#time points: one time when the mouse was between 20 and 30 weeks 
#old, and another when the mouse was greater than 50 weeks old.
#This function combines those into an early time point and a late
#time point.
get_chem_ind <- function(climb.ID, manifest, time.pt.div = 30){

    id.locale <- which(manifest[,"CLIMB.ID"] == climb.ID)[1] #take the first, since each mouse will have multiple assays associated with it
    ind.meta <- as.matrix(manifest[id.locale[1],c("Barcode.1", "Sex", "Diet", "Treatment")])
    mouse.id <- manifest[id.locale,"Barcode.1"]
    climb.locale <- which(chem[,"CLIMB.ID"] == climb.ID)

    if(length(climb.locale) == 0){
        full.chem <- c(ind.meta, rep(NA, 10))
    }else{
        ind.chem <- chem[climb.locale,c("Age..weeks.", "Assay", "Calculated.Concentration")]
        u_assay <- unique(ind.chem[,"Assay"])
        tp1.idx <- unlist(sapply(u_assay, 
            function(x) intersect(which(ind.chem[,"Assay"] == x), 
            which(ind.chem[,"Age..weeks."] < time.pt.div))))
        tp1 <- ind.chem[tp1.idx,"Calculated.Concentration"]       
        if(length(tp1) > 0){
            names(tp1) <- paste0(names(tp1.idx), "_1")
        }

        tp2.idx <- unlist(sapply(u_assay, 
            function(x) intersect(which(ind.chem[,"Assay"] == x), 
            which(ind.chem[,"Age..weeks."] > time.pt.div))))
        tp2 <- ind.chem[tp2.idx,"Calculated.Concentration"]
        if(length(tp2) > 0){       
            names(tp2) <- paste0(names(tp2.idx), "_2")
        }

        full.chem <- c(ind.meta, tp1, tp2)
        names(full.chem) <- c("Mouse.ID", "Sex", "Diet", "Treatment", names(tp1), names(tp2))
    }
    return(full.chem)
}

u_chem_climb <- unique(chem[,"CLIMB.ID"])
chem.list <- lapply(u_chem_climb, function(x) get_chem_ind(x, manifest))
all.mouse.id  <- sapply(chem.list, function(x) x[1])
chem.mouse.id <- all.mouse.id[which(!is.na(all.mouse.id))]

u_assay_names <- unique(unlist(lapply(chem.list, names)))
assay.order <- order(u_assay_names[5:length(u_assay_names)])
u_assay_names <- c(u_assay_names[1:4], u_assay_names[5:length(u_assay_names)][assay.order])

chem.mat <- matrix(NA, nrow = length(chem.mouse.id), ncol = length(u_assay_names))
rownames(chem.mat) <- chem.mouse.id
colnames(chem.mat) <- u_assay_names
for(i in 1:nrow(chem.mat)){
    id.locale <- which(all.mouse.id == chem.mouse.id[i])  
    chem.mat[i,names(chem.list[[id.locale]])] <- chem.list[[id.locale]]
}
chem.mat[which(chem.mat[,"Diet"] == "44% fat + fiber"),"Diet"] <- "HFD"
chem.mat[which(chem.mat[,"Diet"] == "10% fat + fiber"),"Diet"] <- "LFD"
saveRDS(chem.mat, file.path(data.dir, "CC-RIX_chem.RDS"))

assays <- c("Insulin", "TRIG", "GLU", "CHOL", "C-peptide")

for(g in 1:nrow(groups)){
    cat("###", paste(groups[g,], collapse = " "), "\n")
    group.locale <- Reduce("intersect", lapply(group.def, 
        function(x) which(chem.mat[,x] == groups[g,x])))
    #pairs(apply(chem.mat[group.locale,5:ncol(chem.mat)], 2, as.numeric), log = "xy",
    #    main = paste(groups[g,], collapse = " "))
    num.mat <- apply(chem.mat[group.locale,5:ncol(chem.mat)], 2, function(x) rankZ(as.numeric(x)))
    cor.mat <- cor(num.mat, use = "pairwise.complete.obs")
    pheatmap(cor.mat, main = paste(groups[g,], collapse = " "), display_numbers = TRUE)
    cat("\n\n")
}   

```

I am going to save the food and water processing for another time. 

## Gene Expression and Blood Chemistry {.tabset .tabset-fade .tabset-pills}

Align body weight and blood chemistries with gene expression based on
mouse ID.

```{r get_expr}
#This function can get gene expression values for either a single 
#specified gene or for a set of genes with specified weights, as
#we would do when analyzing a composite trait. The function rankZ
#normalizes the gene expression and then multiplies by the genes
#weights to give one gene expression value for each mouse. 

get_num_covar <- function(mouse.id){
    #get mouse covariates
    mouse.covar <- t(sapply(mouse.id, 
        function(x) get_CC_RIX_mouse_info(x, manifest, input.type = "Barcode.1")))
    num.covar <- c("Sex", "Diet", "Treatment", "Timepoint")
    factor.info <- lapply(num.covar, function(x) as.factor(mouse.covar[,x]))
    num.info <- Reduce("cbind", lapply(factor.info, function(x) as.numeric(x)-1))
    colnames(num.info) <- num.covar
    rownames(num.info) <- mouse.covar[,1]
    return(num.info)
}

get_expr <- function(tissue.expr, weighted.gene.vector, adjust.for.covariates = TRUE){
    
    all.mouse.id <- unique(unlist(lapply(tissue.expr, rownames)))
   
    gene.names <- names(weighted.gene.vector)
    common.genes <- lapply(1:length(tissue.expr), function(x) intersect(gene.names, colnames(tissue.expr[[x]])))
    gene.weights <- lapply(1:length(tissue.expr), function(x) matrix(weighted.gene.vector[match(common.genes[[x]], gene.names)], ncol = 1))
    updated.expr <- lapply(1:length(tissue.expr), function(x) tissue.expr[[x]][,match(common.genes[[x]], colnames(tissue.expr[[x]])),drop=FALSE])
    
    if(adjust.for.covariates){
        num.info <- get_num_covar(all.mouse.id)
        adj.expr <- lapply(updated.expr, function(x) adjust(x, num.info))
    }else{
        adj.expr <- updated.expr
    }
    norm.expr <- lapply(adj.expr, function(x) apply(x, 2, rankZ))
    weighted.genes <- lapply(1:length(tissue.expr), function(x) norm.expr[[x]]%*%gene.weights[[x]])
    for(i in 1:length(tissue.expr)){
        rownames(weighted.genes[[i]]) <- rownames(tissue.expr[[i]])
    }
    names(weighted.genes) <- names(tissue.expr)
    return(weighted.genes)
}

#This function matches up individuals in gene expression output 
#from get_expr with individuals in a trait matrix. The names of
#the values in the gene expression out put are individual mouse
#IDs, the rownames of the trait matrix also nees to be mouse IDs.

match_pheno_expr <- function(expr.list, pheno.mat){
    common.ind <- lapply(expr.list, function(x) intersect(rownames(x), rownames(pheno.mat)))
    common.expr.locale <- lapply(1:length(expr.list), function(x) match(common.ind[[x]], rownames(expr.list[[x]])))
    common.pheno.locale <- lapply(1:length(expr.list), function(x) match(common.ind[[x]], rownames(pheno.mat)))
    matched.expr <- lapply(1:length(expr.list), function(x) expr.list[[x]][common.expr.locale[[x]],1,drop=FALSE])
    names(matched.expr) <- names(expr.list)
    matched.pheno <- lapply(1:length(expr.list), function(x) pheno.mat[common.pheno.locale[[x]],,drop=FALSE])
    names(matched.pheno) <- names(expr.list)
    matched.lists <- list("expr" = matched.expr, "traits" = matched.pheno)
    return(matched.lists)
}

```

```{r match_expression_and_traits}
#This code uses the weighted gene vectors set at the top of the
#file and creates corresponding composite expression vectors out 
#of the CC expression data. Change the parameters at the top
#of the file to change what is analyzed here.

#don't adjust because we are looking at groups separately
composite.expr <- get_expr(tissue.expr, expr.vector, adjust.for.covariates = FALSE)
expr.and.chem <- match_pheno_expr(composite.expr, chem.mat)

chem.matched.expr <- expr.and.chem[[1]]
matched.chem <- expr.and.chem[[2]]
```


```{r genes_and_chem, results = "asis", fig.width = 10, fig.height = 5}
all.chem.expr.cor <- vector(mode = "list", length = length(tissue.expr))
names(all.chem.expr.cor) <- names(tissue.expr)

for(i in 1:length(tissue.expr)){
    cat("###", names(tissue.expr)[i], "{.tabset .tabset-fade .tabset-pills}\n")
    group.cor.mat <- matrix(NA, nrow = nrow(groups), ncol = 10)
    colnames(group.cor.mat) <- colnames(matched.chem[[i]])[5:14]
    rownames(group.cor.mat) <- apply(groups, 1, function(x) paste(x, collapse = "_"))
    for(g in 1:nrow(groups)){
        cat("####", paste(groups[g,], collapse = " "), "\n")
        #quartz(width = 10, height = 5)
        group.locale <- Reduce("intersect", lapply(group.def, 
            function(x) which(matched.chem[[i]][,x] == groups[g,x])))

        par(mfrow = c(2,5))
        all.group.cor <- sapply(5:ncol(matched.chem[[i]]), 
                function(j) plot.with.model(rankZ(chem.matched.expr[[i]][group.locale,1]),
                rankZ(as.numeric(matched.chem[[i]][group.locale,j])),
                ylab = colnames(matched.chem[[i]])[j], xlab = "Expression", 
                main = colnames(matched.chem[[i]])[j], report = "cor.test"))
        group.cor.mat[g,] <- all.group.cor[1,]
    cat("\n\n")
    }
    all.chem.expr.cor[[i]] <- group.cor.mat
    cat("\n\n")
}

```

## Gene Expression and Blood Chemistry Summary {.tabset .tabset-fade .tabset-pills}

The following plots summarize the correlations above.

```{r plot_chem_cor_summary, results = "asis", fig.height = 6, fig.width = 6}
global.max <- max(unlist(all.chem.expr.cor))
global.min <- min(unlist(all.chem.expr.cor))
for(i in 1:length(all.chem.expr.cor)){
    cat("###", names(all.chem.expr.cor)[i], "\n")
    par(mar = c(10,7,1,1))
    #sort rows and columns based on the first PC
    #plot.decomp(all.chem.expr.cor[[i]], label.points = TRUE)
    #plot.decomp(t(all.chem.expr.cor[[i]]), label.points = TRUE)
    row.order.2d <- sort.by.then.by(plot.decomp(all.chem.expr.cor[[i]], 
        plot.results = FALSE)$u, col.type = c("n", "n"), return.order = TRUE)
    row.order <- row.order.2d[,1]
    col.order.2d <- sort.by.then.by(plot.decomp(t(all.chem.expr.cor[[i]]), 
        plot.results = FALSE)$u, col.type = c("n", "n"), return.order = TRUE)
    col.order <- col.order.2d[,1]
    imageWithText(t(all.chem.expr.cor[[i]][row.order, col.order]), 
        global.color.scale = TRUE, global.min = global.min, global.max = global.max, 
        use.pheatmap.colors = TRUE, col.text.rotation = 45, cex = 0.8,
        col.text.adj = 1, col.text.shift = 0.08, row.text.shift = 0.08)
    
    #pheatmap(t(all.chem.expr.cor[[i]]), cluster_rows = FALSE, 
    #cluster_cols = FALSE, display_numbers = TRUE)
    cat("\n\n")
}
```

The following box plots further summarize the correlations across
tissues and chemistries. 

### All Tissues and Chemistries

```{r grouped_boxes, fig.width = 10, fig.height = 5}
plot.grouped.boxes(all.chem.expr.cor, type = "Matrix")
abline(h = 0)
```

### All Tissues

```{r cor_by_tissue, fig.width = 4, fig.height = 4}
boxplot(lapply(all.chem.expr.cor, abs), ylab = "correlation")
```

Insulin has the greatest correlation with gene expression

### Tissues and Chemistries Separated

```{r cor_by_chem, fig.width = 9, fig.height = 4}
par(mfrow = c(1,3))
for(i in 1:length(all.chem.expr.cor)){
    boxplot(all.chem.expr.cor[[i]], las = 2, 
    main = names(all.chem.expr.cor)[i], 
    ylim = c(global.min, global.max))
    abline(h = 0)
}
```


## Gene expression and body weight {.tabset .tabset-fade .tabset-pills}

```{r genes_and_bw, fig.width = 10, fig.height = 5, results = "asis"}

bw.mouse.id <- sapply(all.bw, function(x) x[[1]][1])
bw.info.mat <- t(sapply(bw.mouse.id, 
    function(x) get_CC_RIX_mouse_info(x, manifest, input.type = "Barcode.1")))
bw.info.mat[which(bw.info.mat[,"Diet"] == "44% fat + fiber"),"Diet"] <- "HFD"
bw.info.mat[which(bw.info.mat[,"Diet"] == "10% fat + fiber"),"Diet"] <- "LFD"


full.bw.mat <- cbind(bw.info.mat, mouse.weight.val)
rownames(full.bw.mat) <- full.bw.mat[,1]
#head(full.bw.mat)

aligned.expr.bw <- match_pheno_expr(composite.expr, full.bw.mat)

bw.matched.expr <- aligned.expr.bw[[1]]
matched.bw <- aligned.expr.bw[[2]]

all.bw.cor <- vector(mode = "list", length = length(tissue.expr))
names(all.bw.cor) <- names(tissue.expr)

for(i in 1:length(tissue.expr)){
    cat("###", names(tissue.expr)[i],"\n")
    bw.cor.mat <- rep(NA, nrow(groups))
    #quartz(width = 10, height = 5)
    layout(layout.matrix)
    for(g in 1:nrow(groups)){
        group.locale <- Reduce("intersect", lapply(group.def, 
            function(x) which(matched.chem[[i]][,x] == groups[g,x])))        
        bw.cor <- plot.with.model(rankZ(bw.matched.expr[[i]][group.locale]),
            rankZ(as.numeric(matched.bw[[i]][group.locale,"mouse.weight.val"])),
            ylab = "Final BW", xlab = "Expression", main = paste(groups[g,], collapse = " "), 
            report = "cor.test")
        bw.cor.mat[g] <- bw.cor[1]
    }
    all.bw.cor[[i]] <- bw.cor.mat
    cat("\n\n")
}

```

The following plot summarizes the correlations across tissues.

### All Tissues

```{r bw_cor_box}
boxplot(all.bw.cor)
```

## Allele Effects {.tabset .tabset-fade .tabset-pills}

We can't do any mapping in this population, but we can at 
least check to see if the alleles that are present have 
effects in a direction we are expecting. 

In the DO, CAST had a strong negative effect on the 
composite transcript and the composite trait. Here 
we collect the transcripts from the composite
transcript and look for downregulation in PWK-containing
individuals.

If the manual.or.CT parameter is set to CT, the figure 
below shows the allele effects from the DO for 
comparison with allele effects from the CC-RIX.

This piece is only for CT pairs that had overlapping 
composite transcript QTL and composite trait QTL.
Otherwise, we can't compare trait and transcript allele 
effects. So if there are no overlapping QTLs for the 
given CT pair, this part is skipped.

```{r, plot_orig, fig.width = 9, fig.height = 5, results = "asis"}
if(manual.or.CT == "CT"){

    qtl.file <- list.files(here("Results", "CCA_Clusters", "all_traits"), 
        pattern = paste0("Individual.QTL.", tissue.name, ".Composite_Trait", CT, ".Chr"), 
        full.names = TRUE)
    if(length(qtl.file) >0){
        qtl.map <- readRDS(here("Data", "map.RDS"))
        qtl.scan <- lapply(qtl.file, readRDS)
        Ctranscript.peaks <- lapply(qtl.scan, function(x) find_peaks(x$Transcript_LOD, map = qtl.map))
        #Ctrait.peaks <- lapply(qtl.scan, function(x) find_peaks(x$Trait_LOD, map = qtl.map))
        names(Ctranscript.peaks) <- sapply(Ctranscript.peaks, function(x) paste(x["chr"], x["pos"], sep = "_"))

        for(i in 1:length(qtl.scan)){
            chr.label <- strsplit(basename(qtl.file[[i]]), ".", fixed = TRUE)[[1]][6]
            cat("QTL on Chr", chr.label, "\n")
            #quartz(width = 9, height = 5)
            plot_coefCC(qtl.scan[[i]]$Transcript_Coef, map = qtl.map, 
                main = paste(tissue.name, "Composite Transcript", CT))
            plot_coefCC(qtl.scan[[i]]$Trait_Coef, map = qtl.map, 
                main = paste(tissue.name, "Composite Trait", CT))
            cat("\n\n")
        }
    }
}else{
    qtl.file <- NULL
}

```

```{r allele_effects}
match_geno_pheno <- function(genoprobs, mouse.pheno.id, manifest){
    #get mouse info from the mouse id
    all.pheno.id <- t(sapply(mouse.pheno.id, function(x) get_CC_RIX_mouse_info(x, manifest)))

    all.geno.group <- apply(all.pheno.id, 1, function(x) paste(gsub(" X ", "_", x[6]), x[3], sep = "_"))
    #match up with the genotype data
    all.pheno.idx <- match(all.geno.group, rownames(genoprobs[[1]]))
    #take out na's
    not.na.locale <- which(!is.na(all.pheno.idx))
    all.pheno.idx <- all.pheno.idx[not.na.locale]
    mouse.pheno.id <- mouse.pheno.id[not.na.locale]
    
    new.genoprobs <- genoprobs
    #expand genoprobs to incorporate all these individuals
    #assign unique IDs to all individuals
    for(i in 1:length(genoprobs)){
        new.genoprobs[[i]] <- genoprobs[[i]][all.pheno.idx,,]
        rownames(new.genoprobs[[i]]) <- mouse.pheno.id
    }

    return(new.genoprobs)
}

get_geno_region <- function(genoprobs, map, chr, start.pos, end.pos){
    chr.locale <- which(names(map) == chr)
    marker.region <- intersect(which(map[[chr.locale]] >= start.pos), which(map[[chr.locale]] <= end.pos))
    marker.pos <- map[[chr.locale]][marker.region]
    region.geno <- genoprobs[[chr.locale]][,,marker.region]
    names(dimnames(region.geno)[[3]]) <- as.numeric(marker.pos)
    return(region.geno)
}

match_geno_pheno_region <- function(geno.region, mouse.pheno.id, manifest){
    #get mouse info from the mouse id
    all.pheno.id <- t(sapply(mouse.pheno.id, function(x) get_CC_RIX_mouse_info(x, manifest)))

    all.geno.group <- apply(all.pheno.id, 1, function(x) paste(gsub(" X ", "_", x[6]), x[3], sep = "_"))
    #match up with the genotype data
    all.pheno.idx <- match(all.geno.group, rownames(geno.region))
    #take out na's
    not.na.locale <- which(!is.na(all.pheno.idx))
    all.pheno.idx <- all.pheno.idx[not.na.locale]
    mouse.pheno.id <- mouse.pheno.id[not.na.locale]
    
    #expand genoprobs to incorporate all these individuals
    #assign unique IDs to all individuals
    new.region <- geno.region[all.pheno.idx,,]
    rownames(new.region) <- mouse.pheno.id
    
    return(new.region)
}

binned_geno_allele <- function(region.genoprobs, allele){
    genotypes <- c(0, 0.5, 1)
    binned.geno <- sapply(1:dim(region.genoprobs)[3], function(x) bin.vector(region.genoprobs[,allele,x], genotypes))
    colnames(binned.geno) <- dimnames(region.genoprobs)[[3]]
    return(binned.geno)
}

get_effects <- function(geno, pheno, 
    model = c("Additive", "Dominant", "Recessive"), plot.results = FALSE,
    col = "black", ylab = "phenotype", xlab = "genotype", main = "", xlim = c(0,1),
    report = "cor.test"){
    model <- model[1]
    binned.geno <- geno

    if(model == "Dominant"){
        binned.geno <- geno
        binned.geno[which(geno > 0.3)] <- 1
    }
    if(model == "Recessive"){
        binned.geno <- geno
        binned.geno[which(geno < 0.7)] <- 0
    }
    test <- lm(pheno~binned.geno)

    if(plot.results){
        plot.with.model(binned.geno, pheno, ylab = ylab, col = col,
        main = main, xlim = xlim, xlab = xlab, report = report)
    }
    invisible(return(test))
}



test_region_pheno <- function(genoprobs, test.pheno, manifest, map, 
    region.chr, region.start, region.end, allele, return.additive = TRUE,
    plot.results = TRUE, trait.label = "Expression", col = "black"){

    all.mouse.id <- unique(unlist(lapply(test.pheno, rownames)))

    #get mouse covariates
    num.info <- get_num_covar(all.mouse.id)

    #pull out the region of the genome we will be looking at
    region.genotypes <- get_geno_region(genoprobs, map, 
        region.chr, region.start, region.end)
    marker.pos <- as.numeric(names(dimnames(region.genotypes)[[3]]))

    #expand the region to match all mice with measured phenotypes
    matched.geno <- match_geno_pheno_region(geno.region = region.genotypes, 
        mouse.pheno.id = all.mouse.id, manifest)
    

    #pull out the allele of interest and round so that we are
    #not fitting effects to error
    #allele.geno <- binned_geno_allele(matched.geno, allele)
    allele.geno <- round(matched.geno[,allele,], 2)

    marker.effects <- vector(mode = "list", length = length(test.pheno))
    names(marker.effects) <- names(test.pheno)
    
    for(i in 1:length(test.pheno)){
        common.ind <- Reduce("intersect", list(rownames(test.pheno[[i]]), rownames(allele.geno), rownames(num.info)))
        ind.pheno.locale <- match(common.ind, rownames(test.pheno[[i]]))
        ind.geno.locale <- match(common.ind, rownames(allele.geno))
        ind.covar.locale <- match(common.ind, rownames(num.info))

        #adjust for all covariates
        adj.pheno <- adjust(test.pheno[[i]][ind.pheno.locale,,drop=FALSE], num.info[ind.covar.locale,])
        #par(mfrow = c(1,2));hist(test.pheno[[i]][ind.pheno.locale,]);hist(adj.pheno)

        add.tests <- apply(allele.geno, 2, 
            function(x) get_effects(pheno = rankZ(adj.pheno), 
            geno = x[ind.geno.locale], model = "Additive"))
        add.effects <- sapply(add.tests, function(x) coef(x)[2])
        #barplot(add.effects)

        dom.tests <- apply(allele.geno, 2, 
            function(x) get_effects(pheno = rankZ(adj.pheno), 
            geno = x[ind.geno.locale], model = "Dominant"))
        dom.effects <- sapply(dom.tests, function(x) coef(x)[2])
        #barplot(dom.effects)
    
        rec.tests <- apply(allele.geno, 2, 
            function(x) get_effects(pheno = rankZ(adj.pheno), 
            geno = x[ind.geno.locale], model = "Recessive"))
        rec.effects <- sapply(rec.tests, function(x) coef(x)[2])
        #barplot(rec.effects)

        effect.table <- cbind(add.effects, dom.effects, rec.effects)
        rownames(effect.table) <- colnames(allele.geno)  

        marker.effects[[i]] <- effect.table
        #pheatmap(marker.effects[[i]], cluster_rows = FALSE, cluster_cols = FALSE)
    }
    
    models <- c("Additive", "Dominant", "Recessive")
    max.allele.effect <- vector(mode = "list", length = length(test.pheno))
    names(max.allele.effect) <- names(test.pheno)
    if(plot.results){
        for(i in 1:length(marker.effects)){
            common.ind <- intersect(rownames(test.pheno[[i]]), rownames(allele.geno))
            ind.pheno.locale <- match(common.ind, rownames(test.pheno[[i]]))
            ind.geno.locale <- match(common.ind, rownames(allele.geno))
            ind.covar.locale <- match(common.ind, rownames(num.info))

            #adjust for all covariates
            adj.pheno <- adjust(test.pheno[[i]][ind.pheno.locale,,drop=FALSE], num.info[ind.covar.locale,])
            #par(mfrow = c(1,2));hist(test.pheno[[i]][ind.pheno.locale,]);hist(adj.pheno)

            max.model <- which.max(colMeans(marker.effects[[i]]))
            max.marker <- which.max(marker.effects[[i]][,max.model])

            if(length(max.model) == 0 || return.additive){
                max.model <- max.marker <- 1 #for when there is no variation in the genotype
            }

            test.result <- get_effects(geno = allele.geno[ind.geno.locale,max.marker], 
                pheno = rankZ(adj.pheno), plot.results = FALSE, 
                model = models[max.model], xlim = c(0,1), col = col,
                xlab = names(max.marker), ylab = trait.label, report = "cor.test",
                main = paste(names(test.pheno)[i], trait.label, names(allele), models[max.model]))

            boxplot(rankZ(adj.pheno)~bin.vector(allele.geno[ind.geno.locale,max.marker], c(0, 0.5, 1)), 
                col = col, xlab = names(max.marker), ylab = trait.label, 
                main = paste(names(test.pheno)[i], trait.label, names(allele)))

            max.allele.effect[[i]] <- coef(test.result)[2]
        }
    }
    result <- list("marker.effects" = marker.effects, "max.marker.cor" = max.allele.effect)
    return(result)
}

```

### Allele effects on body weight {.tabset .tabset-fade .tabset-pills}

The following plots show allele effects in the region on 
both gene expression and body weight. The marker with the
maximum R2 is marked with a red asterisk.

In the DO, the effects on both expression and the trait were the following:

* AJ - positive effect
* B6 - no effect
* 129 - no effect
* NOD - positive effect
* NZO - no effect
* CAST - negative effect
* PWK - negative effect
* WSB - positive effect

```{r set_region_and_labels}

if(manual.or.CT == "manual"){
    expr.trait.label <- paste(gene.name, "Expression")
}else{
    expr.trait.label <- paste("Expression of\n", tissue.name, "CT", CT, "Genes")
}

test.alleles <- c("AJ" = "AA", "B6" = "BB", "129" = "CC", 
    "NOD" = "DD", "NZO" = "EE", "CAST" = "FF", "PWK" = "GG",
    "WSB" = "HH")
```

```{r bw_marker_effects, fig.width = 9, fig.height = 6, results = "asis"}
#The peak of the QTL for the adipose composite transcript was at 49 Mb in the DO
if(length(qtl.file) > 0){
    max.allele.expr.effect <- lapply(1:length(Ctranscript.peaks), function(x) matrix(NA, nrow = length(test.alleles), ncol = length(tissue.expr)))
    max.allele.bw.effect <- lapply(1:length(Ctranscript.peaks), function(x) matrix(NA, nrow = length(test.alleles), ncol = 1))
    names(max.allele.expr.effect) <- names(max.allele.bw.effect) <- names(Ctranscript.peaks)
    for(i in 1:length(max.allele.expr.effect)){
        rownames(max.allele.expr.effect[[i]]) <- rownames(max.allele.bw.effect[[i]]) <- names(test.alleles)
        colnames(max.allele.expr.effect[[i]]) <- names(tissue.expr)
    }
    for(a in 1:length(test.alleles)){
        cat("####", names(test.alleles)[a], "\n")
        #quartz(width = 9, height = 6)
        layout.mat = matrix(c(1:5,0), nrow = 2, byrow = TRUE)
        layout(layout.mat)
        for(r in 1:length(Ctranscript.peaks)){
            region.chr <- Ctranscript.peaks[[r]][1,"chr"]
            region.start <- Ctranscript.peaks[[r]][1,"pos"] - 1
            region.end <- Ctranscript.peaks[[r]][1,"pos"] + 1
            max.bw.expr.markers <- test_region_pheno(genoprobs, test.pheno = bw.matched.expr, 
                manifest, map, region.chr, region.start, allele = test.alleles[a],
                region.end, trait.label = expr.trait.label, col = CCcolors[a])
        
            max.allele.expr.effect[[r]][a,] <- sapply(max.bw.expr.markers[[2]], function(x) x[1])
        
            #The same allele has a slight negative effect on body weight too.
            #quartz(width = 4, height = 6)
            #par(mfrow = c(2,1))
            test.bw <- list(apply(matched.bw[[1]][,"mouse.weight.val",drop=FALSE], 2, as.numeric))
            rownames(test.bw[[1]]) <- rownames(matched.bw[[1]])

            max.bw.markers <- test_region_pheno(genoprobs, test.pheno = test.bw, 
                manifest, map, region.chr, region.start, allele = test.alleles[a], 
                region.end, plot.results = TRUE, trait.label = "Final Body Weight", 
                col = CCcolors[a])
            #take the additive effect, since this is what we are comparing to in 

            max.allele.bw.effect[[r]][a,] <- sapply(max.bw.markers[[2]], function(x) x[1])

            plot.text(paste("Chromosome", Ctranscript.peaks[[r]]["chr"], 
                "\n", round(as.numeric(Ctranscript.peaks[[r]]["pos"]), 2), 
                "\u00B1", "1 Mb"), cex = 2)
            cat("\n\n")
        }
    }
}

```

### Body Weight Allele Effects Summary

The following plots show the allele effects on expression
in each tissue type. This is an aggregation of the correlations
shown above.

```{r plot_allele_effects, fig.width = 9, fig.height = 6}
if(length(qtl.file) > 0){

    for(r in 1:length(Ctranscript.peaks)){
        ylim <- c(min(max.allele.expr.effect[[r]], na.rm = TRUE), 
        max(max.allele.expr.effect[[r]], na.rm = TRUE))
        par(mfrow = c(2,3))
        for(i in 1:length(tissue.expr)){
            allele.order <- order(max.allele.expr.effect[[r]][,i])
            barplot(max.allele.expr.effect[[r]][allele.order,i], col = CCcolors[allele.order], las = 2,
            ylab = "Maximum Allele Effect on Expression", main = names(tissue.expr)[i])
            abline(h = 0)
        }
        bw.effect.order <- order(max.allele.bw.effect[[r]][,1])
        barplot(max.allele.bw.effect[[r]][bw.effect.order,1], col = CCcolors[bw.effect.order],
        ylab = "Maximum Allele Effect on Weight", main = "Body Weight", las = 2)
    }
}
```


The following plot shows the maximum allele effects in the region on body weight

### Allele effects on blood chemistries {.tabset .tabset-fade .tabset-pills}

Allele effects on blood chemistries are shown below. 
Effects on expression are not recapitulated here, because they are
the same as above. Instead we focus on showing all blood chemistries.

```{r allele_effects_chem, results = "asis", fig.width = 9, fig.height = 6, message = FALSE, warning = FALSE, error = FALSE}
if(length(qtl.file) > 0){

    max.allele.chem.effect <- lapply(1:length(Ctranscript.peaks), function(x) matrix(NA, nrow = length(test.alleles), ncol = 10))
    for(i in 1:length(Ctranscript.peaks)){
        rownames(max.allele.chem.effect[[r]]) <- names(test.alleles)
    }

    for(a in 1:length(test.alleles)){
        cat("####", names(test.alleles)[a], "\n")
        #quartz(width = 9, height = 6)
        
        #quartz(width = 4, height = 6)
        #par(mfrow = c(2,1))
        test.chem <- list(apply(matched.chem[[1]][,5:14,drop=FALSE], 2, as.numeric))
        rownames(test.chem[[1]]) <- rownames(matched.chem[[1]])

        layout.mat <- matrix(c(1:6), nrow = 2, byrow = FALSE)
        layout(layout.mat)
        for(r in 1:length(Ctranscript.peaks)){
            region.chr <- Ctranscript.peaks[[r]][1,"chr"]
            region.start <- Ctranscript.peaks[[r]][1,"pos"] - 1
            region.end <- Ctranscript.peaks[[r]][1,"pos"] + 1

            for(ch in 1:ncol(test.chem[[1]])){
                one.chem <- list(test.chem[[1]][,ch,drop=FALSE])
                max.chem.markers <- test_region_pheno(genoprobs, test.pheno = one.chem, 
                    manifest, map, region.chr, region.start, allele = test.alleles[a], 
                    region.end, plot.results = TRUE, trait.label = colnames(test.chem[[1]])[ch],
                    col = CCcolors[a])
                max.allele.chem.effect[[r]][a,ch] <- sapply(max.chem.markers[[2]], function(x) x[1])
            }
        colnames(max.allele.chem.effect[[r]]) <- colnames(test.chem[[1]])
        }
        
        cat("\n\n")
    }
}
```

### Blood Chemistry Allele Effects Summary

```{r plot_chem_summary, fig.width = 9, fig.height = 9}
if(length(qtl.file) > 0){
    #quartz(width = 9, height = 9)
    par(mfrow = c(3,3))
    for(r in 1:length(max.allele.chem.effect)){
        for(i in 1:ncol(max.allele.chem.effect[[r]])){
            allele.order <- order(max.allele.chem.effect[[r]][,i])
            barplot(max.allele.chem.effect[[r]][allele.order,i], 
            col = CCcolors[allele.order], las = 2, ylab = "r", 
            ylim = c(min(max.allele.chem.effect[[r]], na.rm = TRUE), 
            max(max.allele.chem.effect[[r]], na.rm = TRUE)),
            main = colnames(max.allele.chem.effect[[r]])[i])
            abline(h = 0)
        }
    }
}

```


The following plot is a heatmap of the allele effects above to condense 
the figure even more.

```{r chem_effect_heat, fig.height = 4.5, fig.width = 5}
if(length(qtl.file) > 0){
    for(r in 1:length(max.allele.chem.effect)){
        pheatmap(max.allele.chem.effect[[r]], cluster_rows = FALSE, cluster_cols = FALSE,
        display_numbers = TRUE)
    }
}
```