---
title: "CAPE in using clustered transcripts in multiple tissues"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

```{r set_type}
#set testing to FALSE to run full gene expression networks and to TRUE
#to run a subset.
#testing = TRUE
testing = FALSE
geno.type <- "modules"
include.real.genotypes = FALSE #whether to include true genotypes or just expression
```

The purpose of this workflow is to set up data for CAPE using clustered
transcripts. We use CoExpNets to cluster the transcripts in multiple 
tissues from the same animals. We can then run cape on the multi-tissue
transcript clusters to see if we see interactions between transcripts
from different tissues. 

In this workflow we cluster transcripts separately in liver,
pancreatic islets, and adipose tissue. We run functional enrichment
on each to characterize the modules.

We use these modules in place of genotypes in a CAPE model
to explain traits through interactions bewteen multi-tissue gene expression. 
For traits, we will use the first three eigentraits of the trait 
matrix including all traits except for weight.


```{r load_code}
library("here")
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "gprofiler2", "cape", "CoExpNets", "abind")
load_libraries(all.packages)
```


```{r read_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.clinical.phenotypes")))
if(!data.loaded){
  islets <- load(here("Data", "Attie_DO378_eQTL_viewer_v6.Rdata"))
  other.tissues <- load(here("Data", "attie_all_qtl_viewer_v10_04.20.2020.RData"))
  }
```

## Filter Expression
Filter the expression matrices to include only genes that have at least
a minimum amount of expression. 

Here we use the raw expression matrix to select transcripts from the rank Z
normalized expression matrix. 

```{r adjust_covar}
min.mean = 10

islet.trans <- which(colMeans(dataset.islet.rnaseq$data$raw) > min.mean)
islet.expr <- dataset.islet.rnaseq$data$rz[,islet.trans]
islet.covar <- dataset.islet.rnaseq$covar.matrix
adj.islet <- adjust(islet.expr, islet.covar)

adipose.trans <- which(colMeans(dataset.DO.Cube.Adipose$data$raw) > min.mean)
adipose.expr <- dataset.DO.Cube.Adipose$data$rz[,adipose.trans]
adipose.covar <- dataset.DO.Cube.Adipose$covar.matrix
adj.adipose <- adjust(adipose.expr, adipose.covar)

liver.trans <- which(colMeans(dataset.DO.Cube.Liver$data$raw) > min.mean)
liver.expr <- dataset.DO.Cube.Liver$data$rz[,liver.trans]
liver.covar <- dataset.DO.Cube.Liver$covar.matrix
adj.liver <- adjust(liver.expr, liver.covar)
```


The data set called Attie\_DO378\_eQTL\_viewer\_v6.Rdata contains the following
elements:

1. ensembl.version: the version of ensembl used
2. K: a list of kinship LOCO matrices
3. markers: a table of markers in the data set with names, chromosome, and position in 
both bp and cM
4. dataset.islet.modules: results from WGCNA
5. dataset.clinical.phenotypes: clinical phenotypes including annotations and lod peaks
6. genoprobs: genotype probabilities by chromosome (~4k markers per chromosome)
7. map: marker map from R/qtl
8. dataset.islet.rnaseq: rna seq data set inclusing annotations and lod peaks
9. dataset.islet.hotspots: annotations of QTL/eQTL hotspots 

Extract clinical traits, expression traits, and covariates from the 
islet data set. We will use the first three principal components of 
this trait matrix as our traits.

```{r pheno}
pheno <- as.matrix(dataset.clinical.phenotypes$pheno[,12:(ncol(dataset.clinical.phenotypes$pheno)-1)])
num.pheno <- apply(pheno, 2, as.numeric)
dimnames(num.pheno) <- dimnames(pheno)
covar <- dataset.clinical.phenotypes$covar
gene.table <- dataset.islet.rnaseq$annots
```

## Cluster gene expression using CoExpNets

```{r clust}
#Clustering takes a long time
#set fullAnnotation to F because function defaults to human.
#we will do this ourselves by hand.
islet.net.file <- get.files(here("Results", "Expr_Clusters", "islets"), 
want = "rds", dont.want = "pdf", full.names = TRUE)
if(length(islet.net.file) == 0){
islet.net.file = CoExpNets::getDownstreamNetwork(tissue="islets",
		net.type = "signed", debug=testing, expr.data = adj.islet,
		job.path= here("Results", "Expr_Clusters", "islets"), save.plots = TRUE,
    fullAnnotation = F)
}
islet.net <- readRDS(islet.net.file)

adipose.net.file <- get.files(here("Results", "Expr_Clusters", "adipose"),
want = "rds", dont.want = "pdf", full.names = TRUE)
if(length(adipose.net.file) == 0){
adipose.net.file = CoExpNets::getDownstreamNetwork(tissue="adipose",
		net.type = "signed", debug=testing, expr.data = adj.adipose,
		job.path= here("Results", "Expr_Clusters", "adipose"), save.plots = TRUE,
    fullAnnotation = F)
}
adipose.net <- readRDS(adipose.net.file)

liver.net.file <- get.files(here("Results", "Expr_Clusters", "liver"),
want = "rds", dont.want = "pdf", full.names = TRUE)
if(length(liver.net.file) == 0){
  liver.net.file = CoExpNets::getDownstreamNetwork(tissue="liver", n.iterations=20,
		net.type = "signed", debug=testing, expr.data = adj.liver,
		job.path= here("Results", "Expr_Clusters", "liver"), save.plots = TRUE,
    fullAnnotation = F)
}
liver.net <- readRDS(liver.net.file)
```


## Characterize Modules {.tabset .tabset-fade .tabset-pills}

In each tissue we characterize the modules identified by CoExpNets using
gprofiler2. 

```{r get_module_genes}
module_genes <- function(net.obj){
  moduleV <- net.obj$moduleColors
  modules <- unique(moduleV)
  module.genes <- lapply(modules, function(x) names(moduleV)[which(moduleV == x)])
  names(module.genes) <- modules
  return(module.genes)
}
```

### Liver Module Enrichment

```{r liver_enrichment, fig.height = 8, fig.width = 7}
liver.module.genes <- module_genes(liver.net)
liver.enrich <- lapply(liver.module.genes, function(x) gost(x, organism = "mmusculus",
sources = c("GO", "KEGG", "REACTOME")))
plot.enrichment.group(liver.enrich, max.term.size = 500, plot.label = "Liver Modules")
```

### Adipose Module Enrichment

```{r adipose_enrichment, fig.height = 8, fig.width = 7}
adipose.module.genes <- module_genes(adipose.net)
adipose.enrich <- lapply(adipose.module.genes, function(x) gost(x, organism = "mmusculus",
sources = c("GO", "KEGG", "REACTOME")))
plot.enrichment.group(adipose.enrich, max.term.size = 500, plot.label = "Adipose Modules")
```

### Islet Module Enrichment

```{r islet_enrichment, fig.height = 8, fig.width = 7}
islet.module.genes <- module_genes(islet.net)
islet.enrich <- lapply(islet.module.genes, function(x) gost(x, organism = "mmusculus",
sources = c("GO", "KEGG", "REACTOME")))
plot.enrichment.group(islet.enrich, max.term.size = 500, plot.label = "Islet Modules")
```


## Adjust clinical phenotypes for covariates  {.tabset .tabset-fade .tabset-pills}

Adjust all traits for DO wave. Keep sex as an interactive covariate.

### Before Adjusting

```{r pheno_cor}
pheatmap(cor(num.pheno, use = "pairwise.complete.obs"), 
main = "Phenotype Correlations Before Adjusting for Covariates")
```

### After Adjusting

```{r adj_pheno_cor}
adj.pheno <- adjust(num.pheno, covar[,2:ncol(covar)])
pheatmap(cor(adj.pheno, use = "pairwise.complete.obs"), 
main = "Phenotype Correlations After Adjusting for Covariates")
```

## Create "Genotype" Object and Information Table
In CAPE we use a table of information about the markers to keep track
of where the markers are positioned on the genome. When we use modules,
we lose positional information. Here we create a dummy table labeling 
the modules from each tissue as being from a different chromosome.

```{r genotypes}
common.ind <- Reduce("intersect", list(rownames(liver.net$MEs), 
rownames(adipose.net$MEs), rownames(islet.net$MEs)))
liver.ind.locale <- match(common.ind, rownames(liver.net$MEs))
adipose.ind.locale <- match(common.ind, rownames(adipose.net$MEs))
islet.ind.locale <- match(common.ind, rownames(islet.net$MEs))

liver.MEs <- liver.net$MEs[liver.ind.locale,]
colnames(liver.MEs) <- gsub("ME", "liver_", colnames(liver.MEs))

adipose.MEs <- liver.net$MEs[adipose.ind.locale,]
colnames(adipose.MEs) <- gsub("ME", "adipose_", colnames(adipose.MEs))

islet.MEs <- liver.net$MEs[islet.ind.locale,]
colnames(islet.MEs) <- gsub("ME", "islet_", colnames(islet.MEs))

expr.geno <- cbind(liver.MEs, adipose.MEs, islet.MEs)
```

## Build CAPE objects

Include all traits. Subsets of traits can be specified
in the cape yml file.

```{r to_geno}
#This function takes a matrix of numeric values and converts
#it into a 3D genotype array with the values scaled between
#0 and 1.

add_to_geno <- function(geno.obj, matX){

  #scale to be between 0 and 1 for cape genotypes
  pos.mat <- apply(matX, 2, function(x) x+abs(min(x)))
  #boxplot(pos.mat)
  scaled.mat <- apply(pos.mat, 2, function(x) x/max(x))
  #boxplot(scaled.mat)

  #create a 3D array of the scaled PCs
  new.geno <- array(NA, dim = c(nrow(matX), ncol(geno.obj), ncol(matX)))
  dimnames(new.geno) <- list(rownames(matX), colnames(geno.obj), colnames(scaled.mat))
  for(i in 1:ncol(scaled.mat)){
    one.mat <- matrix(scaled.mat[,i], nrow = nrow(scaled.mat), ncol = ncol(geno.obj))
    new.geno[,,i] <- one.mat
  }

  aug.geno <- abind(geno.obj, new.geno, along = 3)
  return(aug.geno)
}

```

```{r subset_ind}
#Subset one of the objects to only the individuals
#with expression before we make the cape object.
#Then we don't need to do build large objects and 
#subset them later
common.ind <- intersect(rownames(expr.geno), rownames(adj.pheno))
ind.locale <- match(common.ind, rownames(adj.pheno))
sub.obj <- dataset.clinical.phenotypes
sub.obj$pheno <- adj.pheno[ind.locale,]
```

```{r cross_obj}
cape.obj <- qtl2_to_cape(cross = sub.obj, genoprobs, map, covar[,1,drop=FALSE])
data.obj <- cape.obj$data_obj
geno.obj <- cape.obj$geno_obj

#replace data.obj$pheno with the numeric phenotype matrix and covariates
data.obj$pheno <- cbind(adj.pheno[ind.locale,], covar[ind.locale,1,drop=FALSE])

#scale expression data and add to the genotype object
new.geno.obj <- add_to_geno(geno.obj, expr.geno)

#add the information for the new "genotypes" to the data object
data.obj$geno_names <- dimnames(new.geno.obj)
names(data.obj$geno_names) <- c("mouse", "allele", "locus")

#add pseudo-chromosome data
#cbind(liver.MEs, adipose.MEs, islet.MEs)
new.chr <- c(rep(1, ncol(liver.MEs)), rep(2, ncol(adipose.MEs)), rep(3, ncol(islet.MEs)))
new.pos <- c(1:ncol(liver.MEs), 2:ncol(adipose.MEs), 3:ncol(islet.MEs))
data.obj$chromosome <- c(data.obj$chromosome, new.chr)
data.obj$marker_location <- c(data.obj$marker_location, new.pos)
data.obj$marker_num <- 1:length(data.obj$chromosome)


#if we only want to include expression values, take out the actual genotypes
if(!include.real.genotypes){
  true.geno.locale <- 1:dim(geno.obj)[3]
  expr.geno.locale <- setdiff(1:dim(new.geno.obj)[3], true.geno.locale)
  data.obj$chromosome <- data.obj$chromosome[expr.geno.locale]
  data.obj$marker_location <- data.obj$marker_location[expr.geno.locale]
  data.obj$marker_num <- 1:length(data.obj$chromosome)
  #also reduce the "genotypes" to just two alleles.
  #we don't need all 8 if we're just looking at expression
  new.geno.obj <- new.geno.obj[,1:2,expr.geno.locale]
  data.obj$geno_names[[3]] <- data.obj$geno_names[[3]][expr.geno.locale]
  data.obj$geno_names[[2]] <- data.obj$geno_names[[2]][1:2]
}

saveRDS(data.obj, here("Data", paste0("cape_data_", geno.type, ".RDS")))
saveRDS(new.geno.obj, here("Data", paste0("cape_geno_", geno.type, ".RDS")))
```

Now run 2.cape_run_expression.Rmd.
Use the same data set name as set here.
