---
title: "CAPE in using clustered transcripts in multiple tissues"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

```{r set_type}
geno.type <- "modules"
include.real.genotypes = FALSE #whether to include true genotypes or just expression
```

The purpose of this workflow is to set up data for CAPE using clustered
transcripts. We use CoExpNets to cluster the transcripts in multiple 
tissues from the same animals. We can then run cape on the multi-tissue
transcript clusters to see if we see interactions between transcripts
from different tissues. 

We are exploring the idea of using gene expression as an addition to,
or in place of, the genotype matrix. The idea is that we may
be able to identify interactions between genotypes and gene expression,
or between transcripts, that influence clinical traits.

In this setup we will use transcript clusters from CoExpNets. 
For traits, we will use the first three eigentraits of the trait 
matrix including all traits except for weight.

We also write real genotype data to a file for genetic mapping
of transcript modules later.

```{r load_code}
library("here")
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "gProfileR", "cape", "CoExpNets")
load_libraries(all.packages)
```


```{r read_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.clinical.phenotypes")))
if(!data.loaded){
  islets <- load(here("data", "Attie_DO378_eQTL_viewer_v6.Rdata"))
  other.tissues <- load(here("data", "attie_all_qtl_viewer_v10_04.20.2020.RData"))
  }
```

## Filter Expression
Filter the expression matrices to include only genes that have at least
a minimum amount of expression. 

Here we use the raw expression matrix to select transcripts from the rank Z
normalized expression matrix. 

```{r adjust_covar}
min.mean = 10

islet.trans <- which(colMeans(dataset.islet.rnaseq$data$raw) > min.mean)
islet.expr <- dataset.islet.rnaseq$data$rz[,islet.trans]
islet.covar <- dataset.islet.rnaseq$covar.matrix
adj.islet <- adjust(islet.expr, islet.covar)

adipose.trans <- which(colMeans(dataset.DO.Cube.Adipose$data$raw) > min.mean)
adipose.expr <- dataset.DO.Cube.Adipose$data$rz[,adipose.trans]
adipose.covar <- dataset.DO.Cube.Adipose$covar.matrix
adj.adipose <- adjust(adipose.expr, adipose.covar)

liver.trans <- which(colMeans(dataset.DO.Cube.Liver$data$raw) > min.mean)
liver.expr <- dataset.DO.Cube.Liver$data$rz[,liver.trans]
liver.covar <- dataset.DO.Cube.Liver$covar.matrix
adj.liver <- adjust(liver.expr, liver.covar)
```


The data set called Attie\_DO378\_eQTL\_viewer\_v6.Rdata contains the following
elements:

1. ensembl.version: the version of ensembl used
2. K: a list of kinship LOCO matrices
3. markers: a table of markers in the data set with names, chromosome, and position in 
both bp and cM
4. dataset.islet.modules: results from WGCNA
5. dataset.clinical.phenotypes: clinical phenotypes including annotations and lod peaks
6. genoprobs: genotype probabilities by chromosome (~4k markers per chromosome)
7. map: marker map from R/qtl
8. dataset.islet.rnaseq: rna seq data set inclusing annotations and lod peaks
9. dataset.islet.hotspots: annotations of QTL/eQTL hotspots 

Extract clinical traits, expression traits, and covariates.

```{r pheno}
pheno <- as.matrix(dataset.clinical.phenotypes$pheno[,12:(ncol(dataset.clinical.phenotypes$pheno)-1)])
num.pheno <- apply(pheno, 2, as.numeric)
dimnames(num.pheno) <- dimnames(pheno)
covar <- dataset.clinical.phenotypes$covar
gene.table <- dataset.islet.rnaseq$annots
```

## Cluster gene expression using CoExpNets

```{r clust}
#Clustering takes a long time
#set fullAnnotation to F because function defaults to human.
#we will do this ourselves by hand.

islet.net.file = CoExpNets::getDownstreamNetwork(tissue="islets",
		net.type = "signed", debug=F, expr.data = adj.islet,
		job.path= here("Results", "Expr_Clusters", "islets"), save.plots = TRUE,
    fullAnnotation = F)
islet.net <- readRDS(islet.net.file)

adipose.net.file = CoExpNets::getDownstreamNetwork(tissue="adipose",
		net.type = "signed", debug=F, expr.data = adj.adipose,
		job.path= here("Results", "Expr_Clusters", "adipose"), save.plots = TRUE,
    fullAnnotation = F)
adipose.net <- readRDS(adipose.net.file)

liver.net.file = CoExpNets::getDownstreamNetwork(tissue="liver", n.iterations=20,
		net.type = "signed", debug=F, expr.data = adj.liver,
		job.path= here("Results", "Expr_Clusters", "liver"))
liver.net <- readRDS(liver.net.file)
```


## Adjust clinical phenotypes and expression for DO wave  {.tabset .tabset-fade .tabset-pills}

Adjust all traits for DO wave. Keep sex as an interactive covariate.

### Before Adjusting

```{r pheno_cor}
pheatmap(cor(num.pheno, use = "pairwise.complete.obs"), 
main = "Phenotype Correlations Before Adjusting for DO Wave")
```

### After Adjusting

```{r adj_pheno_cor}
adj.pheno <- adjust(num.pheno, covar[,2:ncol(covar)])
adj.expr <- adjust(expr, covar[,2:ncol(covar)])
pheatmap(cor(adj.pheno, use = "pairwise.complete.obs"), 
main = "Phenotype Correlations After Adjusting for DO Wave")
```

## Rename special chromosomes
CAPE automaticall deletes sex chromosomes and mitochondrial chromosomes.
But if we are using expression, we don't need to worry about number mismatches
between males and females. So we rename these chromosomes to be numeric.
We only do this if we are looking at individual transcripts. If we are
looking at modules, we will assign each module to its own pseudo-chromosome.

```{r rename_chromosomes}
if(geno.type == "transcripts"){
  common.genes <- intersect(colnames(expr), gene.table[,"symbol"])
  gene.locale <- match(include.genes, gene.table[,"symbol"])
  gene.ids <- gene.table[gene.locale,"gene_id"]
  gene.chr <- gene.table[gene.locale,"chr"]

  #replace sex and mitochondrial chromosomes with numbers, so
  #we don't delete them in cape.
  gene.chr[which(gene.chr == "X")] <- 20
  gene.chr[which(gene.chr == "Y")] <- 21
  gene.chr[which(gene.chr == "MT")] <- 22
  gene.pos <- gene.table[gene.locale,"middle"]
  gene.info <- cbind(gene.ids, include.genes, as.numeric(gene.chr), as.numeric(gene.pos))
  colnames(gene.info) <- c("gene_id", "symbol", "chr", "middle")
  ordered.info <- sort.by.then.by(gene.info, sort.cols = c(3,4), col.type = c("n", "n"))

  id.locale <- match(ordered.info[,"gene_id"], colnames(adj.expr))
  expr.geno <- adj.expr[,id.locale]
  colnames(expr.geno) <- ordered.info[,"symbol"]
}else{
  expr.geno <- expr
}
```

## Build CAPE objects

Include all traits. Subsets of traits can be specified
in the cape yml file.

```{r to_geno}
#This function takes a matrix of numeric values and converts
#it into a 3D genotype array with the values scaled between
#0 and 1.

add_to_geno <- function(geno.obj, matX){

  #scale to be between 0 and 1 for cape genotypes
  pos.mat <- apply(matX, 2, function(x) x+abs(min(x)))
  #boxplot(pos.mat)
  scaled.mat <- apply(pos.mat, 2, function(x) x/max(x))
  #boxplot(scaled.mat)

  #create a 3D array of the scaled PCs
  new.geno <- array(NA, dim = c(nrow(matX), ncol(geno.obj), ncol(matX)))
  dimnames(new.geno) <- list(rownames(matX), colnames(geno.obj), colnames(scaled.mat))
  for(i in 1:ncol(scaled.mat)){
    one.mat <- matrix(scaled.mat[,i], nrow = nrow(scaled.mat), ncol = ncol(geno.obj))
    new.geno[,,i] <- one.mat
  }

  aug.geno <- abind(geno.obj, new.geno, along = 3)
  return(aug.geno)
}

```

```{r subset_ind}
#Subset one of the objects to only the individuals
#with expression before we make the cape object.
#Then we don't need to do build large objects and 
#subset them later
common.ind <- intersect(rownames(expr.geno), rownames(adj.pheno))
ind.locale <- match(common.ind, rownames(adj.pheno))
sub.obj <- dataset.clinical.phenotypes
sub.obj$pheno <- adj.pheno[ind.locale,]
```

```{r cross_obj}
cape.obj <- qtl2_to_cape(cross = sub.obj, genoprobs, map, covar[,1,drop=FALSE])
data.obj <- cape.obj$data_obj
geno.obj <- cape.obj$geno_obj

#replace data.obj$pheno with the numeric phenotype matrix and covariates
data.obj$pheno <- cbind(adj.pheno[ind.locale,], covar[ind.locale,1,drop=FALSE])

#scale expression data and add to the genotype object
new.geno.obj <- add_to_geno(geno.obj, expr.geno)

#add the information for the new "genotypes" to the data object
data.obj$geno_names <- dimnames(new.geno.obj)
names(data.obj$geno_names) <- c("mouse", "allele", "locus")

if(geno.type == "transcripts"){
  data.obj$chromosome <- c(data.obj$chromosome, ordered.info[,"chr"])
  data.obj$marker_location <- c(data.obj$marker_location, ordered.info[,"middle"])
  data.obj$marker_num <- 1:length(data.obj$chromosome)
  #tail(data.obj$geno_names[[3]], 43)
  #tail(dimnames(new.geno.obj)[[3]], 43)
}else{
  #add pseudo-chromosome data
  data.obj$chromosome <- c(data.obj$chromosome, 1:ncol(adj.expr))
  data.obj$marker_location <- c(data.obj$marker_location, rep(1, ncol(adj.expr)))
  data.obj$marker_num <- 1:length(data.obj$chromosome)
}

#if we only want to include expression values, take out the actual genotypes
if(!include.real.genotypes){
  true.geno.locale <- 1:dim(geno.obj)[3]
  expr.geno.locale <- setdiff(1:dim(new.geno.obj)[3], true.geno.locale)
  data.obj$chromosome <- data.obj$chromosome[expr.geno.locale]
  data.obj$marker_location <- data.obj$marker_location[expr.geno.locale]
  data.obj$marker_num <- 1:length(data.obj$chromosome)
  #also reduce the "genotypes" to just two alleles.
  #we don't need all 8 if we're just looking at expression
  new.geno.obj <- new.geno.obj[,1:2,expr.geno.locale]
  data.obj$geno_names[[3]] <- data.obj$geno_names[[3]][expr.geno.locale]
  data.obj$geno_names[[2]] <- data.obj$geno_names[[2]][1:2]
}

saveRDS(data.obj, here("Data", paste0("cape_data_", geno.type, ".RDS")))
saveRDS(new.geno.obj, here("Data", paste0("cape_geno_", geno.type, ".RDS")))
```

Now run 2.cape_run_expression.Rmd.
Use the same data set name as set here.
