---
title: "Clusters to CAPE"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

```{r set_type}

args <- commandArgs(trailingOnly=T)
exp_name = args[1]
delete_previous <- as.logical(args[2])

if(is.na(delete_previous)){
    #exp.name = "weight_adjusted"
    exp.name = "all_traits"
    delete_previous <- FALSE
}

#set testing to FALSE to run full gene expression networks and to TRUE
#to run a subset.
#testing = TRUE
testing = FALSE
is.interactive = FALSE
#is.interactive = TRUE
geno.type <- "modules"
include.real.genotypes = FALSE #whether to include true genotypes or just expression
```

The purpose of this workflow is to use data derived from adjust_transcripts_traits.Rmd
and cluster_transcripts.Rmd to data objects that can be used in CAPE.

This has not yet been updated for the new data object format.


```{r load_code}
library("here")
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "gprofiler2", "cape", "CoExpNets", "abind")
load_libraries(all.packages)
```

Read in pre-filtered and pre-adjusted transcript data from cluster_transcripts_CCA.Rmd


```{r read_data}
data.dir <- here("Results", "CCA_Clusters", exp.name)
expr.data <- readRDS(file.path(data.dir, "Adjusted_Expression.RDS"))
```

## Cluster gene expression using CoExpNets

```{r clust}
#Clustering takes a long time
#set fullAnnotation to F because function defaults to human.
#we will do this ourselves by hand.
tissue.nets <- vector(mode = "list", length = length(expr.data))
names(tissue.nets) <- names(expr.data)

for(i in 1:length(expr.data)){
  clust_dir_name <- paste(names(expr.data)[i], exp.name, sep = "_")
  cluster_dir <- here("Results", "Expr_Clusters", clust_dir_name)

  if(!file.exists(cluster_dir)){dir.create(cluster_dir)}
  net.file <- get.files(cluster_dir, want = c("net", "rds"), dont.want = "pdf", full.names = TRUE)
  if(length(net.file) == 0){
  net.file = CoExpNets::getDownstreamNetwork(tissue=clust_dir_name,
		net.type = "signed", debug=testing, expr.data = expr.data[[i]],
		job.path= cluster_dir, save.plots = FALSE, fullAnnotation = F)
  }
tissue.nets[[i]] <- readRDS(net.file)
}
```


## Characterize Modules {.tabset .tabset-fade .tabset-pills}

In each tissue we characterized the modules identified by CoExpNets using
gprofiler2. We looked for enrichment in all GO domains, as well as KEGG 
and REACTOME pathways.

The following heatmaps show enriched terms for each set of modules.
Shown are the top 10 most significant terms with fewer than 500 genes.
The islet enrichment plot only shows the top 5 terms per group, since
there are so many modules in this tissue.

```{r get_module_genes}
module_genes <- function(net.obj){
  modules <- gsub("ME", "", colnames(net.obj$MEs))
  moduleV <- net.obj$moduleColors
  module.genes <- lapply(modules, function(x) names(moduleV)[which(moduleV == x)])
  names(module.genes) <- modules
  return(module.genes)
}

characterize_module <- function(enrich.table, n.terms = 10, 
pval_thresh = 0.05, max.term.size = NULL, order.by = "p_value",
decreasing = FALSE){

  low.info.words <- c("to", "of", "the", "process", "or", "by", "in", "and")

  if(length(enrich.table) == 0){return("none")}
  
  if(length(enrich.table) > 1){
    enrichments <- enrich.table[[1]]
    if(length(max.term.size) > 0){
      term.locale <- which(enrichments[,"term_size"] <= max.term.size)
    }else{
      term.locale <- 1:nrow(enrichments)
    }
    if(length(pval_thresh) > 0){
      sig.locale <- which(enrichments[,"p_value"] <= pval_thresh)
    }else{
      sig.locale <- 1:nrow(enrichments)
    }
    take.locale <- intersect(term.locale, sig.locale)
    all.terms <- enrichments[take.locale,"term_name"]
    if(!is.null(order.by)){
      ordered.terms <- enrichments[order(enrichments[,order.by], decreasing = decreasing),]
    }else{
      ordered.terms <- enrichments
    }
    all.terms <- ordered.terms[,"term_name"]
    term.words <- unique(unlist(strsplit(all.terms, " ")))
    pruned.words <- setdiff(term.words, low.info.words)
    top.words <- pruned.words[1:n.terms]
    mod.description <- paste(top.words, collapse = "-")
    return(mod.description)
  }
}
```


```{r tissue_enrichment, fig.height = 25, fig.width = 7, results = "asis"}
for(i in 1:length(tissue.nets)){
  cat("###", names(tissue.nets)[i], "\n")
  module.genes <- module_genes(tissue.nets[[i]])
  clust_dir_name <- paste(names(expr.data)[i], exp.name, sep = "_")
  cluster_dir <- here("Results", "Expr_Clusters", clust_dir_name)

  enrich.file <- file.path(cluster_dir, "group_enrichment.RDS")
  if(!file.exists(enrich.file)){
    enrich <- lapply(module.genes, function(x) gost(x, organism = "mmusculus",
    sources = c("GO", "KEGG", "REACTOME")))
    saveRDS(enrich, enrich.file)
  }else{
    enrich <- readRDS(enrich.file)
  }
  mod.names <- sapply(enrich, 
  function(x) characterize_module(x, n.terms = 5, order.by = "p_value", max.term.size = 500))
  tissue.nets[[i]]$mod.description <- mod.names
  #pdf("~/Desktop/liver.enrichment.pdf", width = 7, height = 25)
  plot.enrichment.group(enrich, max.term.size = 500, 
  plot.label = paste(names(expr.data)[i], "Modules"),
  transformation = "sqrt", sort.by = "p_value")

  cat("\n\n")
  #dev.off()
}
```

## Trait-Module Correlation {.tabset .tabset-fade .tabset-pills}

## Module overlap {.tabset .tabset-fade .tabset-pills}

We were interested in comparing gene modules from the different tissues. 
The figure below shows the correlation matrix for the module eigengenes. 
Interestingly, the module eigengenes tend to cluster by tissue, and not
as much by function. For example, modules labeled with "immune" tend not
to cluster near each other.

```{r mod_overlap}
module.descriptions <- lapply(1:length(tissue.nets), 
function(x) paste0(names(tissue.nets)[x], "_", tissue.nets[[x]]$mod.description))
#get module eigengenes for each tissue
mod.eig <- lapply(tissue.nets, function(x) x$MEs)
common.rows  <- Reduce("intersect", lapply(mod.eig, rownames))
common.idx <- lapply(mod.eig, function(x) match(common.rows, rownames(x)))
eig.mat <- Reduce("cbind", lapply(1:length(mod.eig), function(x) mod.eig[[x]][common.idx[[x]],]))
colnames(eig.mat) <- unlist(module.descriptions)
#pdf("~/Desktop/cor.eig.pdf", width = 20, height = 20)
pheatmap(cor(eig.mat))
#dev.off()
```

## Module mapping  {.tabset .tabset-fade .tabset-pills}

Here we map each module eigengene from CoExpNets. Compare these to the 
mapping of the WGCNA modules in Keller et al. 2018. They are pretty grassy,
except for islet module skyblue3, which has a whopping QTL. I think this 
module is probably the same as Plum1 in Keller 2018.

The islet modules are the only ones we can compare to the Keller paper 
directly. We see a lot of the same signals with similar LOD scores.
We have more modules than were reported previously. We see the same
chromosome 11 hotspot spanning multiple modules. We also see a similar
pleiotropic region on the X.

```{r map_liver_mods, fig.width = 12, fig.height = 5, results = "asis"}
for(i in 1:length(tissue.nets)){
  cat("###", names(tissue.nets)[i], "\n")
  scan.file <- file.path("Results", "Expr_Clusters", 
  paste(names(tissue.nets)[i], exp.name, sep = "_"), 
  paste0(names(tissue.nets)[i], "_scan.RDS"))
  if(!file.exists(scan.file)){
    scan.results <- scan1(genoprobs, tissue.nets[[i]]$MEs)
    saveRDS(scan.results, scan.file)
  }else{
    scan.results <- readRDS(scan.file)
  }

  par(xpd = TRUE)
  mod.dist <- dist(t(scan.results))
  mod.order <- hclust(mod.dist)$order
  multilod.plot(scan.results[,mod.order], map = map, lod.thresh = 5, border.lwd = 1, 
  row.names = colnames(scan.results)[mod.order], row.name.shift = -500, 
  row.text.cex = 0.7, mar = c(2,6,2,0))
  par(xpd = FALSE)
}
```

## Read in adjusted phenotypes

```{r read_pheno}
adj.pheno <- readRDS(here("Results", "CCA_Clusters", exp.name, "Adjusted_Phenotypes.RDS"))
covar <- readRDS(here("Results", "CCA_Clusters", exp.name, "Covariates.RDS"))
```

## Create "Genotype" Object and Information Table
In CAPE we use a table of information about the markers to keep track
of where the markers are positioned on the genome. When we use modules,
we lose positional information. Here we create a dummy table labeling 
the modules from each tissue as being from a different chromosome.

```{r genotypes}
expr.geno <- eig.mat
```

## Build CAPE objects

Include all traits. Subsets of traits can be specified
in the cape yml file.

```{r to_geno}
#This function takes a matrix of numeric values and converts
#it into a 3D genotype array with the values scaled between
#0 and 1.

add_to_geno <- function(geno.obj, matX){

  #scale to be between 0 and 1 for cape genotypes
  pos.mat <- apply(matX, 2, function(x) x+abs(min(x)))
  #boxplot(pos.mat)
  scaled.mat <- apply(pos.mat, 2, function(x) x/max(x))
  #boxplot(scaled.mat)

  #create a 3D array of the scaled PCs
  new.geno <- array(NA, dim = c(nrow(matX), ncol(geno.obj), ncol(matX)))
  dimnames(new.geno) <- list(rownames(matX), colnames(geno.obj), colnames(scaled.mat))
  for(i in 1:ncol(scaled.mat)){
    one.mat <- matrix(scaled.mat[,i], nrow = nrow(scaled.mat), ncol = ncol(geno.obj))
    new.geno[,,i] <- one.mat
  }

  aug.geno <- abind(geno.obj, new.geno, along = 3)
  return(aug.geno)
}

```

```{r subset_ind}
#Subset one of the objects to only the individuals
#with expression before we make the cape object.
#Then we don't need to do build large objects and 
#subset them later
common.ind <- intersect(rownames(expr.geno), rownames(adj.pheno))
ind.locale <- match(common.ind, rownames(adj.pheno))
sub.obj <- list()
sub.obj$pheno <- adj.pheno[ind.locale,]
```

```{r cross_obj}
cape.obj <- qtl2_to_cape(cross = sub.obj, genoprobs, map, covar)
data.obj <- cape.obj$data_obj
geno.obj <- cape.obj$geno_obj

#replace data.obj$pheno with the numeric phenotype matrix and covariates
data.obj$pheno <- cbind(adj.pheno[ind.locale,], covar[ind.locale,1,drop=FALSE])

#scale expression data and add to the genotype object
new.geno.obj <- add_to_geno(geno.obj, expr.geno)

#add the information for the new "genotypes" to the data object
data.obj$geno_names <- dimnames(new.geno.obj)
names(data.obj$geno_names) <- c("mouse", "allele", "locus")

#add pseudo-chromosome data to data.obj
#cbind(liver.MEs, adipose.MEs, islet.MEs)
new.chr <- c(rep(1, ncol(liver.MEs)), rep(2, ncol(adipose.MEs)), rep(3, ncol(islet.MEs)))
new.pos <- c(1:ncol(liver.MEs), 1:ncol(adipose.MEs), 1:ncol(islet.MEs))
data.obj$chromosome <- c(data.obj$chromosome, new.chr)
data.obj$marker_location <- c(data.obj$marker_location, new.pos)
data.obj$marker_num <- 1:length(data.obj$chromosome)


#if we only want to include expression values, take out the actual genotypes
if(!include.real.genotypes){
  true.geno.locale <- 1:dim(geno.obj)[3]
  expr.geno.locale <- setdiff(1:dim(new.geno.obj)[3], true.geno.locale)
  data.obj$chromosome <- data.obj$chromosome[expr.geno.locale]
  data.obj$marker_location <- data.obj$marker_location[expr.geno.locale]
  data.obj$marker_num <- 1:length(data.obj$chromosome)
  #also reduce the "genotypes" to just two alleles.
  #we don't need all 8 if we're just looking at expression
  new.geno.obj <- new.geno.obj[,1:2,expr.geno.locale]
  data.obj$geno_names[[3]] <- data.obj$geno_names[[3]][expr.geno.locale]
  data.obj$geno_names[[2]] <- data.obj$geno_names[[2]][1:2]
}

saveRDS(data.obj, here("Data", paste0("cape_data_", geno.type, ".RDS")))
saveRDS(new.geno.obj, here("Data", paste0("cape_geno_", geno.type, ".RDS")))
```

Now run 2.cape_run_expression.Rmd.
Use the same data set name as set here.
