---
title: "SOM CC-RIX"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to explore the SOM results
to see if we can get a weighted gene feature that is LFD-like
or HFD-like.


```{r get_args}
tissue.name = "adipose"
exp.type = "Groups"
#exp.type = "Differences"
```

## SOM Results

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
needed.packages <- c("oposSOM", "gProfileR", "pheatmap", "qtl2", "grid", "stringr",
  "RColorBrewer")
load_libraries(needed.packages, personal.library = TRUE)
```


```{r read_data}
manifest <- as.matrix(read.csv(here("Data", "CC-RIX", "Original_Samples_Manifest.csv")))

#generated by SOM_CC-RIX.Rmd
adj.expr <- readRDS(here("Results", "SOM", paste0("Adjusted.Expression.", tissue.name, ".RDS")))
tissue.info <- unique(t(sapply(rownames(adj.expr), 
  function(x) get_mouse_info(x, manifest, "Barcode.1"))))
groups <- read.csv(here("Data", "CC-RIX", "Groups.csv"), row.names = 1)
group.def <- colnames(groups)
```

```{r load_som_results}
load(here("Results", "SOM", paste0("CC-RIX_", tissue.name, ".RData")))

group.files <- list.files(here("Results", "SOM", paste0("CC-RIX_", tissue.name, " - Results"), 
  paste0("Summary Sheets - ", exp.type), "CSV Sheets", "Gene Lists - Global"), full.names = TRUE)

#only compare meaningful differences. Those with one change between conditions.
if(exp.type == "Differences"){
  to.keep <- NULL
  file.grid <- t(sapply(strsplit(basename(group.files), ".", fixed = TRUE), function(x) c(x[1], x[3])))
  for(i in 1:nrow(file.grid)){
    if(is.na(file.grid[i,2])){
      to.keep <- c(to.keep, i)
    }else{
      split.group1 <- strsplit(file.grid[i,2], "_")[[1]]
      split.group2 <- strsplit(file.grid[i,1], "_")[[1]]
      n.diffs <- length(which(sapply(1:length(split.group1), function(x) split.group1[x] != split.group2[x])))
      if(n.diffs == 1){
        to.keep <- c(to.keep, i)
      }
    }
  }
group.files <- group.files[to.keep]
}

group.gene.lists <- lapply(group.files, function(x) read.csv(x, skip = 17))
names(group.gene.lists) <- gsub(".csv", "", basename(group.files))
```

For each group profile, we generate an average metagene profile.

There are several ways we can make the profiles. According to 
Wirth et al. 2011, the values listed represent the following:

log FC scale: strong to moderate differential expression
WAD scale: very strong overexpression
log log FC scale: weak to moderate differential expression

I can't find anywhere where these things are thorougly explained, or
even what WAD stands for. I will keep looking, and in the meantime,
I will use the log FC scale for my profiles.

```{r gen_profile}

get_metagene_vals <- function(gene.list, profile.dim = 50, value.name = "logFC"){
  
  metagene.vals <- vector(mode = "list", length = profile.dim^2)
  metagene.names <- rep(NA, profile.dim^2)
  idx <- 1
  for(x in profile.dim:1){
      for(y in profile.dim:1){
          metagene.idx <- row_col_to_idx(x,y, n.row = profile.dim)
          metagene <- paste(x, "x", y)
          metagene.locale <- which(gene.list[,"Metagene"] == metagene)
          metagene.vals[[idx]] <- gene.list[metagene.locale, value.name]
          spot <- spot.assig.mat[which(spot.assig.mat[,2] == metagene.idx),1]
          #spot.locale <- which(names(env$spot.list.kmeans$spots) == spot)
          #plot_portrait(env$spot.list.kmeans$spots[[spot.locale]]$mask)
          names(metagene.vals[[idx]]) <- gene.list[metagene.locale,"ID"]
          metagene.names[idx] <- paste(metagene, spot)
          idx <- idx + 1
      }
  }
  names(metagene.vals) <- metagene.names
  return(metagene.vals)
}

spot.assignments <- lapply(1:length(env$spot.list.kmeans$spots), 
  function(x) cbind(names(env$spot.list.kmeans$spots)[x], env$spot.list.kmeans$spots[[x]]$metagenes))
spot.assig.mat <- Reduce("rbind", spot.assignments)

group.metagenes <- lapply(group.gene.lists, 
  function(x) get_metagene_vals(x, value.name = "logFC"))
saveRDS(group.metagenes, here("Results", "SOM", paste0(exp.type, ".Metagenes.", tissue.name, ".RDS")))

#boxplot(group.metagenes[[1]][1:15])
group.profiles <- lapply(group.metagenes, function(x) 
  sapply(x, function(y) if(length(y) > 0){mean(y, na.rm = TRUE)}else{0}))
```

## Group Profiles

The following plots show the profiles for each group.

```{r profile_plots, fig.width = 10, fig.height = 6}
par(mfrow = c(2,4))
for(i in 1:length(group.profiles)){
  #cat("###", names(group.profiles)[i], "\n")
  plot_portrait(group.profiles[[i]], main = names(group.profiles[i]), 
    byrow = TRUE, rotate = 1)
  #cat("\n\n")
}
```

## Metagene PC

The following plot shows that in PC space, the HFD has
the following strange effects:

* HFD makes females more like males in the first PC.
* HFD makes males more like females in the second PC.
* Metformin pushes all groups a little bit higher along PC2. 

```{r metagene_pc}
metagene.mat <- Reduce("cbind", group.profiles)
colnames(metagene.mat) <- names(group.profiles)
meta.decomp <- plot.decomp(t(metagene.mat), label.points = TRUE, xlim = c(-0.5, 1))
```

## Metagene Correlation

For each individual mouse, use the group metagene
profiles to calculate the distance of the individual
mouse in transcription space to the group profile.

First calculate metagene lists for each individual mouse.
All groups were run together, so they all have the same
metagene list. 

```{r ind_metagenes}
metagene.list <- lapply(group.metagenes[[1]], names)

mouse.metagenes <- lapply(metagene.list, function(x) adj.expr[,x,drop=FALSE])
mean.mouse.metagenes <- sapply(mouse.metagenes, rowMeans)
```

The following plot shows the first four PCs of the mice
in metagene space.

```{r all_pc}
group.cols <- brewer.pal(8, "Paired")
names(group.cols)  <- names(group.metagenes)
mouse.col <- rep(NA, length = nrow(mean.mouse.metagenes))
mouse.class <- apply(tissue.info, 1, function(x) paste(x[3], x[4], x[5], sep = "_"))
for(i in 1:length(names(group.metagenes))){
  mouse.col[which(mouse.class == names(group.metagenes)[i])] <- group.cols[i]
}

not.na <- which(apply(mean.mouse.metagenes, 2, function(x) !all(is.na(x))))

mouse.decomp <- plot.decomp(mean.mouse.metagenes[,not.na], pc = 4, cols = mouse.col)
plot.new()
plot.window(xlim = c(0,1), ylim = c(0,1))
legend(0, 1, col = group.cols, pch = 16, legend = names(group.metagenes))
```

## Correlations to metagenes {.tabset .tabset-fade .tabset-pills}

Get correlations of each mouse's average metagene to the 
profile metagenes.

The following plots show the correlations between
individual mouse expression profiles and the named
metagene profile. Correlations for the mice that are 
in the named group are colored red. The correlations
are low, but the plots clearly show that the mice
in the named group are more highly correlated to the
metagene profile that represents their group than to 
other group profiles.

If we are looking at differences, rather than groups,
there won't be any red bars.

```{r group-like, results = "asis", fig.width = 10, fig.height = 5}
for(g in 1:length(group.profiles)){
  #quartz(width = 10, height = 5)
  group.name <- names(group.profiles)[g]
  cat("###", group.name, "\n")
  group.factors <- strsplit(group.name, "_")[[1]]
  group.idx <- Reduce("intersect", lapply(1:length(group.def), function(x) which(tissue.info[,group.def[x]] == group.factors[x])))
  group.col <- rep("black", nrow(mean.mouse.metagenes))
  group.col[group.idx] <- "red"
  mouse.cor <- apply(mean.mouse.metagenes, 1, function(x) cor(x, group.profiles[[g]], use = "pairwise.complete.obs"))
  cor.order <- order(mouse.cor)
  plot(mouse.cor[cor.order], col = group.col[cor.order], type = "h", 
    ylab = "Correlation", main = group.name)
  cat("\n\n")
}
```

## What next

I have saved the metagene list for each group. I want to use these
in the DO. I want to correlate the DO metagenes to the CC-RIX metagenes
to ask, for example, if the DO mice whose expression profiles are most
highly correlated to the LFD metagenes have the lowest body weight.

In other words, can we find DO HFD mice that have expression profiles
that more closely resemble LFD gene expression profiles?

