---
title: "SOM DO mice"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is examine the use of self-organizing maps 
to analyze transcriptome data for DO mice.

```{r get_args}
args <- commandArgs(trailingOnly = T)
tissue.name <- args[1]
delete_previous <- as.logical(args[2])

if(is.na(tissue.name)){
  tissue.name <- "Adipose"
  delete_previous <- FALSE
}
  is.interactive <- FALSE
  #is.interactive <- TRUE
  qtl.chr <- 11
```

## SOM Results for `r tissue.name`

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
needed.packages <- c("oposSOM", "pheatmap", "qtl2", "grid", "stringr", 
  "VennDiagram")
load_libraries(needed.packages, personal.library = TRUE)
```


Get the expression for `r tissue.name`.

```{r read_data}
vars <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
expr.name <- paste0("dataset.DO.Cube.", tissue.name)
expr.data <- get(expr.name)

#filter expression to transcripts that have at least a mean
#transcription level of 10
mean.lim <- 10
expr.count <- expr.data$data$raw
mean.expr <- colMeans(expr.count)
#hist(log10(mean.expr), breaks = 100)
high.trans <- which(mean.expr > 10)

expr <- expr.data$data$rz[,high.trans]
covar <- expr.data$covar.matrix

pheno <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"),
  stringsAsFactors = FALSE))
num.pheno <- apply(pheno[,11:(ncol(pheno)-1)], 2, as.numeric)
rownames(num.pheno) <- pheno[,1]
```

Adjust phenotypes and expression for covariates.

```{r adjust}
#adjust for diet days, as well as the other covariates
matched.mats <- get.xz(num.pheno, covar)
full.covar <- cbind(matched.mats[[1]][,1], matched.mats[[2]])
adj.pheno <- adjust(matched.mats[[1]][,-1], full.covar)
adj.expr <- adjust(expr, full.covar)
```

I'd like to try using the SOM to classify the expression differences
between animals separated by QTLs. I will start with the Ins_tAUC QTL
on chromosome 11, since this is the largest and narrower than the 
overlapping QTL for HOMA_B.

This QTL has a 4-4 split in the haplotype effects. We will try to divide
the animals along those effects, and classify the differences.

```{r map_traits}
#data(CCcolors)
#barplot(rep(1, length(CCcolors)), col = CCcolors, names = names(CCcolors))

trait.name <- "Ins_tAUC"; chr = 11
trait_map <- scan1(genoprobs, adj.pheno[,trait.name])
qtl_map <- scan1(genoprobs[,qtl.chr], adj.pheno[,trait.name, drop=FALSE])

#pheno.decomp <- plot.decomp(adj.pheno, pc = 20)
#pheno.pc <- pheno.decomp$u
#rownames(pheno.pc) <- rownames(adj.pheno)[pheno.decomp$rows.used]
#trait_map <- scan1(genoprobs, pheno.pc)
#multilod.plot(trait_map, map = map, lod.thresh = 6)
#plot(trait_map, map = map, lodcol = 6, chr = 5)
#chr_map <- scan1coef(genoprobs[,5], pheno.pc[,6])
#plot_coefCC(chr_map, map = map)
#trait_map  <- scan1(genoprobs, pheno.pc[,6])
plot(trait_map, map = map, main = trait.name)
chr_map <- scan1coef(genoprobs[,chr], adj.pheno[,trait.name])
#quartz()
plot_coefCC(chr_map, map = map)
```

NOD (D), 129 (C), B6 (B), A/J (A) are all up.
PWK (G), CAST (F), WSB (H), NZO (E) are all down.
I want to divide the DO individuals into two groups, those with the down alleles,
and those with the up alleles. I'm not sure how possible this is, so we need to 
do some exploring.

Find the peak marker for the QTL.

```{r peak_qtl}
peak.table <- find_peaks(trait_map, map = map, threshold = 10, prob = 0.95)
qtl.markers <- map[[chr]][get.nearest.pt(map[[chr]], peak.table[,"pos"])]
#qtl.markers <- intersect(which(map[[chr]] >= peak.table[,"ci_lo"]), which(map[[chr]] <= peak.table[,"ci_hi"]))
ind.locale <- match(rownames(adj.pheno), rownames(genoprobs[[1]]))
qtl.geno <- genoprobs[[chr]][ind.locale,,qtl.markers,drop=FALSE]
allele.effects <- chr_map[qtl.markers,,drop=FALSE]
```

For each marker in the QTL, check to see how well the up and down
alleles separate animals. We will weight the alleles at each marker
based on their allele effects.

1. We weight the alleles in each marker based on its effect on the trait
2. We sum across rows to give each individual a total genotype 
  score for the allele effects.
3. We bin each individual into one of two groups based on whether 
  is has predominantly negative effect alleles or positive effect alleles
4. We collect the bin assignment for each animal across all markers in the qtl
  and use the votes across all markers to determine the bin for the animal.

The following boxplot shows the bins assigned to each animal across all markers.
There were some animals consistently binned into one group or another across 
all markers, while others were assigned to both groups in varying proportions.

Rather than defining three groups, we will keep the two, and use voting
across markers to decide which animal goes into which bin. Those put into
group 1 at least 50% of the time are assigned to group 1. The horiaontal
line marks the 50% voting threshold.

```{r sep_geno, fig.width = 11, fig.height = 5}

#for each marker find the groups of animals that best split the
#up alleles from the down alleles. Minimize the number of animals
#in the middle group.

all.grouped.markers <- vector(mode = "list", length = length(qtl.markers))
for(i in 1:length(qtl.markers)){
  geno.groups <- vector(mode = "list", length = 2)
  names(geno.groups) <- c("up", "down")
  #weight the genotypes based on allele effects
  weighted.alleles <- sapply(1:ncol(qtl.geno), function(x) qtl.geno[,x,i]*allele.effects[i,x])
  weighted.effect <- rowSums(weighted.alleles)
  #barplot(sort(weighted.effect))  
  geno.groups[[1]] <- which(weighted.effect > 0)
  geno.groups[[2]] <- which(weighted.effect < 0)
  all.grouped.markers[[i]] <- geno.groups
}

#track the position of each animal over all markers
#use the rownames from the expression matrix now, 
#since we want to index the expression for the SOM
ind.groups <- lapply(rownames(adj.expr), 
  function(y) sapply(all.grouped.markers, function(x) grep(y, x)))
group.ind <- sapply(1:length(geno.groups), 
  function(y) sapply(ind.groups, function(x) length(which(x == y))))
colnames(group.ind) <- names(geno.groups)
group.order <- order(group.ind[,1])
if(is.interactive){quartz(width = 11, height = 5)}
barplot(t(group.ind[group.order,]), col = c("red", "blue"), 
  border = NA, main = "bin assignments across markers", xlab = "Individual",
  ylab = "Marker")
abline(h = length(qtl.markers)/2)
#pheatmap(group.ind)
```

We then use voting to assign each animal to a bin based on
its genotypes across all markers. 

This results in two distinct bins for the animals with no overlaps,
as shown below.

```{r bin_vote}
#put animals in groups by voting
assigned.groups <- apply(group.ind, 2, function(x) which(x > length(qtl.markers)/2))
plotVenn(assigned.groups)

#index groups based on expression matrix
group.labels <- rep(NA, nrow(adj.expr))
for(i in 1:length(assigned.groups)){
  group.labels[assigned.groups[[i]]] <- names(geno.groups)[i]
}
```

The following boxplots show trait distributions by our allele grouping.
The trait we selected, `r trait.name`, should have at least a slight
difference based on these groups.

```{r pheno_diff, fig.width = 12, fig.height = 12, eval = FALSE}
#make sure everyone is assigned to a group
#which(is.na(group.labels))

par(mfrow = c(4,5), mar = c(2,2,2,2))
for(i in 2:ncol(adj.pheno)){
  plot(rankZ(adj.pheno[match(rownames(adj.expr), rownames(adj.pheno)),i])~as.factor(group.labels),
  main = colnames(adj.pheno)[i], xlab = "", ylab = colnames(adj.pheno)[i])
}
```


```{r quick_test, eval = FALSE}
#quick test of expression differences by group using a t test.
#we expect most of these to be in the QTL that we picked, but
#hopefully there are some others too that are interesting.

test <- apply(adj.expr, 2, function(x) t.test(x~as.factor(group.labels)))
all.p <- sapply(test, function(x) x$p.value)
qqunif.plot(all.p)

fdr <- p.adjust(all.p, "fdr")
sig.idx <- which(fdr <= 0.1)
#sig.idx <- which(all.p <= 0.005)
sig.genes <- names(fdr)[sig.idx]
length(sig.genes)

pheatmap(adj.expr[,sig.idx], show_rownames = FALSE, show_colnames = FALSE)
#plot.decomp(adj.expr[,sig.idx], main = "individuals by PC")
#plot.decomp(t(adj.expr[,sig.idx]), main = "genes by PC")

library(biomaRt)
mus <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl") 

test.info <- getBM(filters = "ensembl_gene_id", attributes = c("external_gene_name",
  "chromosome_name", "start_position", "end_position"), 
  values = sig.genes, mart = mus)
test.info

enrichment <- gost(sig.genes, organism = "mmusculus")
plot.enrichment(enrichment, num.terms = 30)
```

```{r mediate, eval = FALSE}
chr.scan <- scan1(genoprobs[,11], adj.pheno[,trait.name])
all.drop <- vector(mode = "list", length = length(sig.idx))
names(all.drop) <- sig.genes
for(i in 1:length(sig.idx)){
  med.pheno <- adjust(adj.pheno, adj.expr[,sig.idx[i],drop=FALSE])
  med.scan <- scan1(genoprobs[,11], med.pheno[,trait.name])
  all.drop[[i]] <- chr.scan - med.scan
  #quartz()
  #plot(trait_map, chr = 11, map = map)
  #plot(med.scan, map = map, chr = 11, add = TRUE, col = "red")
}
max.drops <- sapply(all.drop, max)
barplot(max.drops)
max.drops[which.max(max.drops)]
```

Run the SOM using these groups as the classes to distinguish.

```{r som}
som.result.file1 <- here("Documents", "1.Setup_Data", paste0("DO_", tissue.name, ".RData"))
som.result.file2 <- here("Results", "SOM", paste0("DO_", tissue.name, ".RData"))

if(delete_previous){
  unlink(som.result.file1)
  unlink(som.result.file2)
}

if(!file.exists(som.result.file1) && !file.exists(som.result.file2)){
  exp.name <- gsub(".RData", "", basename(som.result.file1))
  env <- opossom.new(list(dataset.name = exp.name))
  env$indata <- t(adj.expr)
  env$group.labels <- group.labels
  opossom.run(env)
}

if(file.exists(som.result.file1)){
  load(som.result.file1)
}
if(file.exists(som.result.file2)){
  load(som.result.file2)
}

som.result <- env$som.result
som.coord <- som.result[[1]]
som.bmu <- som.result[[2]]
som.dim <- sqrt(nrow(som.coord))
```

## Look for phenotype variation among clusters of individuals

Look for trait differences based on grouping. This will be more 
interesting if unsupervised clustering is used. It will show 
more dramatic differences if supervised clustering is used.

```{r pheno_clusters}
groups <- env$group.labels
group.colors <- env$group.colors
u_groups <- unique(groups)
u_colors <- unique(group.colors)
ind.groups <- lapply(u_groups, function(x) names(groups)[which(groups == x)])
names(ind.groups) <- u_groups
```

The plots below show the distribution of phenotypes over each
group identified by the SOM based on transcriptional clustering.

```{r pheno_by_group, width = 12, height = 12}
pheno.ind <- lapply(ind.groups, function(x) which(rownames(adj.pheno) %in% x))
mean.pheno <- matrix(NA, nrow = length(pheno.ind), ncol = ncol(adj.pheno))
rownames(mean.pheno) <- names(pheno.ind)
colnames(mean.pheno) <- colnames(adj.pheno)

if(is.interactive){quartz(width = 8, height = 8)}
layout.mat <- get.layout.mat(9)
layout(layout.mat)
par(mar = c(4,2,2,2))
for(i in 1:ncol(adj.pheno)){  
  group.pheno <- lapply(pheno.ind, function(x) adj.pheno[x,i])
  #med.order <- order(sapply(group.pheno, function(x) median(x, na.rm = TRUE)), decreasing = FALSE)
  boxplot(group.pheno, main = colnames(adj.pheno)[i],
  col = u_colors, las = 2)
  mean.pheno[,i] <- sapply(group.pheno, function(x) median(x, na.rm = TRUE))
}
```

The following heatmap shows the mean phenotype across groups and 
phenotypes. These are scaled by row to accentuate the trait 
differences across groups, while minimizing the trait differences
across traits.

```{r mean_pheno}
if(is.interactive){quartz()}
pheatmap(t(mean.pheno), scale = "row")
```

The following plot is a PC plot of the mean trait values
across the groups. The QTL groups Glu_tAUC and WPIC at 
one end, the bulk of traits in the middle, and the HOMAs
and insulin values at the other end. Ins_tAUC, which 
is the trait for which the QTL was identified, is grouped
with the bulk of trait along the first PC, but is distant
along the second PC, which is explains a tiny amount of 
variance. I'm not sure if any of this is interesting or 
not, but like to make PC plots any time I generate a new
matrix.

## Metagenes

The transcriptome is also clustered. The genes are clustered into a 
grid of metagenes. The distribution of genes per metagene is shown 
below. 

```{r genes_per_metagene}
n.meta <- max(som.bmu)
meta.genes <- lapply(1:n.meta, function(x) names(som.bmu)[which(som.bmu == x)])
n.genes <- sapply(meta.genes, length)

if(is.interactive){quartz()}
hist(n.genes, breaks = 100, main = "Number of Genes per Metagene",
xlab = "Number of Genes")
```

The following boxplot show the absolute value of the correlations
between phenotypes and metagenes.

```{r spot_cor}
#This function gets the average expression for 
#all metagenes given the metagene assignements
#and an expression matrix.
#metagene.assig = som.bmu
metagene_expression <- function(adj.expr, metagene.assig){
  metagene.expr <- matrix(NA, nrow = nrow(adj.expr), ncol = som.dim^2)
  rownames(metagene.expr) <- rownames(adj.expr)
  colnames(metagene.expr) <- 1:ncol(metagene.expr)
  for(i in 1:ncol(metagene.expr)){
    metagene.idx <- which(metagene.assig == i)
    if(length(metagene.idx) > 0){
      scaled.expr <- apply(adj.expr[, metagene.idx,drop=FALSE], 2, rankZ)
      metagene.expr[,i] <- rowMeans(scaled.expr)
    }
  }
  return(metagene.expr)
}

metagene.expr <- metagene_expression(adj.expr, som.bmu)
pheno.metagene.cor <- vector(mode = "list", length = ncol(adj.pheno))
names(pheno.metagene.cor) <- colnames(adj.pheno)

for(m in 1:ncol(adj.pheno)){
  metagene.cor <- apply(metagene.expr, 2, function(x) cor(x, adj.pheno[,m], 
    use = "pairwise.complete.obs"))
  pheno.metagene.cor[[m]] <- matrix(metagene.cor, nrow = som.dim, ncol = som.dim, byrow = FALSE)
  #pheatmap(pheno.metagene.cor[[m]], cluster_rows = FALSE, cluster_cols = FALSE,main = colnames(adj.pheno)[m])
}

if(is.interactive){quartz()}
par(mar = c(8, 4, 4, 4))
boxplot(lapply(pheno.metagene.cor, abs), las = 2)
```

## Correlation Maps  {.tabset .tabset-fade .tabset-pills}

The following plots show maps of the correlations between traits and metagenes.
Each trait has roughly the same correlation pattern with the metagenes. 

In one tab we show the correlations all on the same scale to show that 
correlations with diet days, for example, are weaker than those for the
HOMAs.

The second tab shows each correlation plot on its own scale to better highligh
the overall patterns in the correlations. These plots show that the islet traits,
WPIC, Ins_per_islet, and num_islets, tend to have one correlation pattern with
the transcriptome, while the rest of the traits have a different pattern. 

Overall the patterns are strikingly similar. The correlations are on par with
those seen in WGCNA modules. However in this case, the number of genes in each 
metagene tends to be smaller than the number of genes in WGCNA modules. This is
neither good nor bad.

The correlations are much lower than I had thought they would be, and are
also much lower than what we saw with CCA. With CCA we maximize the correlation
between transcripts and traits. We were able to achive correlations from -0.6 
to 0.6 across all tissues. Here, when we cluster the transcripts without taking
the traits into account, we get much lower correlations. 

The following histogram shows the distribution of correlations between
all metagenes and all traits for `r tissue.name`.

```{r cor_dist}
hist(unlist(pheno.metagene.cor), main = "Trait-Metagene Correlations",
xlab = "Pearson Correlation", breaks = 100)
```

### Global Color Scale
```{r metagene_cor_maps, fig.width = 10, fig.height = 9}

min.cor <- min(unlist(pheno.metagene.cor), na.rm = TRUE)
max.cor <- max(unlist(pheno.metagene.cor), na.rm = TRUE)

layout.mat <- get.layout.mat((length(pheno.metagene.cor)+1))

layout(layout.mat)
par(mar = c(0,0,4,0))
for(m in 1:length(pheno.metagene.cor)){
  plot_portrait(pheno.metagene.cor[[m]], rotate = 3, 
    global.min = min.cor, global.max = max.cor, main = colnames(adj.pheno)[m],
    show.axes = TRUE)
  }
par(mar = c(0,4,4,4))
imageWithTextColorbar(matrix(segment_region(min.cor, max.cor, 10), ncol = 1), 
  use.pheatmap.colors = TRUE, cex = 1)



```

### Individual Color Scale

```{r, ind_color_scale, fig.width = 10, fig.height = 9}
layout(layout.mat)
par(mar = c(0,0,4,0))
for(m in 1:length(pheno.metagene.cor)){
  plot_portrait(pheno.metagene.cor[[m]], rotate = 3, main = colnames(adj.pheno)[m],
  show.axes = TRUE)
}
```

## Metagene Mediation {.tabset .tabset-fade .tabset-pills}

```{r mediator_threshold}
lod.drop.thresh = 3
```

Here we use the expression of each metagene to mediate the 
`r trait.name` QTL.

The following portrait shows the maximum LOD drop by metagene. There 
are some individual metagenes that have relatively high mediation of
the trait. There are also clusters of metagenes that have similar 
mediation profiles. These do not really line up with the k-means
clusters all that well, so we shold probably look at them in their 
own right.

```{r mediate_individual_metagenes}
#adjust our trait for each metagene
all.adj.trait <- sapply(1:ncol(metagene.expr), 
  function(x) if(!all(is.na(metagene.expr[,x]))){adjust(adj.pheno[,trait.name,drop=FALSE], 
  metagene.expr[,x,drop=FALSE])}else{rep(NA, nrow(adj.pheno))})
rownames(all.adj.trait) <- rownames(adj.pheno)

has.vals <- which(apply(all.adj.trait, 2, function(x) !all(is.na(x))))
#only scanning chromosome with qtl
all.med.scan <- scan1(genoprobs[,qtl.chr], all.adj.trait[,has.vals], cores = 4)
lod.drop <- lapply(1:ncol(all.med.scan), function(x) qtl_map - all.med.scan[,x,drop=FALSE])
lod.drop.map <- rep(0, som.dim^2)
lod.drop.map[has.vals] <- sapply(lod.drop, max)

layout(matrix(c(1,1,2,0), nrow = 2), widths = c(1, 0.2), height = c(0.5, 1))
lod.drop.mat <- plot_portrait(lod.drop.map, rotate = 3, main = "Max LOD Drop", show.axes = TRUE)
par(mar = c(0,0,4,4))
imageWithTextColorbar(matrix(segment.region(min(lod.drop.map, na.rm = TRUE),  
  max(lod.drop.map, na.rm = TRUE), 100)), use.pheatmap.colors = TRUE, cex = 1)
```

We clustered the mediating metagenes into groups to test
whether as a group they might mediate the trait QTL even
further.

The following plot shows which metagenes were selected 
and how they were grouped together.

```{r cluster_metagenes, fig.height = 5, fig.width = 5}

high.lod.idx <- which(lod.drop.mat >= lod.drop.thresh)
metagene.idx <- idx_to_metagene_idx(mat.idx = high.lod.idx, som.dim = som.dim)

#group all metagenes that are above the threshold and
#within a distance of two from each other.
node.coord <- which(lod.drop.mat > lod.drop.thresh, arr.ind = TRUE)
node.blocks <- contiguous_blocks(node.coord, node.dist = 2)
cluster.mask <- matrix(0, nrow = som.dim, ncol = som.dim)
for(i in 1:length(node.blocks)){
  for(j in 1:nrow(node.blocks[[i]])){
    cluster.mask[node.blocks[[i]][j,1], node.blocks[[i]][j,2]] <- i
  }
}
if(is.interactive){quartz()}
imageWithText(cluster.mask, use.pheatmap.colors = TRUE)
```

```{r mediate_with_clusters}

cluster_meta_idx <- lapply(node.blocks, function(x) idx_to_metagene_idx(row.col = x, som.dim = som.dim))
cluster_meta_genes <- lapply(cluster_meta_idx, function(x) names(som.bmu)[which(som.bmu %in% x)])

cluster.expr <- lapply(cluster_meta_genes, function(x) adj.expr[,match(x, colnames(adj.expr)), drop = FALSE])
mean.cluster.expr <- sapply(cluster.expr, function(x) rowMeans(apply(x, 2, rankZ)))
cluster.adj <- sapply(1:ncol(mean.cluster.expr), function(x) adjust(adj.pheno[,trait.name,drop=FALSE], mean.cluster.expr[,x,drop=FALSE]))
rownames(cluster.adj) <- rownames(adj.pheno)
cluster.med <- scan1(genoprobs[,qtl.chr], cluster.adj)
cluster.lod.drop <- lapply(1:ncol(cluster.med), function(x) qtl_map - cluster.med[,x,drop=FALSE])
lod.drops <- sapply(cluster.lod.drop, max)

big.drops <- which(lod.drops > lod.drop.thresh)
```

### Trait QTL Mediation Plots {.tabset .tabset-fade .tabset-pills}

The following plots show plots for the mediated QTL for `r trait.name`.
The vertical line shows the position of the maximum LOD drop.
Only those with LOD drops of at least `r lod.drop.thresh` are shown. 
The drops are pretty modest and don't always line up with the trait 
peak QTL.

```{r plot_med, results = "asis", fig.width = 10, fig.height = 5}

for(i in big.drops){
  cat("#### Cluster", i, "\n")
  if(is.interactive){quartz()}
  plot(qtl_map, map = map)
  plot(cluster.med, lodcol = i, map = map, add = TRUE, col = "red")
  med.drop <- qtl_map - cluster.med[,i,drop=FALSE]
  peak.drop <- find_peaks(med.drop, map = map, threshold = lod.drop.thresh)
  max.pos <- peak.drop[which.max(peak.drop[,"lod"]),"pos"]
  abline(v = max.pos)
  cat("\n\n")
}

#big.drop.enrich <- lapply(cluster_meta_genes[big.drops], 
#  function(x) gost(x, organism = "mmusculus"))
#names(big.drop.enrich) <- round(lod.drops[big.drops], 2)
#if(length(unlist(big.drop.enrich)) > 0){
#  plot.enrichment.group(big.drop.enrich)
#}
```

### Reverse Mediation {.tabset .tabset-fade .tabset-pills}


It is possible that a feature that looks like a mediator
is simply highly correlated with the haplotype in the
same way as the trait, and that the apparent mediation is
caused by the tight correlation between the haplotype and
the transcript, and does not speak to a relationship between
the transcript and the trait. 

In other words, the genotype affects both the transcript 
and the trait, but the transcript and the trait are not
causally related.

We can better investigate the relationship between the
mediator and the trait by using partial correlation. 
If we remove the effect of genotype on the transcript,
the residual variance in the transcript should still 
correlate with the trait. If it doesn't, we conclue 
that the haplotype affects both transcript and trait 
independently.

The following code checks residual trait correlations
for all potential mediators.

If there is a big drop in LOD score at the position
of the cluster QTL, AND the cluster is still correlated
with the trait after accounting for genotype, then
the cluster might be a mediator. 

Do we remove the genotype from both the mediator and the 
trait, or just the mediator?

Vertical lines below show the position of peak LOD drop
from the original mediation.

```{r check_reverse, fig.width = 8, fig.height = 8, results = "asis"}
cols <- categorical_pal(8)
for(i in 1:ncol(mean.cluster.expr)){
  cat("#### Cluster", i, "\n")
  cluster.qtl <- scan1(genoprobs[,qtl.chr], mean.cluster.expr[,i])

  if(is.interactive){quartz(width = 8, height = 8)}
  layout.mat <- matrix(c(1,1,2,3), nrow = 2, byrow = TRUE)
  layout(layout.mat)
  ymax <- max(c(cluster.qtl[,1], qtl_map[,1], cluster.med[,i]))  
  plot(qtl_map, map = map, ylim = c(0, ymax))
  plot(cluster.qtl, map = map, col = cols[3], main = paste("Cluster", i), add = TRUE)  
  plot(cluster.med, map = map, lodcol = i, col = cols[1], add = TRUE)
  legend("topleft", col = c("darkblue", cols[c(1,3)]), lty = 1, 
    legend = c("Trait", "Mediated Trait", "Cluster"), lwd = 3)


  #get the genotype at the LOD drop position
  peak.table <- find_peaks(abs(qtl_map - cluster.med[,i,drop=FALSE]), map = map, 
    threshold = lod.drop.thresh)
  #peak.table <- find_peaks(cluster.qtl, map = map)
  max.idx <- which.max(peak.table[,"lod"])
  peak.marker <- map[[qtl.chr]][get.nearest.pt(map[[qtl.chr]], peak.table[max.idx,"pos"])]
  peak.geno <- genoprobs[[qtl.chr]][,,names(peak.marker)]
  abline(v = peak.table[,"pos"])

  plot.with.model(mean.cluster.expr[,i], adj.pheno[,trait.name], 
    xlab = paste("Cluster", i), ylab = trait.name, 
    main = paste("Cluster", i, "vs.", trait.name))

  #adjust the mediator for the genotype
  adj.med <- adjust(mean.cluster.expr[,i,drop=FALSE], peak.geno, retain.intercept = FALSE)
  adj.trait <- adjust(adj.pheno[,trait.name,drop=FALSE], peak.geno, retain.intercept = FALSE)
  #plot.with.model(adj.med, adj.pheno[,trait.name], xlab = paste("Cluster", i, "Adjusted"),
  #  ylab = trait.name, main = paste("Cluster", i, "Adjusted for Haplotpye"))
  plot.with.model(adj.med, adj.trait, xlab = paste("Cluster", i, "Adjusted"),
    ylab = paste(trait.name, "Adjusted"), 
    main = paste("Cluster", i, "vs.", trait.name, "Adjusted for Haplotpye"))
  cat("\n\n")

}

```


```{r slfn3_muscle, eval = FALSE}
gene.id <- "ENSMUSG00000018986"; #Slfn3
gene.idx <- which(colnames(adj.expr) == gene.id)
eqtl <- scan1(genoprobs[,qtl.chr], adj.expr[,gene.idx])
plot(eqtl, map = map)

#Slfn3 has an eQTL that matches the trait QTL on Chr11
#It mediates the QTL, but does that mean it's related 
#to the trait, or just the QTL?

eqtl.coef <- scan1coef(genoprobs[,qtl.chr], adj.expr[,gene.idx])
plot_coefCC(eqtl.coef, map = map)

plot.with.model(adj.expr[,gene.idx], adj.pheno[,trait.name], report = "cor.test")

for(i in 1:length(test.genes)){
  adj.trans <- adjust(adj.expr[,test.genes[i],drop=FALSE], peak.geno)
  plot.with.model(adj.trans, adj.pheno[,trait.name])
}

```