---
title: "SOM to the DO"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow calculate the metagene profiles 
calculated in Explore_SOM.Rmd in the DO mice. 

We first look for correlations between metagenes identified
in the CC-RIX and traits in the DO.

We then investigate whether these metagenes mediate 
any trait QTLs in the DO. 

The metagenes from the CC-RIX are identified using a 
specific group of animals with a particular contrast
in mind. For example, the metagenes might contrast 
HFD and LFD in the female animals with no metformin 
treatment. The tissue and groups used in the CC are 
specified below.

```{r get_args}
tissue.name = "adipose"
group.class1 = "Male"
group.class2 = "None"
contrast = "Diet"

is.interactive = FALSE
#is.interactive = TRUE
```

The results shown here are for `r tissue.name`.

## SOM Results

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
needed.packages <- c("gprofiler2", "pheatmap", "qtl2", "grid", "stringr",
  "RColorBrewer")
load_libraries(needed.packages, personal.library = TRUE)
```


```{r read_data}
#metagene data from CC-RIX 
all.files <- list.files(here("Results", "SOM"))
som.files <- all.files[Reduce("intersect", sapply(c(tissue.name, group.class1, group.class2, contrast), function(x) grep(x, all.files)))]
results.file.idx <- grep("RData", som.files)
results.file <- som.files[results.file.idx]
results.dir <- som.files[-results.file.idx]

#get all results from the SOM environment
load(here("Results", "SOM", results.file))
```

Get the metagene assignments.

```{r metagene_assignments}
som.result <- env$som.result
som.coord <- som.result[[1]]
som.bmu <- som.result[[2]]
som.dim <- sqrt(nrow(som.coord))


#k.means.spots <- env$spot.list.kmeans
#test <- k.means.spots$overview.map
#we need to rotate the matrix three times to make it look like
#the portraits in the oposSOM output.
#x goes from 1 to n left to right.
#y goes from 1 to n bottom to top
#This is different than R matrices, which start counting from the
#top left, so we need to account for this.
#imageWithText(rotate.mat(rotate.mat(rotate.mat(test))), col.scale = "blue")


#use the best matching units to assign genes to positions
cc.metagenes <- lapply(1:(som.dim^2), function(x) names(som.bmu)[which(som.bmu == x)])
names(cc.metagenes) <- apply(som.coord, 1, function(x) paste(x[1:2], collapse = "_"))

cc.expr <- env$indata #expression data from CC-RIX. Input data to SOM
cc.groups <- env$group.labels #individual assignments to groups
u_groups <- unique(cc.groups)

group.expr <- lapply(u_groups, function(x) lapply(cc.metagenes, function(y) cc.expr[match(y, rownames(cc.expr)), which(cc.groups == x),drop=FALSE]))
cc.profiles <- lapply(group.expr, function(x) sapply(x, colMeans))
names(cc.profiles) <- u_groups

group.grid <- read.csv(here("Data", "CC-RIX", "Groups.csv"), row.names = 1)
group.def <- colnames(group.grid)

all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.DO.Cube.Adipose")))
if(!data.loaded){
    do.data <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
}

tissue.idx <- grep(tissue.name, do.data, ignore.case = TRUE)
tissue.data <- get(do.data[tissue.idx])

tissue.expr <- tissue.data$data$rz
tissue.covar <- tissue.data$covar.matrix

#match gene expression and covariates along individuals
matched.mats <- get.xz(tissue.expr, tissue.covar)
matched.expr <- matched.mats[[1]]
matched.covar <- matched.mats[[2]]

gene.info <- tissue.data$annot.mrna

pheno <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE))
num.pheno <- apply(pheno[,11:(ncol(pheno)-1)], 2, as.numeric)
rownames(num.pheno) <- pheno[,1]
```

We will look at males and females separately as we did 
with the CC-RIX, so we won't adjust out sex. Just adjust
for Generation.

```{r adjust}
gen.locale <- grep("generation", colnames(matched.covar), ignore.case = TRUE)
adj.expr <- adjust(matched.expr, matched.covar[,gen.locale])
f.locale <- which(matched.covar[,1] == 0)
m.locale <- which(matched.covar[,1] == 1)

#regress out diet days
num.pheno <- adjust(num.pheno[,2:ncol(num.pheno)], num.pheno[,1,drop=FALSE]) 
adj.pheno <- adjust(num.pheno, matched.covar[,gen.locale])
```

## Calculate DO Metagenes

Pull out metagenes and spots for this tissue.

There is one set of metagenes for the particular analysis. 
The only thing that varies in the DO is whether we are looking
at male or female mice. 

In the CC-RIX, there are lots of different groups,
each of which has its own metagene expression. 


```{r do_metagenes}
get.do.metagene <- function(gene.names, group.expr){
    gene.idx <- match(gene.names, colnames(group.expr))
    gene.idx <- gene.idx[which(!is.na(gene.idx))]
    return(group.expr[,gene.idx,drop=FALSE])
}

groups <- names(cc.profiles)

#get DO metagene values for males and females separately
#the metagenes contain the same genes across all groups
#the only difference is which DO individuals are included
#The only groups in the DO are the two sexes.
sub.expr <- lapply(c(0,1), function(x) adj.expr[which(matched.covar[,1] == x),])
names(sub.expr) <- c("Female", "Male")

do.metagenes <- lapply(sub.expr, function(y) lapply(cc.metagenes, 
    function(x) get.do.metagene(x, y)))
```

Calculate the correlations between the metagene profiles from the 
CC-RIX and the do metagenes.

The following portraits show the CC-RIX metagenes compared 
with the mean DO metagenes for the groups that are most 
closely matched. All DO were on high fat diet with no metformin 
treatment, so here we only compare those groups in males and 
females.

```{r metagene_profile_comparison, fig.width = 8, fig.height = 4}
do.metagene.cor <- vector(mode = "list", length = length(groups))
names(do.metagene.cor) <- groups
do.profile <- lapply(do.metagenes, function(x) sapply(x, rowMeans))

#has.genes <- which(apply(do.profile[[1]], 2, function(x) !all(is.na(x))))
#plot.decomp(do.profile[[1]][,has.genes], pc = 4)
#plot.decomp(do.profile[[2]][,has.genes], pc = 4)

if(is.interactive){quartz(width = 8, height = 4)}
par(mfrow = c(1,2))
plot_portrait(colMeans(do.profile[[1]]), main = paste("DO", names(do.profile)[1]), rotate = 3)
plot_portrait(colMeans(do.profile[[2]]), main = paste("DO", names(do.profile)[2]), rotate = 3)

mean.cc.profiles <- lapply(cc.profiles, colMeans)
do.metagene.cor <- lapply(do.profile, 
    function(y) apply(y, 1, function(x) sapply(mean.cc.profiles, 
    function(y) cor(x, y, use = "pairwise.complete.obs"))))

```

The following boxplot shows the distribution of correlations between
DO metagenes and CC-RIX metagenes for males and females.

```{r, similarity_dist, fig.width = 9, fig.height = 6}
for(g in 1:length(do.metagene.cor)){
    if(is.interactive){quartz()}
    boxplot(t(do.metagene.cor[[g]]), ylim = c(-0.8, 0.8), main = names(do.metagene.cor)[g])
    abline(h = 0)
}
```

## Metagene-Trait Correlations  {.tabset .tabset-fade .tabset-pills}

Are individual metagenes in the DO correlated with 
traits?

The boxplots below show the distributions of correlations
between metagenes and traits for each group

```{r cor_weight, fig.width = 9, fig.height = 6}

#calculate mean metagene profiles for each DO individual

common.ind <- lapply(do.profile, function(x) intersect(rownames(x), rownames(adj.pheno)))
do.idx <- lapply(1:length(common.ind), function(x) match(common.ind[[x]], rownames(do.profile[[x]])))
pheno.idx <- lapply(common.ind, function(x) match(x, rownames(adj.pheno)))

metagene.cor.mat <- lapply(1:length(common.ind), 
    function(z) apply(adj.pheno[pheno.idx[[z]],], 2, 
    function(x) apply(do.profile[[z]][do.idx[[z]],], 2, 
    function(y) cor(x, y, use = "pairwise.complete.obs"))))

if(is.interactive){quartz(width = 9, height = 6)}
par(mar = c(8, 4, 4, 4))
boxplot(abs(metagene.cor.mat[[1]]), las = 2, main = names(do.profile)[1])
boxplot(abs(metagene.cor.mat[[2]]), las = 2, main = names(do.profile)[2])
```

### Correlation Portraits {.tabset .tabset-fade .tabset-pills}

The following plots show the correlation portraits for males and females.

```{r cor_profiles, results = "asis"}
for(g in 1:length(do.profile)){
    cat("####", names(do.profile)[g], "\n")
    par(mfrow = c(4,5), mar = c(0,1,2,1))
    global.min = min(unlist(metagene.cor.mat[[g]]), na.rm = TRUE)
    global.max = max(unlist(metagene.cor.mat[[g]]), na.rm = TRUE)
    for(i in 1:ncol(metagene.cor.mat[[g]])){
        plot_portrait(metagene.cor.mat[[g]][,i], global.min = global.min, 
            global.max = global.max, main = colnames(metagene.cor.mat[[g]])[i])
    }
    cat("\n\n")
}

```


## Mediating with Metagenes {.tabset .tabset-fade .tabset-pills}

Below we look for metagenes that mediate trait QTLs. 

```{r map_traits}
#scan the physiological traits
trait.map <- lapply(c(0,1), 
    function(x) scan1(genoprobs, adj.pheno[which(matched.covar[,1] == x),], cores = 4))
```

```{r map_meta, fig.width = 10, fig.height = 7, results = "asis"}
cor.thresh = 0.6
qtl.thresh <- 6
lod.drop.thresh = 4

#par(mfrow = c(2,1))
#lodcol = 5
#plot(trait.map[[1]], map = map, lodcol = lodcol, main = colnames(trait.map[[1]])[lodcol])
#plot(trait.map[[2]], map = map, lodcol = lodcol, main = colnames(trait.map[[1]])[lodcol])


for(g in 1:length(do.profile)){
    group.name <- names(do.profile)[g]
    cat("###", group.name, "{.tabset .tabset-fade .tabset-pills}\n")

    #plot(sort(abs(metagene.cor.mat[[g]])));abline(h = cor.thresh)
    #find metagenes with high correlations to each trait
    high.cor.idx <- apply(metagene.cor.mat[[g]], 2, function(x) which(x >= cor.thresh))
    has.vals <- which(sapply(high.cor.idx, length) > 0)    

    if(length(has.vals) == 0){next()}

    #loop through traits and do mediation separately for each one
    for(tr in has.vals){   
        if(length(high.cor.idx[[tr]]) == 0){next()}
        spot.mat <- do.profile[[g]][,high.cor.idx[[tr]],drop=FALSE]
        if(ncol(spot.mat) == 1){
        meta.spot <- spot.mat
        }else{
            meta.spot  <- plot.decomp(spot.mat, plot.results = FALSE)$u[,1,drop=FALSE]
        }
        #meta.spot <- matrix(rowMeans(spot.mat), ncol = 1)
        rownames(meta.spot) <- rownames(do.profile[[g]])
        adj.traits <- adjust(adj.pheno[,tr,drop=FALSE], meta.spot)

        med.scan <- scan1(genoprobs, adj.traits, cores = 4)

        lod.drop <- trait.map[[g]][,tr,drop=FALSE] - med.scan
        all.max.min <- t(apply(lod.drop, 2, function(x) c(min(x), max(x))))
        colnames(all.max.min) <- c("min_drop", "max_drop")
    
        drop_peaks <- find_peaks(lod.drop, map = map, threshold = lod.drop.thresh)

        if(nrow(drop_peaks) == 0){next()}

        cat("####", colnames(adj.pheno)[tr], "\n")
        if(is.interactive){quartz(width = 9, height = 7)}

        #plot the original scan for the trait overlayed
        #with the mediated scan in red
        par(mfrow = c(2,1), mar = c(4,4,2,4))
        plot(trait.map[[g]], map = map, lodcol = tr, main = colnames(adj.pheno)[tr])
        plot(med.scan, map = map, col = "red", add = TRUE)
        abline(h = qtl.thresh)
        #also plot the lod drops along the genome
        plot(lod.drop, map = map, ylim = all.max.min*1.1, main = "LOD Drop")
        abline(h = lod.drop.thresh)

        drop.chr <- drop_peaks[,"chr"]

        for(sm in 1:length(drop.chr)){
            if(is.interactive){quartz(width = 9, height = 7)}
            layout(matrix(c(1,1,2,3), nrow = 2, byrow = TRUE), widths = c(0.4, 0.6))
            plot(trait.map[[g]], map = map, lodcol = tr, chr = drop.chr[[sm]], 
                main = paste(colnames(adj.traits), "Mediation with Chr", drop.chr[sm] ,"subset"))
            plot(med.scan, map = map, col = "red", add = TRUE, chr = drop.chr[[sm]])
            #spot.row.col <- som.coord[high.cor.idx[[tr]],]
            #spot.idx <- row_col_to_idx(spot.row.col[,1], spot.row.col[,2], som.dim)
            spot.idx <- high.cor.idx[[tr]]
            mediator.mask <- rep(0, som.dim^2)
            mediator.mask[high.cor.idx[[tr]]] <- 1
            plot_portrait(mediator.mask, rotate = 3)
            meta.genes <- unlist(lapply(spot.idx, function(x) colnames(do.metagenes[[g]][x][[1]])))
            gene.enrich <- gost(meta.genes, organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME"))
            plot.enrichment(gene.enrich, num.term = 15, max.term.size = 2000, order.by = "p_value")
            #plot.text(paste(colnames(meta.spot)[drop.chr.locale[[sm]]], collapse = "\n"))
            cat("\n\n")
        } #end looping through chromosomes with big lod drops
        cat("\n\n")
    } #end looping through traits

    cat("\n\n")
} #end looping through sexes
```

```{r mapping_extra, eval = FALSE}
find_peaks(group.scan, map = map, threshold = 6)

coef.scan <- scan1coef(genoprobs[,"X"], group.scores, kinship = K[[20]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

coef.scan <- scan1coef(genoprobs[,"11"], group.scores, kinship = K[[11]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

```

```{r DO_portraits, eval = FALSE}
test <- sapply(do.metagenes[[group.locale]], rowMeans)
par(mfrow = c(5,5), mar = c(1,1,1,1))
for(i in 26:50){
    plot_portrait(test[i,])
}
```