---
title: "SOM to the DO"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow calculate the metagene profiles 
calculated in Explore_SOM.Rmd in the DO mice. 

We first look for correlations between metagenes identified
in the CC-RIX and traits in the DO.

We then investigate whether these metagenes mediate 
any trait QTLs in the DO. 

The metagenes from the CC-RIX are identified using a 
specific group of animals with a particular contrast
in mind. For example, the metagenes might contrast 
HFD and LFD in the female animals with no metformin 
treatment. The tissue and groups used in the CC are 
specified below.

```{r get_args}
tissue.name = "adipose"
contrast = "Diet"

is.interactive = FALSE
#is.interactive = TRUE
```

The results shown here are for `r tissue.name`.

## SOM Results

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
needed.packages <- c("gprofiler2", "pheatmap", "qtl2", "grid", "stringr",
  "RColorBrewer")
load_libraries(needed.packages, personal.library = TRUE)
```


```{r read_data}
#metagene data from CC-RIX 
all.files <- list.files(here("Results", "SOM"))
som.files <- all.files[Reduce("intersect", sapply(c(tissue.name, contrast), function(x) grep(x, all.files)))]
results.file.idx <- grep("RData", som.files)
results.file <- som.files[results.file.idx]
#results.dir <- som.files[-results.file.idx]

#get all results from the SOM environment
load(here("Results", "SOM", results.file))
```

Get the metagene assignments.

```{r metagene_assignments}
som.result <- env$som.result
som.coord <- som.result[[1]]
som.bmu <- som.result[[2]]
som.dim <- sqrt(nrow(som.coord))

#k.means.spots <- env$spot.list.kmeans
#test <- k.means.spots$overview.map
#we need to rotate the matrix three times to make it look like
#the portraits in the oposSOM output.
#x goes from 1 to n left to right.
#y goes from 1 to n bottom to top
#This is different than R matrices, which start counting from the
#top left, so we need to account for this.
#imageWithText(rotate.mat(rotate.mat(rotate.mat(test))), col.scale = "blue")


#use the best matching units to assign genes to positions
cc.metagenes <- lapply(1:(som.dim^2), function(x) names(som.bmu)[which(som.bmu == x)])
names(cc.metagenes) <- apply(som.coord, 1, function(x) paste(x[1:2], collapse = "_"))

cc.expr <- env$indata #expression data from CC-RIX. Input data to SOM
cc.groups <- env$group.labels #individual assignments to groups
u_groups <- unique(cc.groups)
```

```{r expr_pc, warning = FALSE}
#This function takes in a matrix with individuals
#in rows and metagene expression in columns. It
#performs svd and returns the first PC of the 
#group to use as a portrait representative of the
#full group.
group_profile <- function(profile.group){
    if(ncol(profile.group) == 0){
        return(NULL)
    }
    if(ncol(profile.group) == 1){
        return(profile.group)
    }
    decomp <- plot.decomp(t(profile.group), plot.results = FALSE)
    profile.pc <- rep(NA, som.dim^2)
    profile.pc[decomp$rows.used] <- decomp$u[,1]
    return(profile.pc)
}
```

```{r cc_profiles}
#get metagene expression for all individuals and all metagenes
cc.metagene.expr <- lapply(u_groups, function(x) lapply(cc.metagenes, function(y) cc.expr[match(y, rownames(cc.expr)), which(cc.groups == x),drop=FALSE]))
#create matrix of mean expression for each metagene by individual for each group
cc.metagene.profiles <- lapply(cc.metagene.expr, function(x) sapply(x, colMeans))
names(cc.metagene.profiles) <- u_groups

group.grid <- read.csv(here("Data", "CC-RIX", "Groups.csv"), row.names = 1)
group.def <- colnames(group.grid)
```

```{r do_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.DO.Cube.Adipose")))
if(!data.loaded){
    do.data <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
}

tissue.idx <- grep(tissue.name, do.data, ignore.case = TRUE)
tissue.data <- get(do.data[tissue.idx])

tissue.expr <- tissue.data$data$rz
tissue.covar <- tissue.data$covar.matrix

#match gene expression and covariates along individuals
matched.mats <- get.xz(tissue.expr, tissue.covar)
matched.expr <- matched.mats[[1]]
matched.covar <- matched.mats[[2]]

gene.info <- as.matrix(tissue.data$annot.mrna)

pheno <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE))
num.pheno <- apply(pheno[,11:(ncol(pheno)-1)], 2, as.numeric)
rownames(num.pheno) <- pheno[,1]
```

Put all individuals together to increase power. Adjust out
sex and generation.

```{r adjust}
gen.locale <- grep("generation", colnames(matched.covar), ignore.case = TRUE)
#adjust for Sex and Generation
adj.expr <- adjust(matched.expr, matched.covar)

#in addition, regress out diet days from the traits
num.pheno <- adjust(num.pheno[,2:ncol(num.pheno)], num.pheno[,1,drop=FALSE]) 
adj.pheno <- adjust(num.pheno, matched.covar)
```

The following portraits show the profile for each group
in the CC-RIX.

```{r cc_group_portraits, fig.width = 8, fig.height = 4}
#PC profiles better match the profiles from oposSOM thans means
cc.pc.profiles <- sapply(cc.metagene.profiles, group_profile)

global.max <- max(unlist(cc.pc.profiles), na.rm = TRUE)
global.min <- min(unlist(cc.pc.profiles), na.rm = TRUE)
layout.mat <- get.layout.mat(ncol(cc.pc.profiles))
layout(layout.mat)
par(mar = c(1,1,3,1))
for(i in 1:ncol(cc.pc.profiles)){
    plot_portrait(cc.pc.profiles[,i], rotate = 3, main = colnames(cc.pc.profiles)[i],
    global.min = global.min, global.max = global.max)
}
```

## Calculate DO Metagenes

Pull out metagenes and spots for this tissue.

There is one set of metagenes for the particular analysis. 
The only thing that varies in the DO is whether we are looking
at male or female mice. 

In the CC-RIX, there are lots of different groups,
each of which has its own metagene expression. 


```{r do_metagenes}
get.do.metagene <- function(gene.names, group.expr){
    gene.idx <- match(gene.names, colnames(group.expr))
    gene.idx <- gene.idx[which(!is.na(gene.idx))]
    return(group.expr[,gene.idx,drop=FALSE])
}

#get DO metagene values. There are no treatment groups
#in the DO. We adjusted for sex and generation.
do.metagenes <- lapply(cc.metagenes, function(x) get.do.metagene(x, adj.expr))
do.profiles <- sapply(do.metagenes, rowMeans)
do.pc.profile <- group_profile(do.profiles)
if(is.interactive){quartz()}
plot_portrait(do.pc.profile, rotate = 3, main = "Overall DO Profile")
```

Calculate the correlations between profiles of the DO and CC-RIX groups.
The following boxplot shows the distribution of correlations between the 
DO metagene profiles and the CC metagene profiles. 
We use the absolute value because we are calculating group profiles using
PCs, and the 

There is no obvious 
trend. I expected the DO animals to have profiles close to those of the 
HFD CC-RIX, but that doesn't seem to be true overall.

```{r metagene_profile_comparison}
do.metagene.cor <- lapply(1:ncol(cc.pc.profiles), 
    function(x) apply(do.profiles, 1, 
    function(y) cor(cc.pc.profiles[,x], y, use = "pairwise.complete.obs")))
names(do.metagene.cor) <- u_groups

par(mar = c(10, 4, 4, 4))
boxplot(lapply(do.metagene.cor, abs), las = 2)
```


## Metagene-Trait Correlations  {.tabset .tabset-fade .tabset-pills}

Are metagene profiles in the DO correlated with traits?

The boxplots below show the distributions of correlations
between metagenes and traits for each group.

```{r cor_weight, fig.width = 9, fig.height = 6}
common.ind <- intersect(rownames(do.profiles), rownames(adj.pheno))
do.idx <- match(common.ind, rownames(do.profiles))
pheno.idx <- match(common.ind, rownames(adj.pheno))

metagene.trait.cor <- apply(adj.pheno[pheno.idx,], 2, 
    function(x) apply(do.profiles[do.idx,], 2, 
    function(y) cor(rankZ(x),rankZ(y), use = "pairwise.complete.obs")))


if(is.interactive){quartz(width = 9, height = 6)}
par(mar = c(8, 4, 4, 4))
boxplot(abs(metagene.trait.cor), las = 2, main = "DO Metagene-Trait Correlations")
```

### Correlation Portraits {.tabset .tabset-fade .tabset-pills}

The following plots show the correlation portraits for the DO.
These show where metagenes with strong correlations to traits
are located.

```{r cor_profiles, results = "asis"}
par(mfrow = c(4,5), mar = c(0,1,2,1))
global.min = min(unlist(metagene.trait.cor), na.rm = TRUE)
global.max = max(unlist(metagene.trait.cor), na.rm = TRUE)
for(i in 1:ncol(metagene.trait.cor)){
    plot_portrait(metagene.trait.cor[,i], global.min = global.min, 
        global.max = global.max, main = colnames(metagene.trait.cor)[i],
        rotate = 3)
}
```


## Mediating with Metagenes {.tabset .tabset-fade .tabset-pills}

Below we look for metagenes that mediate trait QTLs. 

```{r map_traits}
#scan the physiological traits
trait.map <- scan1(genoprobs, adj.pheno, cores = 4)
```

```{r map_meta, fig.width = 10, fig.height = 7, results = "asis"}
qtl.thresh <- 6
lod.drop.thresh = 4

#map all metagenes
metagene.eqtl.file <- here("Results", "SOM", paste0("DO.", tissue.name, ".Metagene.eQTL.RDS"))
no.na <- which(apply(do.profiles, 2, function(x) all(!is.na(x))))
metagene.bins <- bin.sequence(no.na, 100)
if(!file.exists(metagene.eqtl.file)){
    metagene.eqtl <- NULL
    for(i in 1:length(metagene.bins)){
        report.progress(i, length(metagene.bins))
        metagene.map <- scan1(genoprobs, do.profiles[,metagene.bins[[i]]], cores = 4)
        metagene.eqtl <- rbind(metagene.eqtl, find_peaks(metagene.map, map = map, threshold = qtl.thresh))
    }
    saveRDS(metagene.eqtl, metagene.eqtl.file)
}else{
    metagene.eqtl <- readRDS(metagene.eqtl.file)
}

#find metagenes with high correlations to each trait
#high.cor.idx <- apply(metagene.trait.cor, 2, function(x) which(x >= cor.thresh))
#has.vals <- which(sapply(high.cor.idx, length) > 0)    

#find all traits with QTL LOD scores greater than the specified threshold
trait.peaks <- find_peaks(trait.map, map = map, threshold = qtl.thresh)
traits.with.peaks <- unique(trait.peaks[,2])
trait.idx <- match(traits.with.peaks, colnames(adj.pheno))

#loop through traits and do mediation separately for each one
for(tr in trait.idx){

    #if(length(high.cor.idx[[tr]]) == 0){next()}
    trait.qtls <- trait.peaks[which(trait.peaks[,2] == colnames(adj.pheno)[tr]),,drop=FALSE]

    if(nrow(trait.qtls) == 0){next()}

    #for each QTL, identify metagenes with overlapping QTL
    overlapping.metagenes <- lapply(1:nrow(trait.qtls), 
        function(x) overlapping_peaks(trait.qtls[x,"chr"], trait.qtls[x,"pos"]-5, 
        trait.qtls[x,"pos"]+5, metagene.eqtl))

    meta.row.col <- lapply(overlapping.metagenes, function(x) t(sapply(strsplit(x[,"lodcolumn"], "_"), function(y) as.numeric(y))))
    meta.idx <- lapply(meta.row.col, function(x) row_col_to_idx(x[,1], x[,2], som.dim))
    #check to make sure we indexed the metagenes correction and that they actually map where we think they do
    #test <- scan1(genoprobs, do.profiles[,meta.idx[[2]],drop=FALSE])
    #multilod.plot(test, map = map, lod.thresh = 6)

    spot.mat <- do.profiles[,sort(unique(unlist(meta.idx)))]
    #sapply(spot.mat, dim)
    
    #adjust trait for individual metagenes individually
    adj.traits <- sapply(1:ncol(spot.mat), function(x) adjust(adj.pheno[,tr,drop=FALSE], spot.mat[,x,drop=FALSE]))
    rownames(adj.traits) <- rownames(adj.pheno)
    colnames(adj.traits) <- colnames(spot.mat)
    
    #and scan again
    med.scan <- scan1(genoprobs, adj.traits, cores = 4)
    
    #check for LOD drops - mediated QTL
    lod.drop <- lapply(1:ncol(med.scan), function(x) trait.map[,tr,drop=FALSE] - med.scan[,x,drop=FALSE])
    all.max.min <- t(sapply(lod.drop, function(x) c(min(x), max(x))))
    #boxplot(all.max.min)

    drop_peaks <- lapply(lod.drop, 
        function(x) find_peaks(x, map = map, threshold = lod.drop.thresh))
    big.drop.idx <- which(sapply(drop_peaks, nrow) > 0)
    #drop.table <- Reduce("rbind", drop_peaks[big.drop.idx])

    if(sum(sapply(drop_peaks, nrow)) == 0){next()}

    cat("###", colnames(adj.pheno)[tr], "{.tabset .tabset-fade .tabset-pills}\n")

    for(sm in 1:length(big.drop.idx)){

        cat("\n\n")

        cat("#### Mediation on Chr", drop_peaks[[big.drop.idx[sm]]][,"chr"], "\n")
        if(is.interactive){quartz(width = 9, height = 7)}
        layout(matrix(c(1,1,2,3), nrow = 2, byrow = TRUE), widths = c(0.4, 0.6))
        med.chr <- drop_peaks[[big.drop.idx[sm]]][,"chr"]

        plot(trait.map, map = map, lodcol = tr, chr = med.chr, 
            main = paste(colnames(adj.traits)[big.drop.idx[sm]], "Mediation for Chr", 
            paste(med.chr, collapse = ", ")))
        plot(med.scan, lodcol = big.drop.idx[sm], map = map, col = "red", add = TRUE, 
            chr = drop_peaks[[big.drop.idx[sm]]][,"chr"])
        abline(h = qtl.thresh)

        mediator.mask <- rep(0, som.dim^2)
        metagene.name <- colnames(adj.traits)[big.drop.idx[sm]]
        mediator.mask[unlist(meta.idx)[big.drop.idx[sm]]] <- 1
        plot_portrait(mediator.mask, rotate = 3)
        meta.genes <- colnames(do.metagenes[[which(names(do.metagenes) == metagene.name)]])
        gene.enrich <- gost(meta.genes, organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME"))
        plot.enrichment(gene.enrich, num.term = 15, max.term.size = 2000, order.by = "p_value")
        cat("\n\n")
    } #end looping through chromosomes with big lod drops
    cat("\n\n")
} #end looping through traits

```

```{r mapping_extra, eval = FALSE}
find_peaks(group.scan, map = map, threshold = 6)

coef.scan <- scan1coef(genoprobs[,"X"], group.scores, kinship = K[[20]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

coef.scan <- scan1coef(genoprobs[,"11"], group.scores, kinship = K[[11]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

```

```{r DO_portraits, eval = FALSE}
test <- sapply(do.metagenes[[group.locale]], rowMeans)
par(mfrow = c(5,5), mar = c(1,1,1,1))
for(i in 26:50){
    plot_portrait(test[i,])
}
```