---
title: "SOM to the DO"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow export the metagene profiles 
calculated in Explore_SOM.Rmd to the DO mice. 

We want to see whether mice with gene expression
profiles more highly correlated with the LFD CC-RIX
have lower body weight than those more highly correlated
with the HFD CC-RIX. If this is the case, we can map
with that correlation to see if there are any loci 
that generate LFD-like expression profiles even 
when exposed to HFD.


```{r get_args}
tissue.name = "adipose"
exp.type = "Groups"
#exp.type = "Differences"
is.interactive = FALSE
#is.interactive = TRUE
```

The results shown here are for `r tissue.name`.

## SOM Results

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
needed.packages <- c("gprofiler2", "pheatmap", "qtl2", "grid", "stringr",
  "RColorBrewer")
load_libraries(needed.packages, personal.library = TRUE)
```


```{r read_data}
#metagene data from CC-RIX 
cc.metagenes <- readRDS(here("Results", "SOM", paste0(exp.type, ".Metagenes.", tissue.name, ".RDS")))
group.grid <- read.csv(here("Data", "CC-RIX", "Groups.csv"), row.names = 1)
group.def <- colnames(group.grid)

all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.DO.Cube.Adipose")))
if(!data.loaded){
    do.data <- load(here("Data", "dataset.DO.CUBE.multissue.RData"))
}

tissue.idx <- grep(tissue.name, do.data, ignore.case = TRUE)
tissue.data <- get(do.data[tissue.idx])

tissue.expr <- tissue.data$data$rz
tissue.covar <- tissue.data$covar.matrix

#match gene expression and covariates along individuals
matched.mats <- get.xz(tissue.expr, tissue.covar)
matched.expr <- matched.mats[[1]]
matched.covar <- matched.mats[[2]]

gene.info <- tissue.data$annot.mrna

pheno <- as.matrix(read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE))
num.pheno <- apply(pheno[,11:(ncol(pheno)-1)], 2, as.numeric)
rownames(num.pheno) <- pheno[,1]
```

We will look at males and females separately as we did 
with the CC-RIX, so we won't adjust out sex. Just adjust
for Generation.

```{r adjust}
gen.locale <- grep("generation", colnames(matched.covar), ignore.case = TRUE)
adj.expr <- adjust(matched.expr, matched.covar[,gen.locale])
f.locale <- which(matched.covar[,1] == 0)
m.locale <- which(matched.covar[,1] == 1)

#regress out diet days
num.pheno <- adjust(num.pheno[,2:ncol(num.pheno)], num.pheno[,1,drop=FALSE]) 
adj.pheno <- adjust(num.pheno, matched.covar[,gen.locale])
```

## Calculate DO Metagenes and Spots

Pull out metagenes and spots for this tissue.

There is one set of metagenes for the particular analysis. 
The only thing that varies in the DO is whether we are looking
at male or female mice. 

In the CC-RIX, there are lots of different groups,
each of which has its own metagene expression. 


```{r do_metagenes}
get.do.metagene <- function(gene.names, group.expr){
    gene.idx <- match(gene.names, colnames(group.expr))
    gene.idx <- gene.idx[which(!is.na(gene.idx))]
    return(group.expr[,gene.idx,drop=FALSE])
}

get.do.spots <- function(group.metagenes, group.expr){
    spots <- sapply(lapply(names(group.metagenes), function(x) strsplit(x, " ")), function(y) y[[1]][4])
    u_spots <- sort(unique(spots))
    spot.genes <- lapply(u_spots, function(x) unlist(lapply(group.metagenes[which(spots == x)], names)))
    spot.expr <- lapply(spot.genes, function(x) get.do.metagene(x, group.expr))
    names(spot.expr) <- u_spots
    return(spot.expr)
}


groups <- names(cc.metagenes)

#get DO metagene values for males and females separately
#the metagenes contain the same genes across all groups
#the only difference is which DO individuals are included
#The only groups in the DO are the two sexes.
sub.expr <- lapply(c(0,1), function(x) adj.expr[which(matched.covar[,1] == x),])
names(sub.expr) <- c("Female", "Male")

do.metagenes <- lapply(sub.expr, function(y) lapply(cc.metagenes[[1]], 
    function(x) get.do.metagene(names(x), y)))
do.spots <- lapply(sub.expr, function(x) get.do.spots(cc.metagenes[[1]], x))
```

Calculate the correlations between the metagene profiles from the CC-RIX
and the do metagenes.

The following portraits show the CC-RIX metagenes compared 
with the mean DO metagenes for the groups that are most 
closely matched. All DO were on high fat diet with no metformin 
treatment, so here we only compare those groups in males and 
females.

```{r metagene_profile_comparison, fig.width = 8, fig.height = 4}
do.metagene.cor <- vector(mode = "list", length = length(groups))
names(do.metagene.cor) <- groups
do.profile <- lapply(do.metagenes, function(x) sapply(x, rowMeans))

#has.genes <- which(apply(do.profile[[1]], 2, function(x) !all(is.na(x))))
#plot.decomp(do.profile[[1]][,has.genes], pc = 4)
#plot.decomp(do.profile[[2]][,has.genes], pc = 4)

for(g in 1:length(groups)){
    cc.profile <- sapply(cc.metagenes[[g]], mean)    
    if(g == 1){
        if(is.interactive){quartz(width = 8, height = 4)}
        par(mfrow = c(1,2))
        plot_portrait(colMeans(do.profile[[1]]), main = paste("DO", names(do.profile)[1]))
        plot_portrait(colMeans(do.profile[[2]]), main = paste("DO", names(do.profile)[2]))
    }
    do.metagene.cor[[g]] <- lapply(do.profile, function(y) apply(y, 1, function(x) cor(x, cc.profile, use = "pairwise.complete.obs")))
}
```

The following boxplot shows the distribution of correlations between
DO metagenes and CC-RIX metagenes for each group.

If we are looking at group differences, the correlations below tell
us how ...

```{r, similarity_dist, fig.width = 9, fig.height = 6}
par(mfrow = c(2,4))
for(g in 1:length(do.metagene.cor)){
    boxplot(do.metagene.cor[[g]], main = groups[g], ylim = c(-0.8, 0.8))
    abline(h = 0)
}
```

## Metagene-Trait Correlations  {.tabset .tabset-fade .tabset-pills}

Are individual metagenes in the DO correlated with 
traits?

The boxplots below show the distributions of correlations
between metagenes and traits for each group

```{r cor_weight, fig.width = 9, fig.height = 6}

#calculate mean metagene profiles for each DO individual

common.ind <- lapply(do.profile, function(x) intersect(rownames(x), rownames(adj.pheno)))
do.idx <- lapply(1:length(common.ind), function(x) match(common.ind[[x]], rownames(do.profile[[x]])))
pheno.idx <- lapply(common.ind, function(x) match(x, rownames(adj.pheno)))

metagene.cor.mat <- lapply(1:length(common.ind), 
    function(z) apply(adj.pheno[pheno.idx[[z]],], 2, 
    function(x) apply(do.profile[[z]][do.idx[[z]],], 2, 
    function(y) cor(x, y, use = "pairwise.complete.obs"))))

if(is.interactive){quartz(width = 9, height = 6)}
par(mar = c(8, 4, 4, 4))
boxplot(abs(metagene.cor.mat[[1]]), las = 2, main = names(do.profile)[1])
boxplot(abs(metagene.cor.mat[[2]]), las = 2, main = names(do.profile)[2])
```

### Correlation Portraits {.tabset .tabset-fade .tabset-pills}

The following plots show the correlation portraits for all groups.

```{r cor_profiles, results = "asis"}
for(g in 1:length(do.profile)){
    cat("####", names(do.profile)[g], "\n")
    par(mfrow = c(4,5), mar = c(0,1,2,1))
    global.min = min(unlist(metagene.cor.mat[[g]]), na.rm = TRUE)
    global.max = max(unlist(metagene.cor.mat[[g]]), na.rm = TRUE)
    for(i in 1:ncol(metagene.cor.mat[[g]])){
        plot_portrait(metagene.cor.mat[[g]][,i], global.min = global.min, 
            global.max = global.max, main = colnames(metagene.cor.mat[[g]])[i])
    }
    cat("\n\n")
}

```


## Mediating with Metagenes {.tabset .tabset-fade .tabset-pills}

Below we look for metagenes that mediate trait QTLs. 

```{r map_meta, fig.width = 10, fig.height = 7, results = "asis"}
cor.thresh = 0.6
qtl.thresh <- 6
lod.drop.thresh = 5

trait.map <- lapply(c(0,1), 
    function(x) scan1(genoprobs, adj.pheno[which(matched.covar[,1] == x),], cores = 4))

#par(mfrow = c(2,1))
#lodcol = 5
#plot(trait.map[[1]], map = map, lodcol = lodcol, main = colnames(trait.map[[1]])[lodcol])
#plot(trait.map[[2]], map = map, lodcol = lodcol, main = colnames(trait.map[[1]])[lodcol])

for(g in 1:length(do.profile)){
    group.name <- names(do.profile)[g]
    cat("###", group.name, "{.tabset .tabset-fade .tabset-pills}\n")

    #find metagenes with high correlations to any trait
    high.cor.idx <- which(abs(metagene.cor.mat[[g]]) >= cor.thresh)
    high.cor.row.col <- idx_to_row_col(high.cor.idx, nrow(metagene.cor.mat[[g]]))
    high.meta.names <- rownames(metagene.cor.mat[[g]])[unique(high.cor.row.col[,1])]

    high.metagene.row.col <- t(sapply(strsplit(unique(high.meta.names), " "), 
        function(x) as.numeric(c(x[1], x[3]))))
    high.metagenes.idx <- row_col_to_idx(high.metagene.row.col[,1], high.metagene.row.col[,2], 50)
    high.cor.profile <- rep(0, length = 2500)
    high.cor.profile[high.metagenes.idx] <- 1
    #par(mfrow = c(4,5), mar = c(0,0,2,0))
    #for(i in 1:ncol(adj.pheno)){
    #    high.cor.profile <- rep(0, length = 2500)
    #    high.cor.profile[high.metagenes.idx] <- metagene.cor.mat[[g]][high.metagenes.idx,i]
    #    plot_portrait(high.cor.profile, main = colnames(adj.pheno)[i])
    #}

    #kmeans.spots <- sapply(strsplit(high.meta.names, " "), function(x) x[4])
    #u_spots <- unique(kmeans.spots)
    meta.spot <- do.profile[[g]][,high.meta.names,drop=FALSE]
    adj.traits <- adjust(adj.pheno, meta.spot)
    #boxplot(adj.traits, las = 2)
    med.scan <- scan1(genoprobs, adj.traits, cores = 4)

    lod.drop <- trait.map[[g]] - med.scan
    all.max.min <- t(apply(lod.drop, 2, function(x) c(min(x), max(x))))
    drop_peaks <- find_peaks(lod.drop, map = map, threshold = lod.drop.thresh)    
    #nrow(drop_peaks)
    big.drop.traits <- drop_peaks[,1]
    big.drop.chr <- drop_peaks[,"chr"]
    u_drop_traits <- unique(big.drop.traits)

    for(tr in u_drop_traits){
        #test each metagene to see if any are responsible for the 
        #big lod drops seen earlier.
        cat("####", colnames(adj.pheno)[tr], "\n")
        if(is.interactive){quartz(width = 9, height = 7)}

        #plot the original scan for the trait overlayed
        #with the mediated scan in red
        par(mfrow = c(2,1), mar = c(4,4,2,4))
        plot(trait.map[[g]], map = map, lodcol = tr, main = colnames(adj.pheno)[tr])
        plot(med.scan, map = map, lodcol = tr, col = "red", add = TRUE)
        abline(h = qtl.thresh)
        #also plot the lod drops along the genome
        plot(lod.drop, map = map, lodcol = tr, ylim = all.max.min[tr,]*1.1, main = "LOD Drop")
        abline(h = lod.drop.thresh)

        #scan individual metagenes for eQTL on the same chromosome
        meta.scan <- scan1(genoprobs, meta.spot, cores = 4)
        meta.peaks <- find_peaks(meta.scan, map = map, threshold = qtl.thresh)
        drop.chr <- big.drop.chr[which(big.drop.traits == tr)]
        drop.chr.locale <- lapply(drop.chr, function(x) meta.peaks[which(meta.peaks[,"chr"] == x),1])
        names(drop.chr.locale) <- drop.chr

        #remove any elements of length zero i.e. where 
        #no transcripts have an eQTL > qtl.thresh
        has.meta <- which(sapply(drop.chr.locale, length) > 0)
        drop.chr.locale <- drop.chr.locale[has.meta]

        if(length(drop.chr.locale) == 0){next()}

        secondary.adj <- lapply(drop.chr.locale, function(x) adjust(adj.pheno, meta.spot[,x,drop=FALSE]))
        secondary.med <- lapply(1:length(secondary.adj), function(x) scan1(genoprobs[,drop.chr[x]], secondary.adj[[x]][,tr,drop=FALSE]))
        med.mats <- lapply(secondary.med, function(x) Reduce("cbind", x))

        if(is.interactive){quartz(width = 9, height = 7)}
        #plot the original scan for the trait overlayed
        #with the mediation of just a subset of the original metagenes in red
        for(sm in 1:length(secondary.med)){
            if(is.interactive){quartz(width = 9, height = 7)}
            layout(matrix(c(1,1,2,3), nrow = 2, byrow = TRUE), widths = c(0.4, 0.6))
            plot(trait.map[[g]], map = map, lodcol = tr, chr = drop.chr[[sm]], 
                main = paste(colnames(adj.pheno)[tr], "Mediation with Chr", names(drop.chr.locale)[sm] ,"subset"))
            plot(secondary.med[[sm]], map = map, col = "red", add = TRUE)
            spot.split.idx <- strsplit(colnames(meta.spot)[drop.chr.locale[[sm]]], " ")
            spot.row.col <- t(sapply(spot.split.idx, function(x) as.numeric(c(x[1], x[3]))))
            spot.idx <- row_col_to_idx(spot.row.col[,1], spot.row.col[,2], 50)
            mediator.mask <- rep(0, 2500)
            mediator.mask[spot.idx] <- 1
            plot_portrait(mediator.mask)
            meta.genes <- unlist(lapply(spot.idx, function(x) colnames(do.metagenes[[g]][x][[1]])))
            gene.enrich <- gost(meta.genes, organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME"))
            plot.enrichment(gene.enrich, num.term = 15, max.term.size = 2000, order.by = "p_value")
            #plot.text(paste(colnames(meta.spot)[drop.chr.locale[[sm]]], collapse = "\n"))
        } #end looping through chromosomes with mediated QTL
        cat("\n\n")
    } #end looping through traits

    #the spots seem too broad to effectively capture the high-correlation metagenes
    #spot.scan <- scan1(genoprobs, high.cor.spot.expr, cores = 4)
    #multilod.plot(spot.scan, map = map, lod.thresh = 6, chr.label.y = 0.5)
    cat("\n\n")
} #end looping through sexes
```

```{r mapping_extra, eval = FALSE}
find_peaks(group.scan, map = map, threshold = 6)

coef.scan <- scan1coef(genoprobs[,"X"], group.scores, kinship = K[[20]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

coef.scan <- scan1coef(genoprobs[,"11"], group.scores, kinship = K[[11]])
plot_coefCC(coef.scan, map = map, main = paste(map_group, "Chr X"))

```

```{r DO_portraits, eval = FALSE}
test <- sapply(do.metagenes[[group.locale]], rowMeans)
par(mfrow = c(5,5), mar = c(1,1,1,1))
for(i in 26:50){
    plot_portrait(test[i,])
}
```