
## Clustered Metagenes

The SOM clusters metagenes into clusters using k-means clustering. 
Neighboring metagenes are very highly correlated with each other. 
The clustering of metagenes allows us to reduce the dimentionality 
even more. 

The barplot below shows the number of genes in each of the K-means 
clusters. They are remarkably even in size.

```{r metagenes, fig.width = 9, fig.height = 5}
kmeta <- env$spot.list.kmeans$spots
cluster.sizes <- sapply(kmeta, function(x) length(x$genes))
barplot(cluster.sizes)
```

## Metagene Cluster Trends

look at correlation matrices for clustered metagenes. Here we find all the 
metagenes in a single cluster and decompose the cluster using SVD. We then
use the first principle component of the metagene matrix as the clustered
metagene. We calculated the correlation of each of these clustered metagenes
to the clinical traits, and scanned each. 

The correlations with the traits are very comparable to WGCNA eigengene
correlations with traits. The mapping is also par with mapping of other 
gene modules. Maybe a little better.

```{r metagene_clusters}
kmeans <- env$spot.list.kmeans
kmap <- kmeans$overview.map
u_k <- sort(unique(as.vector(kmap)))

k.gene.cor <- matrix(NA, nrow = length(u_k), ncol = ncol(adj.pheno))
rownames(k.gene.cor) <- names(kmeans$spots)
colnames(k.gene.cor) <- colnames(adj.pheno)

k.gene.mat <- matrix(NA, nrow = nrow(adj.pheno), ncol = length(u_k))
colnames(k.gene.mat) <- names(kmeans$spots)
rownames(k.gene.mat) <- rownames(adj.pheno)

common.ind <- intersect(rownames(adj.pheno), colnames(metagenes))
pheno.locale <- match(common.ind, rownames(adj.pheno))
metagene.locale <- match(common.ind, colnames(metagenes))
expr.locale <- match(common.ind, rownames(adj.expr))

for(k in 1:length(u_k)){
  #k.locale <- which(kmap == u_k[k])
  #k.genes <- metagenes[k.locale,]
  #k.gene <- plot.decomp(t(k.genes), plot.results = FALSE)$u[,1]

  k.genes <- kmeta[[k]]$genes
  k.gene <- plot.decomp(adj.expr[,k.genes], plot.results = FALSE)$u[,1]

  k.gene.mat[,k] <- k.gene
  #k.cor <- apply(adj.pheno[pheno.locale,], 2, function(x) cor(x, k.gene[metagene.locale], use = "pairwise.complete.obs"))
  k.cor <- apply(adj.pheno[pheno.locale,], 2, function(x) cor(x, k.gene[expr.locale], use = "pairwise.complete.obs"))
  k.gene.cor[k,] <- k.cor
}

pheatmap(k.gene.cor, display_numbers = TRUE)
```

The following code runs genetic mapping on metagenes. I tried 
mapping the metagenes directly and got very strange mapping.
So instead I am mapping the first principle compoenent of the 
expression matrix defined by the metagene clusters.

```{r kgene_scan, fig.width = 12, fig.height = 6}
metagene.scan.file <- here("Results", "SOM", paste0("Metagene.Scan.", tissue.name, ".RDS"))

if(!file.exists(metagene.scan.file)){
  k.gene.scan <- scan1(genoprobs, k.gene.mat)
  saveRDS(k.gene.scan, metagene.scan.file)
}else{
  k.gene.scan <- readRDS(metagene.scan.file)
}

par(xpd = NA)
multilod.plot(k.gene.scan, map = map, lod.thresh = 4, row.names = colnames(k.gene.scan))
par(xpd = TRUE)
```

## LOD traces for clustered metagenes {.tabset .tabset-fade .tabset-pills}

The following plots show the LOD trace for each metagene along with its
cis-trans eQTL plot to help show where the metagene QTLs might be coming
from.

```{r eqtl_fun}
#get a table with eQTL chromosome, position,
#and lod score.
get_eqtl_table <- function(gene.ids){
    gene.locale <- match(gene.ids, lod.table[[1]])
    gene.locale <- gene.locale[which(!is.na(gene.locale))]
    sub.table <- as.matrix(lod.table[gene.locale,])
    split.qtl.pos <- strsplit(sub.table[,2], "_")
    qtl.chr <- sapply(split.qtl.pos, function(x) x[1])
    qtl.pos <- as.numeric(sapply(split.qtl.pos, function(x) x[2]))/1e6
    final.table <- cbind(sub.table[,1], qtl.chr, qtl.pos)
    return(final.table)
}

#get a table with transcript chromosome and
#position
get_transcript_table <- function(gene.ids){
    trans.locale <- match(gene.ids, gene.info[[1]])
    sub.table <- as.matrix(gene.info[trans.locale,])
    final.table <- sub.table[,c(1,4,5)]
    return(final.table)
}
```

```{r plot_ind, fig.width = 12, fig.height = 8, results = "asis"}
lod.table <- expr.data$lod$additive
gene.info <- expr.data$annot.mrna

all.marker.chr <- unlist(sapply(1:length(map), function(x) rep(names(map)[x], length(map[[x]]))))
all.marker.pos <- as.numeric(unlist(map))
rel.marker.pos <- get_relative_genomic_position(all.marker.chr, all.marker.pos, map)
chr.col <- rep("darkgray", length(all.marker.chr))
chr.col[which(as.numeric(all.marker.chr) %% 2 == 1)] <- "#1f78b4"

layout.mat <- matrix(c(1,2), ncol = 1)

for(i in 1:ncol(k.gene.scan)){
  if(is.interactive){quartz(width = 10, height = 6)}
  cat("###", colnames(k.gene.scan)[i], "\n")
  layout(layout.mat, heights = c(0.5, 1))

  par(mar = c(0,2,2,2))
  #plot(k.gene.scan, map = map, lodcol = i)
  #quartz()
  plot(rel.marker.pos, k.gene.scan[,i], col = chr.col, type = "h", 
    axes = FALSE, ylab = "LOD Score")
  axis(2);abline(h = 6)

  cluster.genes <- kmeta[[i]]$genes
  #mod.eig <- plot.decomp(adj.expr[,cluster.genes], plot.results = FALSE)$u[,1,drop=FALSE]
  #rownames(mod.eig) <- rownames(adj.pheno)
  #eig.scan <- scan1(genoprobs, mod.eig)
  #plot(rel.marker.pos, eig.scan[,1], col = chr.col, type = "h", 
  #  axes = FALSE, ylab = "LOD Score")
  #axis(2);abline(h = 6)
 
  eqtl.table <- get_eqtl_table(cluster.genes)
  transcript.pos.table <- get_transcript_table(cluster.genes)
  lod.col <- colors.from.values(sqrt(as.numeric(eqtl.table[,3])), 
  col.scale = "blue", light.dark = "d")
  par(mar = c(2,2,0,2))
  plot.cistrans.table(eqtl.table, transcript.pos.table, map = map,
    add = FALSE, cex = 0.5, col = lod.col, label.cex = 0.7)  
  
  mtext(side = 3, outer = TRUE, text = paste("Metagene", colnames(k.gene.scan)[i]),
  line = -2.5)
  cat("\n\n")
}
```

## Peak Allele Effects {.tabset .tabset-fade .tabset-pills}

```{r scan_allele_effects}
metagene.peak.scan.file <- here("Results", "SOM", 
  paste0("Metagene.LOD.Scores.", tissue.name, ".RDS"))
metagene.peak.allele.file <- here("Results", "SOM", 
  paste0("Metagene.Allele.Effects.", tissue.name, ".RDS"))

peak.table <- find_peaks(k.gene.scan, map = map, threshold = 6)

if(!file.exists(metagene.peak.scan.file)){
  metagene.peak.lod.scans <- metagene.peak.allele.scans <- vector(mode = "list", length = nrow(peak.table))
  scan.names <- apply(peak.table, 1, function(x) paste0(x[2], "_Chr", str_trim(x[3])))
  for(i in 1:nrow(peak.table)){
    peak.chr <- as.vector(peak.table[i,"chr"])
    if(peak.chr != "X"){
      peak.pos <- as.numeric(peak.table[i,"pos"])
      peak.trait <- peak.table[i,"lodcolumn"]
      chr.locale <- which(names(map) == peak.chr)
      lod.scan <- scan1(genoprobs[,chr.locale], k.gene.mat[,peak.trait])
      metagene.peak.lod.scans[[i]] <- lod.scan
      coef.scan <- scan1coef(genoprobs[,chr.locale], k.gene.mat[,peak.trait])
      #coef.scan <- scan1blup(genoprobs[,chr.locale], k.gene.mat[,peak.trait])
      metagene.peak.allele.scans[[i]] <- coef.scan
    }
  }
  saveRDS(metagene.peak.allele.scans, metagene.peak.allele.file)
  saveRDS(metagene.peak.lod.scans, metagene.peak.scan.file)
}else{
  metagene.peak.allele.scans <- readRDS(metagene.peak.allele.file)
  metagene.peak.lod.scans <- readRDS(metagene.peak.scan.file)
}
```

```{r plot_scans, results = "asis", fig.height = 8, fig.widdth = 12}
for(i in 1:length(metagene.peak.lod.scans)){
  peak.chr <- peak.table[i,"chr"]
  peak.pos <- as.numeric(peak.table[i,"pos"])
  if(peak.chr != "X"){
    cat("###", peak.table[i,"lodcolumn"], "Chr", peak.chr, "\n")
    par(mfrow = c(2,1), mar = c(2,2,2,2))
    plot(metagene.peak.lod.scans[[i]], map = map);abline(v = peak.pos)
    plot_coefCC(metagene.peak.allele.scans[[i]], map = map)
    abline(v = peak.pos)
    cat("\n\n")
  }
}
```


```{r cluster_analysis, eval = FALSE}
cluster.name = "L"
p.thresh <- 20
cluster.locale <- which(names(kmeta) == cluster.name)
cluster.enrich <- kmeta[[cluster.locale]]$Fisher.p
log.p <- -log10(cluster.enrich)
par(mar = c(4, 18, 2, 2))
barplot(sort(log.p[which(log.p > p.thresh)]), horiz = TRUE, las = 2, xlab = "-log10(p)",
main = paste("Metagene", cluster.name))

cl.names <- lapply(kmeta, function(x) tail(sort(-log10(x$Fisher.p)), 1))

barplot(unlist(cl.names), las = 2, horiz = TRUE)
```

## Cluster Enrichments {.tabset .tabset-fade .tabset-pills}

The following plots show the locations and top associated biological
processes of each metagene.

```{r cluster_enrichments, results = "asis", fig.width = 10, fig.height = 6}
layout.matrix <- matrix(c(1,2,0,2), byrow = TRUE, ncol = 2)
for(k in 1:length(kmeta)){
  cat("### Cluster", names(kmeta)[k], "\n")
  if(is.interactive){quartz(width = 10, height = 6)}
  layout(layout.matrix, widths = c(0.7, 1), heights = c(1,0.5))
  par(mar = c(0,2,2,0))
  plot_portrait(kmeta[[k]]$mask, rotate = 3)
  enrich <- sort(-log10(kmeta[[k]]$Fisher.p))
  top.enrich <- tail(enrich, 20)
  par(mar = c(4, 18, 4, 4))
  barplot(top.enrich, horiz = TRUE, las = 2, main = paste("Metagene", names(kmeta)[k]))
  cat("\n\n")
}
```